<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Collision removal functionality • ISAnalytics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Collision removal functionality">
<meta property="og:description" content="ISAnalytics">
<meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISAnalytics</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">1.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/aggregate_function_usage.html">Working with aggregate functions</a>
    </li>
    <li>
      <a href="../articles/collision_removal.html">Collision removal functionality</a>
    </li>
    <li>
      <a href="../articles/how_to_import_functions.html">How to use import functions</a>
    </li>
    <li>
      <a href="../articles/sharing_analyses.html">Sharing analyses with ISAnalytics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://calabrialab.github.io/ISAnalytics/">
    <span class="fas fa-code-branch"></span>
     
    RELEASE
  </a>
</li>
<li>
  <a href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="collision_removal_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Collision removal functionality</h1>
                        <h4 class="author">Giulia Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget, Via Olgettina 60, 20132 Milano - Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 class="date">7 October 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/master/vignettes/collision_removal.Rmd"><code>vignettes/collision_removal.Rmd</code></a></small>
      <div class="hidden name"><code>collision_removal.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
</div>
<div id="installation-and-options" class="section level1">
<h1 class="hasAnchor">
<a href="#installation-and-options" class="anchor"></a>Installation and options</h1>
<p><code>ISAnalytics</code> can be installed quickly in different ways:</p>
<ul>
<li>You can install it via <a href="http://bioconductor.org">Bioconductor</a>
</li>
<li>You can install it via GitHub using the package <code>devtools</code>
</li>
</ul>
<p>There are always 2 versions of the package active:</p>
<ul>
<li>
<code>RELEASE</code> is the latest stable version</li>
<li>
<code>DEVEL</code> is the development version, it is the most up-to-date version where all new features are introduced</li>
</ul>
<div id="installation-from-bioconductor" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-from-bioconductor" class="anchor"></a>Installation from bioconductor</h2>
<p>RELEASE version:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span></code></pre></div>
<p>DEVEL version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="co"># The following initializes usage of Bioc devel</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span></code></pre></div>
</div>
<div id="installation-from-github" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-from-github" class="anchor"></a>Installation from GitHub</h2>
<p>RELEASE:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">devtools</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"RELEASE_3_13"</span>,
                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,
                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>DEVEL:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">devtools</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"master"</span>,
                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,
                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="setting-options" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-options" class="anchor"></a>Setting options</h2>
<p><code>ISAnalytics</code> has a verbose option that allows some functions to print additional information to the console while they’re executing. To disable this feature do:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DISABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># ENABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Some functions also produce report in a user-friendly HTML format, to set this feature:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DISABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.reports"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># ENABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.reports"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://calabrialab.github.io/ISAnalytics">ISAnalytics</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span></code></pre></div>
</div>
<div id="what-is-a-collision-and-why-should-you-care" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-a-collision-and-why-should-you-care" class="anchor"></a>What is a collision and why should you care?</h2>
<p>We’re not going into too much detail here, but we’re going to explain in a very simple way what a “collision” is and how the function in this package deals with them.</p>
<p>We say that an integration (aka a unique combination of chromosome, integration locus and strand) is a <em>collision</em> if this combination is shared between different independent samples: an independent sample is a unique combination of <code>ProjectID</code> and <code>SubjectID</code> (where subjects usually represent patients). The reason behind this is that it’s highly improbable to observe the very same integration in two different subjects and this phenomenon might be an indicator of some kind of contamination in the sequencing phase or in PCR phase, for this reason we might want to exclude such contamination from our analysis. <code>ISAnalytics</code> provides a function that processes the imported data for the removal or reassignment of these “problematic” integrations, <code><a href="../reference/remove_collisions.html">remove_collisions()</a></code>.</p>
<p>The processing is done using the sequence count value, so the corresponding matrix is needed for this operation.</p>
</div>
<div id="the-logic-behind-the-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-logic-behind-the-function" class="anchor"></a>The logic behind the function</h2>
<p>The <code><a href="../reference/remove_collisions.html">remove_collisions()</a></code> function follows several logical steps to decide whether an integration is a collision and if it is it decides whether to re-assign it or remove it entirely based on different criterias.</p>
<div id="identifying-the-collisions" class="section level3">
<h3 class="hasAnchor">
<a href="#identifying-the-collisions" class="anchor"></a>Identifying the collisions</h3>
<p>As we said before, a collision is a triplet made of <code>chr</code>, <code>integration locus</code> and <code>strand</code>, which is shared between different independent samples, aka a pair made of <code>ProjectID</code> and <code>SubjectID</code>. The function uses the information stored in the association file to assess which independent samples are present and counts the number of independent samples for each integration: those who have a count &gt; 1 are considered collisions.</p>
<table class="table">
<caption>Example of collisions: the same integration (1, 123454, +) is found in 2 different independent samples ((SUBJ01, PJ01) &amp; (SUBJ02, PJ01))</caption>
<colgroup>
<col width="4%">
<col width="21%">
<col width="8%">
<col width="10%">
<col width="29%">
<col width="12%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="right">seqCount</th>
<th align="left">CompleteAmplificationID</th>
<th align="left">SubjectID</th>
<th align="left">ProjectID</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">123454</td>
<td align="left">+</td>
<td align="right">653</td>
<td align="left">SAMPLE1</td>
<td align="left">SUBJ01</td>
<td align="left">PJ01</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">123454</td>
<td align="left">+</td>
<td align="right">456</td>
<td align="left">SAMPLE2</td>
<td align="left">SUBJ02</td>
<td align="left">PJ01</td>
</tr>
</tbody>
</table>
</div>
<div id="re-assign-vs-remove" class="section level3">
<h3 class="hasAnchor">
<a href="#re-assign-vs-remove" class="anchor"></a>Re-assign vs remove</h3>
<p>Once the collisions are identified, the function follows 3 steps where it tries to re-assign the combination to a single independent sample. The criterias are:</p>
<ol style="list-style-type: decimal">
<li>Compare dates: if it’s possible to have an absolute ordering on dates, the integration is re-assigned to the sample that has the earliest date. If two samples share the same date it’s impossible to decide, so the next criteria is tested</li>
<li>Compare replicate number: if a sample has the same integration in more than one replicate, it’s more probable the integration is not an artifact. If it’s possible to have an absolute ordering, the collision is re-assigned to the sample whose grouping is largest</li>
<li>Compare the sequence count value: if the previous criteria wasn’t sufficient to make a decision, for each group of independent samples it’s evaluated the sum of the sequence count value - for each group there is a cumulative value of the sequence count and this is compared to the value of other groups. If there is a single group which has a ratio n times bigger than other groups, this one is chosen for re-assignment. The factor n is passed as a parameter in the function (<code>reads_ratio</code>), the default value is 10.</li>
</ol>
<p>If none of the criterias were sufficient to make a decision, the integration is simply removed from the matrix.</p>
</div>
</div>
</div>
<div id="usage" class="section level1">
<h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h1>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">data</span>(<span class="st">"integration_matrices"</span>, <span class="dt">package =</span> <span class="st">"ISAnalytics"</span>)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="kw">data</span>(<span class="st">"association_file"</span>, <span class="dt">package =</span> <span class="st">"ISAnalytics"</span>)</span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co">## Multi quantification matrix</span></span>
<span id="cb8-4"><a href="#cb8-4"></a>no_coll &lt;-<span class="st"> </span><span class="kw">remove_collisions</span>(<span class="dt">x =</span> integration_matrices,</span>
<span id="cb8-5"><a href="#cb8-5"></a>                             <span class="dt">association_file =</span> association_file,</span>
<span id="cb8-6"><a href="#cb8-6"></a>                             <span class="dt">report_path =</span> <span class="ot">NULL</span>)</span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co">#&gt; Identifying collisions...</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="co">#&gt; Processing collisions...</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="co">#&gt; </span></span>
<span id="cb8-10"><a href="#cb8-10"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</span>
<span id="cb8-12"><a href="#cb8-12"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="co">#&gt; Finished!</span></span>
<span id="cb8-15"><a href="#cb8-15"></a><span class="co">## Matrix list</span></span>
<span id="cb8-16"><a href="#cb8-16"></a>separated &lt;-<span class="st"> </span><span class="kw">separate_quant_matrices</span>(integration_matrices)</span>
<span id="cb8-17"><a href="#cb8-17"></a>no_coll_list &lt;-<span class="st"> </span><span class="kw">remove_collisions</span>(<span class="dt">x =</span> separated,</span>
<span id="cb8-18"><a href="#cb8-18"></a>                             <span class="dt">association_file =</span> association_file,</span>
<span id="cb8-19"><a href="#cb8-19"></a>                             <span class="dt">report_path =</span> <span class="ot">NULL</span>)</span>
<span id="cb8-20"><a href="#cb8-20"></a><span class="co">#&gt; Identifying collisions...</span></span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="co">#&gt; Processing collisions...</span></span>
<span id="cb8-22"><a href="#cb8-22"></a><span class="co">#&gt; </span></span>
<span id="cb8-23"><a href="#cb8-23"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb8-24"><a href="#cb8-24"></a><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</span>
<span id="cb8-25"><a href="#cb8-25"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb8-26"><a href="#cb8-26"></a><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</span>
<span id="cb8-27"><a href="#cb8-27"></a><span class="co">#&gt; Finished!</span></span>
<span id="cb8-28"><a href="#cb8-28"></a><span class="co">## Only sequence count</span></span>
<span id="cb8-29"><a href="#cb8-29"></a>no_coll_single &lt;-<span class="st"> </span><span class="kw">remove_collisions</span>(<span class="dt">x =</span> separated<span class="op">$</span>seqCount,</span>
<span id="cb8-30"><a href="#cb8-30"></a>                             <span class="dt">association_file =</span> association_file,</span>
<span id="cb8-31"><a href="#cb8-31"></a>                             <span class="dt">quant_cols =</span> <span class="kw">c</span>(<span class="dt">seqCount =</span> <span class="st">"Value"</span>),</span>
<span id="cb8-32"><a href="#cb8-32"></a>                             <span class="dt">report_path =</span> <span class="ot">NULL</span>)</span>
<span id="cb8-33"><a href="#cb8-33"></a><span class="co">#&gt; Identifying collisions...</span></span>
<span id="cb8-34"><a href="#cb8-34"></a><span class="co">#&gt; Processing collisions...</span></span>
<span id="cb8-35"><a href="#cb8-35"></a><span class="co">#&gt; </span></span>
<span id="cb8-36"><a href="#cb8-36"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb8-37"><a href="#cb8-37"></a><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</span>
<span id="cb8-38"><a href="#cb8-38"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb8-39"><a href="#cb8-39"></a><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</span>
<span id="cb8-40"><a href="#cb8-40"></a><span class="co">#&gt; Finished!</span></span></code></pre></div>
<p>Important notes on the association file:</p>
<ul>
<li>You have to be sure your association file is properly filled out. The function requires you to specify a date column (by default “SequencingDate”), you have to ensure this column doesn’t contain NA values or incorrect values.</li>
</ul>
<p>The function accepts different inputs, namely:</p>
<ul>
<li>A multi-quantification matrix: this is always the recommended approach</li>
<li>A named list of matrices where names are quantification types in <code><a href="../reference/quantification_types.html">quantification_types()</a></code>
</li>
<li>The single sequence count matrix: this is not the recommended approach since it requires a realignment step for other quantification matrices if you have them.</li>
</ul>
<p>If the option <code>ISAnalytics.reports</code> is active, an interactive report in HTML format will be produced at the specified path.</p>
</div>
<div id="re-align-other-matrices" class="section level1">
<h1 class="hasAnchor">
<a href="#re-align-other-matrices" class="anchor"></a>Re-align other matrices</h1>
<p>If you’ve given as input the standalone sequence count matrix to <code><a href="../reference/remove_collisions.html">remove_collisions()</a></code>, to realign other matrices you have to call the function <code><a href="../reference/realign_after_collisions.html">realign_after_collisions()</a></code>, passing as input the processed sequence count matrix and the named list of other matrices to realign. <strong>NOTE: the names in the list must be quantification types.</strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">other_realigned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/realign_after_collisions.html">realign_after_collisions</a></span><span class="op">(</span>
  sc_matrix <span class="op">=</span> <span class="va">no_coll_single</span>,
  other_matrices <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>fragmentEstimate <span class="op">=</span> <span class="va">separated</span><span class="op">$</span><span class="va">fragmentEstimate</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p><code>R</code> session information.</p>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  setting  value                       
#&gt;  version  R version 4.1.1 (2021-08-10)
#&gt;  os       macOS Catalina 10.15.7      
#&gt;  system   x86_64, darwin17.0          
#&gt;  ui       X11                         
#&gt;  language (EN)                        
#&gt;  collate  en_US.UTF-8                 
#&gt;  ctype    en_US.UTF-8                 
#&gt;  tz       UTC                         
#&gt;  date     2021-10-07                  
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  package      * version date       lib source        
#&gt;  BiocManager    1.30.16 2021-06-15 [1] CRAN (R 4.1.0)
#&gt;  BiocParallel   1.26.2  2021-08-22 [1] Bioconductor  
#&gt;  BiocStyle    * 2.20.2  2021-06-17 [1] Bioconductor  
#&gt;  bookdown       0.24    2021-09-02 [1] CRAN (R 4.1.0)
#&gt;  bslib          0.3.1   2021-10-06 [1] CRAN (R 4.1.1)
#&gt;  cachem         1.0.6   2021-08-19 [1] CRAN (R 4.1.0)
#&gt;  cli            3.0.1   2021-07-17 [1] CRAN (R 4.1.0)
#&gt;  colorspace     2.0-2   2021-06-24 [1] CRAN (R 4.1.0)
#&gt;  crayon         1.4.1   2021-02-08 [1] CRAN (R 4.1.0)
#&gt;  data.table     1.14.2  2021-09-27 [1] CRAN (R 4.1.0)
#&gt;  desc           1.4.0   2021-09-28 [1] CRAN (R 4.1.0)
#&gt;  digest         0.6.28  2021-09-23 [1] CRAN (R 4.1.0)
#&gt;  dplyr          1.0.7   2021-06-18 [1] CRAN (R 4.1.0)
#&gt;  ellipsis       0.3.2   2021-04-29 [1] CRAN (R 4.1.0)
#&gt;  evaluate       0.14    2019-05-28 [1] CRAN (R 4.1.0)
#&gt;  fansi          0.5.0   2021-05-25 [1] CRAN (R 4.1.0)
#&gt;  fastmap        1.1.0   2021-01-25 [1] CRAN (R 4.1.0)
#&gt;  fs             1.5.0   2020-07-31 [1] CRAN (R 4.1.0)
#&gt;  generics       0.1.0   2020-10-31 [1] CRAN (R 4.1.0)
#&gt;  ggplot2        3.3.5   2021-06-25 [1] CRAN (R 4.1.0)
#&gt;  ggrepel        0.9.1   2021-01-15 [1] CRAN (R 4.1.0)
#&gt;  glue           1.4.2   2020-08-27 [1] CRAN (R 4.1.0)
#&gt;  gtable         0.3.0   2019-03-25 [1] CRAN (R 4.1.0)
#&gt;  highr          0.9     2021-04-16 [1] CRAN (R 4.1.0)
#&gt;  hms            1.1.1   2021-09-26 [1] CRAN (R 4.1.0)
#&gt;  htmltools      0.5.2   2021-08-25 [1] CRAN (R 4.1.0)
#&gt;  httr           1.4.2   2020-07-20 [1] CRAN (R 4.1.0)
#&gt;  ISAnalytics  * 1.3.6   2021-10-07 [1] Bioconductor  
#&gt;  jquerylib      0.1.4   2021-04-26 [1] CRAN (R 4.1.0)
#&gt;  jsonlite       1.7.2   2020-12-09 [1] CRAN (R 4.1.0)
#&gt;  knitr          1.36    2021-09-29 [1] CRAN (R 4.1.0)
#&gt;  lifecycle      1.0.1   2021-09-24 [1] CRAN (R 4.1.0)
#&gt;  lubridate      1.7.10  2021-02-26 [1] CRAN (R 4.1.0)
#&gt;  magrittr     * 2.0.1   2020-11-17 [1] CRAN (R 4.1.0)
#&gt;  memoise        2.0.0   2021-01-26 [1] CRAN (R 4.1.0)
#&gt;  munsell        0.5.0   2018-06-12 [1] CRAN (R 4.1.0)
#&gt;  pillar         1.6.3   2021-09-26 [1] CRAN (R 4.1.0)
#&gt;  pkgconfig      2.0.3   2019-09-22 [1] CRAN (R 4.1.0)
#&gt;  pkgdown        1.6.1   2020-09-12 [1] CRAN (R 4.1.0)
#&gt;  plyr           1.8.6   2020-03-03 [1] CRAN (R 4.1.0)
#&gt;  purrr          0.3.4   2020-04-17 [1] CRAN (R 4.1.0)
#&gt;  R6             2.5.1   2021-08-19 [1] CRAN (R 4.1.0)
#&gt;  ragg           1.1.3   2021-06-09 [1] CRAN (R 4.1.0)
#&gt;  Rcapture       1.4-3   2019-12-16 [1] CRAN (R 4.1.0)
#&gt;  Rcpp           1.0.7   2021-07-07 [1] CRAN (R 4.1.0)
#&gt;  readr          2.0.2   2021-09-27 [1] CRAN (R 4.1.0)
#&gt;  RefManageR   * 1.3.0   2020-11-13 [1] CRAN (R 4.1.0)
#&gt;  rlang          0.4.11  2021-04-30 [1] CRAN (R 4.1.0)
#&gt;  rmarkdown      2.11    2021-09-14 [1] CRAN (R 4.1.0)
#&gt;  rprojroot      2.0.2   2020-11-15 [1] CRAN (R 4.1.0)
#&gt;  sass           0.4.0   2021-05-12 [1] CRAN (R 4.1.0)
#&gt;  scales         1.1.1   2020-05-11 [1] CRAN (R 4.1.0)
#&gt;  sessioninfo  * 1.1.1   2018-11-05 [1] CRAN (R 4.1.0)
#&gt;  stringi        1.7.5   2021-10-04 [1] CRAN (R 4.1.1)
#&gt;  stringr        1.4.0   2019-02-10 [1] CRAN (R 4.1.0)
#&gt;  systemfonts    1.0.2   2021-05-11 [1] CRAN (R 4.1.0)
#&gt;  textshaping    0.3.5   2021-06-09 [1] CRAN (R 4.1.0)
#&gt;  tibble         3.1.5   2021-09-30 [1] CRAN (R 4.1.0)
#&gt;  tidyr          1.1.4   2021-09-27 [1] CRAN (R 4.1.0)
#&gt;  tidyselect     1.1.1   2021-04-30 [1] CRAN (R 4.1.0)
#&gt;  tzdb           0.1.2   2021-07-20 [1] CRAN (R 4.1.0)
#&gt;  utf8           1.2.2   2021-07-24 [1] CRAN (R 4.1.0)
#&gt;  vctrs          0.3.8   2021-04-29 [1] CRAN (R 4.1.0)
#&gt;  withr          2.4.2   2021-04-18 [1] CRAN (R 4.1.0)
#&gt;  xfun           0.26    2021-09-14 [1] CRAN (R 4.1.0)
#&gt;  xml2           1.3.2   2020-04-23 [1] CRAN (R 4.1.0)
#&gt;  yaml           2.2.1   2020-02-01 [1] CRAN (R 4.1.0)
#&gt;  zip            2.2.0   2021-05-31 [1] CRAN (R 4.1.0)
#&gt; 
#&gt; [1] /Users/runner/work/_temp/Library
#&gt; [2] /Library/Frameworks/R.framework/Versions/4.1/Resources/library</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.13/BiocStyle">BiocStyle</a></em> (Oleś, 2021) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (Xie, 2021) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> (Allaire, Xie, McPherson, Luraschi, Ushey, Atkins, Wickham, Cheng, Chang, and Iannone, 2021) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR">RefManageR</a></em> (McLean, 2017).</p>
<p>
<a id="bib-allaire2021rmarkdown"></a><a href="#cite-allaire2021rmarkdown">[1]</a><cite> J. Allaire, Y. Xie, J. McPherson, et al. <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.11. 2021. URL: <a href="https://github.com/rstudio/rmarkdown">https://github.com/rstudio/rmarkdown</a>.</cite>
</p>
<p>
<a id="bib-mclean2017refmanager"></a><a href="#cite-mclean2017refmanager">[2]</a><cite> M. W. McLean. “RefManageR: Import and Manage BibTeX and BibLaTeX References in R”. In: <em>The Journal of Open Source Software</em> (2017). DOI: <a href="https://doi.org/10.21105/joss.00338">10.21105/joss.00338</a>.</cite>
</p>
<p>
<a id="bib-ole2021biocstyle"></a><a href="#cite-ole2021biocstyle">[3]</a><cite> A. Oleś. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.20.2. 2021. URL: <a href="https://github.com/Bioconductor/BiocStyle">https://github.com/Bioconductor/BiocStyle</a>.</cite>
</p>
<p>
<a id="bib-xie2021knitr"></a><a href="#cite-xie2021knitr">[4]</a><cite> Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.36. 2021. URL: <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.</cite>
</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
