<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>How to use import functions • ISAnalytics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="How to use import functions">
<meta property="og:description" content="ISAnalytics">
<meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISAnalytics</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">1.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/aggregate_function_usage.html">Working with aggregate functions</a>
    </li>
    <li>
      <a href="../articles/collision_removal.html">Collision removal functionality</a>
    </li>
    <li>
      <a href="../articles/how_to_import_functions.html">How to use import functions</a>
    </li>
    <li>
      <a href="../articles/report_system.html">ISAnalytics report system</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://calabrialab.github.io/ISAnalytics/">
    <span class="fas fa-code-branch"></span>
     
    RELEASE
  </a>
</li>
<li>
  <a href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="how_to_import_functions_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="how_to_import_functions_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="how_to_import_functions_files/jquery-3.5.1/jquery.min.js"></script><link href="how_to_import_functions_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="how_to_import_functions_files/datatables-binding-0.18/datatables.js"></script><link href="how_to_import_functions_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="how_to_import_functions_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="how_to_import_functions_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="how_to_import_functions_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="how_to_import_functions_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>How to use import functions</h1>
                        <h4 class="author">Giulia Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget, Via Olgettina 60, 20132 Milano - Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 class="date">3 August 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/master/vignettes/how_to_import_functions.Rmd"><code>vignettes/how_to_import_functions.Rmd</code></a></small>
      <div class="hidden name"><code>how_to_import_functions.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-isanalytics-import-functions-family" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-isanalytics-import-functions-family" class="anchor"></a>Introduction to <code>ISAnalytics</code> import functions family</h1>
<p>In this vignette we’re going to explain more in detail how functions of the import family should be used, the most common workflows to follow and more.</p>
</div>
<div id="installation-and-options" class="section level1">
<h1 class="hasAnchor">
<a href="#installation-and-options" class="anchor"></a>Installation and options</h1>
<p><code>ISAnalytics</code> can be installed quickly in different ways:</p>
<ul>
<li>You can install it via <a href="http://bioconductor.org">Bioconductor</a>
</li>
<li>You can install it via GitHub using the package <code>devtools</code>
</li>
</ul>
<p>There are always 2 versions of the package active:</p>
<ul>
<li>
<code>RELEASE</code> is the latest stable version</li>
<li>
<code>DEVEL</code> is the development version, it is the most up-to-date version where all new features are introduced</li>
</ul>
<div id="installation-from-bioconductor" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-from-bioconductor" class="anchor"></a>Installation from bioconductor</h2>
<p>RELEASE version:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span></code></pre></div>
<p>DEVEL version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="co"># The following initializes usage of Bioc devel</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span></code></pre></div>
</div>
<div id="installation-from-github" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-from-github" class="anchor"></a>Installation from GitHub</h2>
<p>RELEASE:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">devtools</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"RELEASE_3_13"</span>,
                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,
                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>DEVEL:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">devtools</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"master"</span>,
                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,
                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="setting-options" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-options" class="anchor"></a>Setting options</h2>
<p><code>ISAnalytics</code> has a verbose option that allows some functions to print additional information to the console while they’re executing. To disable this feature do:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DISABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># ENABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Some functions also produce report in a user-friendly HTML format, to set this feature:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DISABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.reports"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># ENABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.reports"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://calabrialab.github.io/ISAnalytics">ISAnalytics</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span></code></pre></div>
</div>
<div id="designed-to-work-with-vispa2-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#designed-to-work-with-vispa2-pipeline" class="anchor"></a>Designed to work with VISPA2 pipeline</h2>
<p>The vast majority of the functions included in this package is designed to work in combination with VISPA2 pipeline (Giulio Spinozzi Andrea Calabria, 2017). If you don’t know what it is, we strongly recommend you to take a look at these links:</p>
<ul>
<li>Article: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5702242/">VISPA2: Article</a><br>
</li>
<li>BitBucket Wiki: <a href="https://bitbucket.org/andreacalabria/vispa2/wiki/Home">VISPA2 Wiki</a>
</li>
</ul>
</div>
<div id="file-system-structure-generated" class="section level2">
<h2 class="hasAnchor">
<a href="#file-system-structure-generated" class="anchor"></a>File system structure generated</h2>
<p>VISPA2 produces a standard file system structure starting from a folder you specify as your workbench or root. The structure always follows this schema:</p>
<ul>
<li>root/
<ul>
<li>Optional intermediate folders
<ul>
<li>Projects (PROJECTID)
<ul>
<li>bam</li>
<li>bcmuxall</li>
<li>bed</li>
<li>iss
<ul>
<li>Pools (concatenatePoolIDSeqRun)</li>
</ul>
</li>
<li>quality</li>
<li>quantification
<ul>
<li>Pools (concatenatePoolIDSeqRun)</li>
</ul>
</li>
<li>report</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>Most of the functions implemented expect a standard file system structure as the one described above.</p>
</div>
</div>
<div id="notation" class="section level1">
<h1 class="hasAnchor">
<a href="#notation" class="anchor"></a>Notation</h1>
<p>We call an <em>“integration matrix”</em> a tabular structure characterized by:</p>
<ul>
<li>3 mandatory columns of genomic features that characterize a viral insertion site in the genome: <code>chr</code>, <code>integration_locus</code> and <code>strand</code>
</li>
<li>2 (optional) annotation columns: <code>GeneName</code> and <code>GeneStrand</code>
</li>
<li>A variable number n of sample columns containing the quantification of the corresponding integration site</li>
</ul>
<div id="htmlwidget-25543a7bd466120506b8" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-25543a7bd466120506b8">{"x":{"filter":"none","data":[["1","6","7"],[12324,657532,657532],["+","+","+"],["NFATC3","LOC100507487","EDIL3"],["+","+","-"],[4553,76,null],[5345,545,56],[null,5,null]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>chr<\/th>\n      <th>integration_locus<\/th>\n      <th>strand<\/th>\n      <th>GeneName<\/th>\n      <th>GeneStrand<\/th>\n      <th>exp1<\/th>\n      <th>exp2<\/th>\n      <th>exp3<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","scrollX":120,"columnDefs":[{"className":"dt-right","targets":[1,5,6,7]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>The package uses a more compact form of these matrices, limiting the amount of NA values and optimizing time and memory consumption. For more info on this take a look at: <a href="https://r4ds.had.co.nz/tidy-data.html">Tidy data</a></p>
<p>While integration matrices contain the actual data, we also need associated sample metadata to perform the vast majority of the analyses. <code>ISAnalytics</code> expects the metadata to be contained in a so called <em>“association file”</em>, which is a simple tabular file with a set of standard column headers.</p>
<p>To generate a blank association file you can use the function <code>generate_blank_association_file</code>. You can also view the standard column names with <code><a href="../reference/association_file_columns.html">association_file_columns()</a></code>.</p>
</div>
<div id="metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#metadata" class="anchor"></a>Importing metadata</h1>
<p>To import metadata we use <code><a href="../reference/import_association_file.html">import_association_file()</a></code>. This function is not only responsible for reading the file into the R environment as a data frame, but it is capable to perform a file system alignment operation, that is, for each project and pool contained in the file, it scans the file system starting from the provided root to check if the corresponding folders (contained in the appropriate column) can be found. Remember that to work properly, this operation expects a standard folder structure, such as the one provided by VISPA2. This function also produces an interactive HTML report, to know more about this feature see <code>vignette(report_system)</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fs_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"fs.zip"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span>
<span class="va">root</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unzip_file_system.html">unzip_file_system</a></span><span class="op">(</span><span class="va">fs_path</span>, <span class="st">"fs"</span><span class="op">)</span>
<span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_options.html">with_options</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>ISAnalytics.reports <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, code <span class="op">=</span> <span class="op">{</span>
  <span class="va">af_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"asso.file.tsv.gz"</span>, 
                         package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span>
  <span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_association_file.html">import_association_file</a></span><span class="op">(</span><span class="va">af_path</span>, root <span class="op">=</span> <span class="va">root</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="co">#&gt; *** Association file import summary ***</span>
<span class="co">#&gt; ℹ For detailed report please set option 'ISAnalytics.reports' to TRUE</span>
<span class="co">#&gt; * Parsing problems detected: FALSE</span>
<span class="co">#&gt; * Date parsing problems: FALSE</span>
<span class="co">#&gt; * Column problems detected: FALSE</span>
<span class="co">#&gt; * NAs found in important columns: FALSE</span>
<span class="co">#&gt; * File system alignment: no problems detected</span></code></pre></div>
<div id="htmlwidget-8a30395b3e765690457b" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8a30395b3e765690457b">{"x":{"filter":"none","data":[["PJ01","PJ01","PJ01","PJ01","PJ01","PJ01"],["ET#382.46","ET#381.40","ET#381.9","ET#381.71","ET#381.2","ET#382.28"],["POOL01","POOL01","POOL01","POOL01","POOL01","POOL01"],["LTR75LC38","LTR53LC32","LTR83LC66","LTR27LC94","LTR69LC52","LTR37LC2"],["PT001","PT001","PT001","PT001","PT001","PT001"],["lenti","lenti","lenti","lenti","lenti","lenti"],["GLOBE","GLOBE","GLOBE","GLOBE","GLOBE","GLOBE"],[null,null,null,null,null,null],["PB","BM","BM","BM","PB","BM"],["0060","0180","0180","0180","0180","0060"],["SONIC","SONIC","SONIC","SONIC","SONIC","SONIC"],["SLiM","SLiM","SLiM","SLiM","SLiM","SLiM"],["LTR75LC38","LTR53LC32","LTR83LC66","LTR27LC94","LTR69LC52","LTR37LC2"],[null,null,null,null,null,null],["MNC","MNC","MNC","MNC","MNC","MNC"],["LTR75.LC38","LTR53.LC32","LTR83.LC66","LTR27.LC94","LTR69.LC52","LTR37.LC2"],[null,null,null,null,null,null],["HiSeq","HiSeq","HiSeq","HiSeq","HiSeq","HiSeq"],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],["PT001-103","PT001-81","PT001-81","PT001-81","PT001-74","PT001-107"],[3,2,1,3,1,2],["2016-03-16","2016-07-15","2016-07-15","2016-07-15","2016-07-15","2016-03-16"],[23.184,181.44,181.44,181.44,23.058,171.36],[null,null,null,null,null,null],[null,null,null,null,null,null],["2016-11-02","2016-11-02","2016-11-02","2016-11-02","2016-11-02","2016-11-02"],["2016-11-02","2016-11-02","2016-11-02","2016-11-02","2016-11-02","2016-11-02"],["ET#380.46","ET#379.40","ET#379.9","ET#379.71","ET#379.2","ET#380.28"],["2016-11-02","2016-11-02","2016-11-02","2016-11-02","2016-11-02","2016-11-02"],[null,null,null,null,null,null],[null,null,null,null,null,null],["ET#382.46","ET#381.40","ET#381.9","ET#381.71","ET#381.2","ET#382.28"],["2016-11-03","2016-11-03","2016-11-03","2016-11-03","2016-11-03","2016-11-03"],["2016-11-07","2016-11-07","2016-11-07","2016-11-07","2016-11-07","2016-11-07"],["2016-11-15","2016-11-15","2016-11-15","2016-11-15","2016-11-15","2016-11-15"],[0.3,0.27,0.27,0.27,0.24,0.42],["hg19","hg19","hg19","hg19","hg19","hg19"],[1,1,1,1,1,1],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],["POOL01-1","POOL01-1","POOL01-1","POOL01-1","POOL01-1","POOL01-1"],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],["PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC","PJ01_POOL01_LTR53LC32_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC","PJ01_POOL01_LTR83LC66_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC","PJ01_POOL01_LTR27LC94_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC","PJ01_POOL01_LTR69LC52_PT001_PT001-74_lenti_GLOBE_PB_1_SLiM_0180_MNC","PJ01_POOL01_LTR37LC2_PT001_PT001-107_lenti_GLOBE_BM_1_SLiM_0060_MNC"],["ID00000000000000007433","ID00000000000000007340","ID00000000000000007310","ID00000000000000007370","ID00000000000000007303","ID00000000000000007417"],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null],["/PJ01","/PJ01","/PJ01","/PJ01","/PJ01","/PJ01"],[null,null,null,null,null,null],[null,null,null,null,null,null],["02","06","06","06","06","02"],["01","01","01","01","01","01"],[23.18,181.44,181.44,181.44,23.06,171.36],["/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01"],["/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/quantification/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/quantification/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/quantification/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/quantification/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/quantification/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/quantification/POOL01-1"],["/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/iss/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/iss/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/iss/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/iss/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/iss/POOL01-1","/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpOU5jQI/fs/PJ01/iss/POOL01-1"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>ProjectID<\/th>\n      <th>FUSIONID<\/th>\n      <th>PoolID<\/th>\n      <th>TagSequence<\/th>\n      <th>SubjectID<\/th>\n      <th>VectorType<\/th>\n      <th>VectorID<\/th>\n      <th>ExperimentID<\/th>\n      <th>Tissue<\/th>\n      <th>TimePoint<\/th>\n      <th>DNAFragmentation<\/th>\n      <th>PCRMethod<\/th>\n      <th>TagIDextended<\/th>\n      <th>Keywords<\/th>\n      <th>CellMarker<\/th>\n      <th>TagID<\/th>\n      <th>NGSProvider<\/th>\n      <th>NGSTechnology<\/th>\n      <th>ConverrtedFilesDir<\/th>\n      <th>ConverrtedFilesName<\/th>\n      <th>SourceFileFolder<\/th>\n      <th>SourceFileNameR1<\/th>\n      <th>SourceFileNameR2<\/th>\n      <th>DNAnumber<\/th>\n      <th>ReplicateNumber<\/th>\n      <th>DNAextractionDate<\/th>\n      <th>DNAngUsed<\/th>\n      <th>LinearPCRID<\/th>\n      <th>LinearPCRDate<\/th>\n      <th>SonicationDate<\/th>\n      <th>LigationDate<\/th>\n      <th>1stExpoPCRID<\/th>\n      <th>1stExpoPCRDate<\/th>\n      <th>2ndExpoID<\/th>\n      <th>2ndExpoDate<\/th>\n      <th>FusionPrimerPCRID<\/th>\n      <th>FusionPrimerPCRDate<\/th>\n      <th>PoolDate<\/th>\n      <th>SequencingDate<\/th>\n      <th>VCN<\/th>\n      <th>Genome<\/th>\n      <th>SequencingRound<\/th>\n      <th>Genotype<\/th>\n      <th>TestGroup<\/th>\n      <th>MOI<\/th>\n      <th>Engraftment<\/th>\n      <th>Transduction<\/th>\n      <th>Notes<\/th>\n      <th>AddedField1<\/th>\n      <th>AddedField2<\/th>\n      <th>AddedField3<\/th>\n      <th>AddedField4<\/th>\n      <th>concatenatePoolIDSeqRun<\/th>\n      <th>AddedField6_RelativeBloodPercentage<\/th>\n      <th>AddedField7_PurityTestFeasibility<\/th>\n      <th>AddedField8_FacsSeparationPurity<\/th>\n      <th>Kapa<\/th>\n      <th>ulForPool<\/th>\n      <th>CompleteAmplificationID<\/th>\n      <th>UniqueID<\/th>\n      <th>StudyTestID<\/th>\n      <th>StudyTestGroup<\/th>\n      <th>MouseID<\/th>\n      <th>Tigroup<\/th>\n      <th>Tisource<\/th>\n      <th>PathToFolderProjectID<\/th>\n      <th>SamplesNameCheck<\/th>\n      <th>TimepointDays<\/th>\n      <th>TimepointMonths<\/th>\n      <th>TimepointYears<\/th>\n      <th>ng DNA corrected<\/th>\n      <th>Path<\/th>\n      <th>Path_quant<\/th>\n      <th>Path_iss<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","scrollX":120,"columnDefs":[{"className":"dt-right","targets":[24,26,39,41,45,46,54,55,56,57,61,62,67,70]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><div id="function-arguments" class="section level2">
<h2 class="hasAnchor">
<a href="#function-arguments" class="anchor"></a>Function arguments</h2>
<p>You can change several arguments in the function call to modify the behavior of the function.</p>
<ul>
<li>
<code>root</code>
<ul>
<li>Set it to <code>NULL</code> if you only want to import the association file without file system alignment. Beware that some of the automated import functionalities won’t work!</li>
<li>Set it to a non-empty string (path on disk): in this case, the column <code>PathToFolderProjectID</code> in the file should contain <strong>relative</strong> file paths, so if for example your root is set to “/home” and your project folder in the association file is set to “/PJ01”, the function will check that the directory exists under “/home/PJ01”</li>
<li>Set it to an empty string: ideal if you want to store paths in the association file as <strong>absolute</strong> file paths. In this case if your project folder is in “/home/PJ01” you should have this path in the <code>PathToFolderProjectID</code> column and set <code>root</code> = ""</li>
</ul>
</li>
<li>
<code>tp_padding</code>: this argument is used to pad the <code>TimePoint</code> column in the association file so that time points have all the same length</li>
<li>
<code>dates_format</code>: a string that is useful for properly parsing dates from tabular formats</li>
<li>
<code>separator</code>: the column separator used in the file. Defaults to “\t”, other valid separators are “,” (comma), “;” (semi-colon)</li>
<li>
<code>filter_for</code>: you can set this argument to a <strong>named</strong> list of filters, where names are column names. For example <code><a href="https://rdrr.io/r/base/list.html">list(ProjectID = "PJ01")</a></code> will return only those rows whose attribute “ProjectID” equals “PJ01”</li>
<li>
<code>import_iss</code>: either <code>TRUE</code> or <code>FALSE</code>. If set to <code>TRUE</code>, performs an internal call to <code><a href="../reference/import_Vispa2_stats.html">import_Vispa2_stats()</a></code> (see next section), and appends the imported files to metadata</li>
<li>
<code>convert_tp</code>: either <code>TRUE</code> or <code>FALSE</code>. Converts the “TimePoint” column in months and years (with custom logic).</li>
<li>
<code>report_path</code>
<ul>
<li>Set it to <code>NULL</code> to avoid the production of a report</li>
<li>Set it to a folder (if it doesn’t exist, it gets automatically created)</li>
<li>Set it to a file</li>
</ul>
</li>
<li>
<code>...</code>: additional named arguments to pass to <code><a href="../reference/import_Vispa2_stats.html">import_Vispa2_stats()</a></code> if you chose to import VISPA2 stats</li>
</ul>
<p><em>NOTE</em>: the function supports files in various formats as long as the correct separator is provided. It also accepts files in <code>*.xlsx</code> and <code>*.xls</code> formats but we do not recommend using these since the report won’t include a detailed summary of potential parsing problems.</p>
<p>The interactive report includes useful information such as</p>
<ul>
<li>General issues: parsing problems, missing columns, NA values in important columns etc. This allows you to immediately spot problems and correct them before proceeding with the analyses</li>
<li>File system alignment issues: very useful to know if all data can be imported or folders are missing</li>
<li>Info on VISPA2 stats (if <code>import_iss</code> was <code>TRUE</code>)</li>
</ul>
</div>
</div>
<div id="importing-vispa2-stats-files" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-vispa2-stats-files" class="anchor"></a>Importing VISPA2 stats files</h1>
<p>VISPA2 automatically produces summary files for each pool holding information that can be useful for other analyses downstream, so it is recommended to import them in the first steps of the workflow. To do that, you can use <code>import_VISPA2_stats</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_options.html">with_options</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>ISAnalytics.reports <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="op">{</span>
    <span class="va">vispa_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_Vispa2_stats.html">import_Vispa2_stats</a></span><span class="op">(</span>
        association_file <span class="op">=</span> <span class="va">af</span>,
        join_with_af <span class="op">=</span> <span class="cn">FALSE</span>
    <span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-01a9168ac27db5f49d11" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-01a9168ac27db5f49d11">{"x":{"filter":"none","data":[["PJ01|POOL01-1","PJ01|POOL01-1","PJ01|POOL01-1","PJ01|POOL01-1","PJ01|POOL01-1","PJ01|POOL01-1"],["POOL01-1","POOL01-1","POOL01-1","POOL01-1","POOL01-1","POOL01-1"],["LTR75LC38","LTR53LC32","LTR83LC66","LTR27LC94","LTR69LC52","LTR37LC2"],[43586699,43586699,43586699,43586699,43586699,43586699],[2256176,2256176,2256176,2256176,2256176,2256176],[645026,652208,451519,426500,18300,729327],[645026,652177,451512,426499,18300,729327],[630965,649044,449669,425666,18290,727219],[211757,303300,204810,185752,6962,318653],[402477,322086,227275,223915,10487,369117],[219452,222646,149385,143283,5907,235640],[null,null,null,null,null,null],[null,null,null,null,null,null],[null,null,null,null,null,null]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>RUN_NAME<\/th>\n      <th>POOL<\/th>\n      <th>TAG<\/th>\n      <th>PHIX_MAPPING<\/th>\n      <th>PLASMID_MAPPED_BYPOOL<\/th>\n      <th>BARCODE_MUX<\/th>\n      <th>LTR_IDENTIFIED<\/th>\n      <th>TRIMMING_FINAL_LTRLC<\/th>\n      <th>LV_MAPPED<\/th>\n      <th>BWA_MAPPED_OVERALL<\/th>\n      <th>ISS_MAPPED_OVERALL<\/th>\n      <th>RAW_READS<\/th>\n      <th>QUALITY_PASSED<\/th>\n      <th>ISS_MAPPED_PP<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","scrollX":120,"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8,9,10]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>The function requires as input the imported and file system aligned association file and it will scan the <code>iss</code> folder for files that match some known prefixes (defaults are already provided but you can change them as you see fit). You can either choose to join the imported data frames with the association file in input and obtain a single data frame or keep it as it is, just set the parameter <code>join_with_af</code> accordingly. At the end of the process an HTML report is produced, signaling potential problems.</p>
<p>You can directly call this function when you import the association file by setting the <code>import_iss</code> argument of <code>import_association_file</code> to <code>TRUE</code>.</p>
</div>
<div id="importing-a-single-integration-matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-a-single-integration-matrix" class="anchor"></a>Importing a single integration matrix</h1>
<p>If you want to import a single integration matrix you can do so by using the <code><a href="../reference/import_single_Vispa2Matrix.html">import_single_Vispa2Matrix()</a></code> function. This function reads the file and converts it into a tidy structure: several different formats can be read, since you can specify the column separator.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">matrix_path</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="http://fs.r-lib.org/reference/path.html">path</a></span><span class="op">(</span><span class="va">root</span>,
                        <span class="st">"PJ01"</span>,
                        <span class="st">"quantification"</span>,
                        <span class="st">"POOL01-1"</span>,
                        <span class="st">"PJ01_POOL01-1_seqCount_matrix.no0.annotated.tsv.gz"</span><span class="op">)</span>
<span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_single_Vispa2Matrix.html">import_single_Vispa2Matrix</a></span><span class="op">(</span><span class="va">matrix_path</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-00552e1d2aa5b213eb61" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-00552e1d2aa5b213eb61">{"x":{"filter":"none","data":[["16","16","16","16","16","16"],[68164148,1762026,15966129,1762026,29843197,1650839],["+","-","-","-","-","+"],["NFATC3","MAPK8IP3","FOPNL","MAPK8IP3","MVP","LOC105371046"],["+","+","-","+","+","+"],["PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC","PJ01_POOL01_LTR53LC32_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC","PJ01_POOL01_LTR53LC32_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC","PJ01_POOL01_LTR83LC66_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC","PJ01_POOL01_LTR83LC66_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC","PJ01_POOL01_LTR83LC66_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC"],[182,727,821,37,983,389]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>chr<\/th>\n      <th>integration_locus<\/th>\n      <th>strand<\/th>\n      <th>GeneName<\/th>\n      <th>GeneStrand<\/th>\n      <th>CompleteAmplificationID<\/th>\n      <th>Value<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","scrollX":120,"columnDefs":[{"className":"dt-right","targets":[1,6]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>Other arguments you can pass to the function are</p>
<ul>
<li>
<code>to_exclude</code>: a character vector that contains column names that need to be excluded when imported. This is more targeted towards files that do have all the columns of an integration matrix as presented in section @ref(notation) and other additional columns. By default this argument is set to <code>NULL</code>
</li>
<li>
<code>keep_excluded</code>: if set to <code>TRUE</code> all columns contained in <code>to_exclude</code> are preserved as additional annotation columns</li>
</ul>
</div>
<div id="automated-integration-matrices-import" class="section level1">
<h1 class="hasAnchor">
<a href="#automated-integration-matrices-import" class="anchor"></a>Automated integration matrices import</h1>
<p>Integration matrices import can be automated when when the association file is imported with the file system alignment option. <code>ISAnalytics</code> provides a function, <code><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices()</a></code>, that allows to do just that in a fast and efficient way.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_options.html">with_options</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>ISAnalytics.reports <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="op">{</span>
    <span class="va">matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices</a></span><span class="op">(</span><span class="va">af</span>,
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span>,
        mode <span class="op">=</span> <span class="st">"AUTO"</span>
    <span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<div id="function-arguments-1" class="section level2">
<h2 class="hasAnchor">
<a href="#function-arguments-1" class="anchor"></a>Function arguments</h2>
<p>Let’s see how the behavior of the function changes when we change arguments.</p>
<div id="association_file-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#association_file-argument" class="anchor"></a><code>association_file</code> argument</h3>
<p>You can supply a data frame object, imported via <code><a href="../reference/import_association_file.html">import_association_file()</a></code> (see Section @ref(metadata)) or a string (the path to the association file on disk). In the first scenario it is necessary to perform file system alignment, since the function scans the folders contained in the column <code>Path_quant</code>, while in the second case you should also provide as additional <strong>named</strong> argument (to <code>...</code>) an appropriate <code>root</code>: the function will internally call <code><a href="../reference/import_association_file.html">import_association_file()</a></code>, if you don’t have specific needs we recommend doing the 2 steps separately and provide the association file as a data frame.</p>
</div>
<div id="quantification_type-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#quantification_type-argument" class="anchor"></a><code>quantification_type</code> argument</h3>
<p>For each pool there may be multiple available quantification types, that is, different matrices containing the same samples and same genomic features but a different quantification. A typical workflow contemplates <code>seqCount</code> and <code>fragmentEstimate</code>, all the supported quantification types can be viewed with <code><a href="../reference/quantification_types.html">quantification_types()</a></code>.</p>
</div>
<div id="matrix_type-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#matrix_type-argument" class="anchor"></a><code>matrix_type</code> argument</h3>
<p>As we mentioned in Section @ref(notation), annotation columns are optional and may not be included in some matrices. This argument allows you to specify the function to look for only a specific type of matrix, either <code>annotated</code> or <code>not_annotated</code>. Please note that in order to do that, for now, the function needs to assume some standard file name notation, that is, for <code>annotated</code> matrices, the function will look for the <code>.no0.annotated</code> suffix in the file name.</p>
</div>
<div id="workers-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#workers-argument" class="anchor"></a><code>workers</code> argument</h3>
<p>Sets the number of parallel workers to set up. This highly depends on the hardware configuration of your machine.</p>
</div>
<div id="multi_quant_matrix-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#multi_quant_matrix-argument" class="anchor"></a><code>multi_quant_matrix</code> argument</h3>
<p>When importing more than one quantification at once, it can be very handy to have all data in a single data frame rather than two. If set to <code>TRUE</code> the function will internally call <code><a href="../reference/comparison_matrix.html">comparison_matrix()</a></code> and produce a single data frames that has a dedicated column for each quantification. For example, for the matrices we’ve imported before:</p>
<div id="htmlwidget-8521af7de90232bdf642" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8521af7de90232bdf642">{"x":{"filter":"none","data":[["16","4","5","12","2","20"],[68164148,129390130,84009671,54635693,181930711,35920986],["+","+","-","-","+","+"],["NFATC3","LOC100507487","EDIL3","CBX5","UBE2E3","MANBAL"],["+","+","-","-","+","+"],["PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC","PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC","PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC","PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC","PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC","PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC"],[102.945718327,3.0135895554,5.03374159776,9.1293375748,50.5027340914,16.3944223177],[182,4,5,9,83,39]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>chr<\/th>\n      <th>integration_locus<\/th>\n      <th>strand<\/th>\n      <th>GeneName<\/th>\n      <th>GeneStrand<\/th>\n      <th>CompleteAmplificationID<\/th>\n      <th>fragmentEstimate<\/th>\n      <th>seqCount<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","scrollX":120,"columnDefs":[{"className":"dt-right","targets":[1,6,7]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="report_path-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#report_path-argument" class="anchor"></a><code>report_path</code> argument</h3>
<p>As other import functions, also <code><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices()</a></code> produces an interactive report, use this argument to set the appropriate path were the report should be saved.</p>
</div>
<div id="mode-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#mode-argument" class="anchor"></a><code>mode</code> argument</h3>
<p>This argument can take one of two values, <code>AUTO</code> or <code>INTERACTIVE</code>. The <code>INTERACTIVE</code> workflow, as the name suggests, needs user console input but allows a fine tuning of the import process. On the other hand, <code>AUTO</code> allows a fully automated workflow but has of course some limitations.</p>
<p><strong>What do you want to import?</strong><br>
In a fully automated mode, the function will try to import everything that is contained in the input association file. This means that if you need to import only a specific set of projects/pools, you will need to filter the association file accordingly prior calling the function (you can easily do that via the <code>filter_for</code> argument as explained in Section @ref(metadata)). In interactive mode the function will ask you to type what you want to import.</p>
<p><strong>How to deal with duplicates?</strong><br>
When scanning folders for files that match a given pattern (in our case the function looks for matrices that match the quantification type and the matrix type), it is very possible that the same folder contains multiple files for the same quantification. Of course this is not recommended, we suggest to move the duplicated files in a sub directory or remove them if they’re not necessary, but in case this happens, in interactive mode, the function asks directly which files should be considered. Of course this is not possible in automated mode, therefore you need to set two other arguments (described in the next sub sections) to “help” the function discriminate between duplicates. Please note that if such discrimination is not possible no files are imported.</p>
</div>
<div id="patterns-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#patterns-argument" class="anchor"></a><code>patterns</code> argument</h3>
<p>This argument is relevant only if <code>mode</code> is set to <code>AUTO</code>. Providing a set of patterns (interpreted as regular expressions) helps the function to choose between duplicated files if any are found. If you’re confident your folders don’t contain any duplicates feel free to ignore this argument.</p>
</div>
<div id="matching_opt-argument" class="section level3">
<h3 class="hasAnchor">
<a href="#matching_opt-argument" class="anchor"></a><code>matching_opt</code> argument</h3>
<p>This argument is relevant only if <code>mode</code> is set to <code>AUTO</code> and <code>patterns</code> isn’t <code>NULL</code>. Tells the function how to match the given patterns if multiple are supplied: <code>ALL</code> means keep only those files whose name matches all the given patterns, <code>ANY</code> means keep only those files whose name matches any of the given patterns and <code>OPTIONAL</code> expresses a preference, try to find files that contain the patterns and if you don’t find any return whatever you find.</p>
</div>
<div id="argument" class="section level3">
<h3 class="hasAnchor">
<a href="#argument" class="anchor"></a><code>...</code> argument</h3>
<p>Additional named arguments to supply to both <code><a href="../reference/import_association_file.html">import_association_file()</a></code> and <code><a href="../reference/comparison_matrix.html">comparison_matrix()</a></code>.</p>
</div>
</div>
<div id="notes" class="section level2">
<h2 class="hasAnchor">
<a href="#notes" class="anchor"></a>Notes</h2>
<p>Earlier versions of the package featured two separated functions, <code><a href="../reference/import_parallel_Vispa2Matrices_auto.html">import_parallel_Vispa2Matrices_auto()</a></code> and <code><a href="../reference/import_parallel_Vispa2Matrices_interactive.html">import_parallel_Vispa2Matrices_interactive()</a></code>. Those functions are now officially deprecated (since <code>ISAnalytics 1.3.3</code>) and will be defunct on the next release cycle.</p>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p><code>R</code> session information.</p>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  setting  value                       
#&gt;  version  R version 4.1.0 (2021-05-18)
#&gt;  os       macOS Catalina 10.15.7      
#&gt;  system   x86_64, darwin17.0          
#&gt;  ui       X11                         
#&gt;  language (EN)                        
#&gt;  collate  en_US.UTF-8                 
#&gt;  ctype    en_US.UTF-8                 
#&gt;  tz       UTC                         
#&gt;  date     2021-08-03                  
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  package      * version date       lib source        
#&gt;  BiocManager    1.30.16 2021-06-15 [1] CRAN (R 4.1.0)
#&gt;  BiocParallel   1.26.1  2021-07-04 [1] Bioconductor  
#&gt;  BiocStyle    * 2.20.2  2021-06-17 [1] Bioconductor  
#&gt;  bit            4.0.4   2020-08-04 [1] CRAN (R 4.1.0)
#&gt;  bit64          4.0.5   2020-08-30 [1] CRAN (R 4.1.0)
#&gt;  bookdown       0.22    2021-04-22 [1] CRAN (R 4.1.0)
#&gt;  bslib          0.2.5.1 2021-05-18 [1] CRAN (R 4.1.0)
#&gt;  cachem         1.0.5   2021-05-15 [1] CRAN (R 4.1.0)
#&gt;  cli            3.0.1   2021-07-17 [1] CRAN (R 4.1.0)
#&gt;  colorspace     2.0-2   2021-06-24 [1] CRAN (R 4.1.0)
#&gt;  crayon         1.4.1   2021-02-08 [1] CRAN (R 4.1.0)
#&gt;  crosstalk      1.1.1   2021-01-12 [1] CRAN (R 4.1.0)
#&gt;  data.table     1.14.0  2021-02-21 [1] CRAN (R 4.1.0)
#&gt;  desc           1.3.0   2021-03-05 [1] CRAN (R 4.1.0)
#&gt;  digest         0.6.27  2020-10-24 [1] CRAN (R 4.1.0)
#&gt;  dplyr          1.0.7   2021-06-18 [1] CRAN (R 4.1.0)
#&gt;  DT             0.18    2021-04-14 [1] CRAN (R 4.1.0)
#&gt;  ellipsis       0.3.2   2021-04-29 [1] CRAN (R 4.1.0)
#&gt;  evaluate       0.14    2019-05-28 [1] CRAN (R 4.1.0)
#&gt;  fansi          0.5.0   2021-05-25 [1] CRAN (R 4.1.0)
#&gt;  fastmap        1.1.0   2021-01-25 [1] CRAN (R 4.1.0)
#&gt;  fs             1.5.0   2020-07-31 [1] CRAN (R 4.1.0)
#&gt;  generics       0.1.0   2020-10-31 [1] CRAN (R 4.1.0)
#&gt;  ggplot2        3.3.5   2021-06-25 [1] CRAN (R 4.1.0)
#&gt;  ggrepel        0.9.1   2021-01-15 [1] CRAN (R 4.1.0)
#&gt;  glue           1.4.2   2020-08-27 [1] CRAN (R 4.1.0)
#&gt;  gtable         0.3.0   2019-03-25 [1] CRAN (R 4.1.0)
#&gt;  hms            1.1.0   2021-05-17 [1] CRAN (R 4.1.0)
#&gt;  htmltools      0.5.1.1 2021-01-22 [1] CRAN (R 4.1.0)
#&gt;  htmlwidgets    1.5.3   2020-12-10 [1] CRAN (R 4.1.0)
#&gt;  httr           1.4.2   2020-07-20 [1] CRAN (R 4.1.0)
#&gt;  ISAnalytics  * 1.3.4   2021-08-03 [1] Bioconductor  
#&gt;  jquerylib      0.1.4   2021-04-26 [1] CRAN (R 4.1.0)
#&gt;  jsonlite       1.7.2   2020-12-09 [1] CRAN (R 4.1.0)
#&gt;  knitr          1.33    2021-04-24 [1] CRAN (R 4.1.0)
#&gt;  lifecycle      1.0.0   2021-02-15 [1] CRAN (R 4.1.0)
#&gt;  lubridate      1.7.10  2021-02-26 [1] CRAN (R 4.1.0)
#&gt;  magrittr     * 2.0.1   2020-11-17 [1] CRAN (R 4.1.0)
#&gt;  memoise        2.0.0   2021-01-26 [1] CRAN (R 4.1.0)
#&gt;  munsell        0.5.0   2018-06-12 [1] CRAN (R 4.1.0)
#&gt;  pillar         1.6.2   2021-07-29 [1] CRAN (R 4.1.0)
#&gt;  pkgconfig      2.0.3   2019-09-22 [1] CRAN (R 4.1.0)
#&gt;  pkgdown        1.6.1   2020-09-12 [1] CRAN (R 4.1.0)
#&gt;  plyr           1.8.6   2020-03-03 [1] CRAN (R 4.1.0)
#&gt;  purrr          0.3.4   2020-04-17 [1] CRAN (R 4.1.0)
#&gt;  R.methodsS3    1.8.1   2020-08-26 [1] CRAN (R 4.1.0)
#&gt;  R.oo           1.24.0  2020-08-26 [1] CRAN (R 4.1.0)
#&gt;  R.utils        2.10.1  2020-08-26 [1] CRAN (R 4.1.0)
#&gt;  R6             2.5.0   2020-10-28 [1] CRAN (R 4.1.0)
#&gt;  ragg           1.1.3   2021-06-09 [1] CRAN (R 4.1.0)
#&gt;  Rcapture       1.4-3   2019-12-16 [1] CRAN (R 4.1.0)
#&gt;  Rcpp           1.0.7   2021-07-07 [1] CRAN (R 4.1.0)
#&gt;  readr          2.0.0   2021-07-20 [1] CRAN (R 4.1.0)
#&gt;  RefManageR   * 1.3.0   2020-11-13 [1] CRAN (R 4.1.0)
#&gt;  rlang          0.4.11  2021-04-30 [1] CRAN (R 4.1.0)
#&gt;  rmarkdown      2.9     2021-06-15 [1] CRAN (R 4.1.0)
#&gt;  rprojroot      2.0.2   2020-11-15 [1] CRAN (R 4.1.0)
#&gt;  sass           0.4.0   2021-05-12 [1] CRAN (R 4.1.0)
#&gt;  scales         1.1.1   2020-05-11 [1] CRAN (R 4.1.0)
#&gt;  sessioninfo  * 1.1.1   2018-11-05 [1] CRAN (R 4.1.0)
#&gt;  stringi        1.7.3   2021-07-16 [1] CRAN (R 4.1.0)
#&gt;  stringr        1.4.0   2019-02-10 [1] CRAN (R 4.1.0)
#&gt;  systemfonts    1.0.2   2021-05-11 [1] CRAN (R 4.1.0)
#&gt;  textshaping    0.3.5   2021-06-09 [1] CRAN (R 4.1.0)
#&gt;  tibble         3.1.3   2021-07-23 [1] CRAN (R 4.1.0)
#&gt;  tidyr          1.1.3   2021-03-03 [1] CRAN (R 4.1.0)
#&gt;  tidyselect     1.1.1   2021-04-30 [1] CRAN (R 4.1.0)
#&gt;  tzdb           0.1.2   2021-07-20 [1] CRAN (R 4.1.0)
#&gt;  utf8           1.2.2   2021-07-24 [1] CRAN (R 4.1.0)
#&gt;  vctrs          0.3.8   2021-04-29 [1] CRAN (R 4.1.0)
#&gt;  vroom          1.5.3   2021-07-14 [1] CRAN (R 4.1.0)
#&gt;  withr          2.4.2   2021-04-18 [1] CRAN (R 4.1.0)
#&gt;  xfun           0.24    2021-06-15 [1] CRAN (R 4.1.0)
#&gt;  xml2           1.3.2   2020-04-23 [1] CRAN (R 4.1.0)
#&gt;  yaml           2.2.1   2020-02-01 [1] CRAN (R 4.1.0)
#&gt;  zip            2.2.0   2021-05-31 [1] CRAN (R 4.1.0)
#&gt; 
#&gt; [1] /Users/runner/work/_temp/Library
#&gt; [2] /Library/Frameworks/R.framework/Versions/4.1/Resources/library</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.13/BiocStyle">BiocStyle</a></em> (Oleś, 2021) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (Xie, 2021) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> (Allaire, Xie, McPherson, Luraschi, Ushey, Atkins, Wickham, Cheng, Chang, and Iannone, 2021) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR">RefManageR</a></em> (McLean, 2017).</p>
<p>
<a id="bib-allaire2021rmarkdown"></a><a href="#cite-allaire2021rmarkdown">[1]</a><cite> J. Allaire, Y. Xie, J. McPherson, et al. <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.9. 2021. URL: <a href="https://github.com/rstudio/rmarkdown">https://github.com/rstudio/rmarkdown</a>.</cite>
</p>
<p>
<a id="bib-giuliospinozzivispa2"></a><a href="#cite-giuliospinozzivispa2">[2]</a><cite> S. B. Giulio Spinozzi Andrea Calabria. “VISPA2: a scalable pipeline for high-throughput identification and annotation of vector integration sites”. In: <em>BMC Bioinformatics</em> (Nov. 25, 2017). DOI: <a href="https://doi.org/10.1186/s12859-017-1937-9">10.1186/s12859-017-1937-9</a>.</cite>
</p>
<p>
<a id="bib-mclean2017refmanager"></a><a href="#cite-mclean2017refmanager">[3]</a><cite> M. W. McLean. “RefManageR: Import and Manage BibTeX and BibLaTeX References in R”. In: <em>The Journal of Open Source Software</em> (2017). DOI: <a href="https://doi.org/10.21105/joss.00338">10.21105/joss.00338</a>.</cite>
</p>
<p>
<a id="bib-ole2021biocstyle"></a><a href="#cite-ole2021biocstyle">[4]</a><cite> A. Oleś. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.20.2. 2021. URL: <a href="https://github.com/Bioconductor/BiocStyle">https://github.com/Bioconductor/BiocStyle</a>.</cite>
</p>
<p>
<a id="bib-xie2021knitr"></a><a href="#cite-xie2021knitr">[5]</a><cite> Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.33. 2021. URL: <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.</cite>
</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
