<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="
Statistical approach for the validation of common insertion sites
significance based on the comparison of the integration frequency
at the CIS gene with respect to other genes contained in the
surrounding genomic regions. For more details please refer to
this paper:
https://ashpublications.org/blood/article/117/20/5332/21206/Lentiviral-vector-common-integration-sites-in"><title>Grubbs test for Common Insertion Sites (CIS). — CIS_grubbs • ISAnalytics</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Grubbs test for Common Insertion Sites (CIS). — CIS_grubbs"><meta property="og:description" content="
Statistical approach for the validation of common insertion sites
significance based on the comparison of the integration frequency
at the CIS gene with respect to other genes contained in the
surrounding genomic regions. For more details please refer to
this paper:
https://ashpublications.org/blood/article/117/20/5332/21206/Lentiviral-vector-common-integration-sites-in"><meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ISAnalytics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/ISAnalytics.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/workflow_start.html">Setting up the workflow and first steps</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="nav-link" href="https://calabrialab.github.io/ISAnalytics/dev/">
    <span class="fas fa fas fa-code-branch"></span>
     
    DEVEL
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Grubbs test for Common Insertion Sites (CIS).</h1>
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/HEAD/R/analysis-functions.R" class="external-link"><code>R/analysis-functions.R</code></a></small>
      <div class="d-none name"><code>CIS_grubbs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="figures/lifecycle-stable.svg" alt="[Stable]"></a>
Statistical approach for the validation of common insertion sites
significance based on the comparison of the integration frequency
at the CIS gene with respect to other genes contained in the
surrounding genomic regions. For more details please refer to
this paper:
<a href="https://ashpublications.org/blood/article/117/20/5332/21206/Lentiviral-vector-common-integration-sites-in" class="external-link">https://ashpublications.org/blood/article/117/20/5332/21206/Lentiviral-vector-common-integration-sites-in</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CIS_grubbs</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  genomic_annotation_file <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  grubbs_flanking_gene_bp <span class="op">=</span> <span class="fl">1e+05</span>,</span>
<span>  threshold_alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_missing_as_df <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  results_as_list <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>An integration matrix, must include the <code><a href="mandatory_IS_vars.html">mandatory_IS_vars()</a></code>
columns and the <code><a href="mandatory_IS_vars.html">annotation_IS_vars()</a></code> columns</p></dd>


<dt>genomic_annotation_file</dt>
<dd><p>Database file for gene annotation,
see details.</p></dd>


<dt>grubbs_flanking_gene_bp</dt>
<dd><p>Number of base pairs flanking a gene</p></dd>


<dt>threshold_alpha</dt>
<dd><p>Significance threshold</p></dd>


<dt>by</dt>
<dd><p>Either <code>NULL</code> or a character vector of column names. If not
NULL, the function will perform calculations for each group and return
a list of data frames with the results. E.g. for <code>by = "SubjectID"</code>,
CIS will be computed for each distinct SubjectID found in the table
("SubjectID" column must be included in the input data frame).</p></dd>


<dt>return_missing_as_df</dt>
<dd><p>Returns those genes present in the input df
but not in the refgenes as a data frame?</p></dd>


<dt>results_as_list</dt>
<dd><p>If <code>TRUE</code>
return the group computations as a named list, otherwise return a single
df with an additional column containing the group id</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data frame</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    
<div class="section">
<h3 id="genomic-annotation-file">Genomic annotation file<a class="anchor" aria-label="anchor" href="#genomic-annotation-file"></a></h3>


<p>A data frame containing
genes annotation for the specific genome.
From version <code>1.5.4</code> the argument <code>genomic_annotation_file</code> accepts only
data frames or package provided defaults.
The user is responsible for importing the appropriate tabular files if
customization is needed.
The annotations for the human genome (hg19) and
murine genome (mm9) are already
included in this package: to use one of them just
set the argument <code>genomic_annotation_file</code> to either <code>"hg19"</code> or
<code>"mm9"</code>.
If for any reason the user is performing an analysis on another genome,
this file needs to be changed respecting the USCS Genome Browser
format, meaning the input file headers should include:</p>
<p>name2, chrom, strand, min_txStart, max_txEnd, minmax_TxLen, average_TxLen, name, min_cdsStart, max_cdsEnd, minmax_CdsLen, average_CdsLen</p>
</div>

    </div>
    <div class="section level2">
    <h2 id="required-tags">Required tags<a class="anchor" aria-label="anchor" href="#required-tags"></a></h2>
    

<p>The function will explicitly check for the presence of these tags:</p><ul><li><p>chromosome</p></li>
<li><p>locus</p></li>
<li><p>is_strand</p></li>
<li><p>gene_symbol</p></li>
<li><p>gene_strand</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other Analysis functions: 
<code><a href="HSC_population_size_estimate.html">HSC_population_size_estimate</a>()</code>,
<code><a href="compute_abundance.html">compute_abundance</a>()</code>,
<code><a href="cumulative_is.html">cumulative_is</a>()</code>,
<code><a href="gene_frequency_fisher.html">gene_frequency_fisher</a>()</code>,
<code><a href="is_sharing.html">is_sharing</a>()</code>,
<code><a href="iss_source.html">iss_source</a>()</code>,
<code><a href="sample_statistics.html">sample_statistics</a>()</code>,
<code><a href="top_integrations.html">top_integrations</a>()</code>,
<code><a href="top_targeted_genes.html">top_targeted_genes</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"integration_matrices"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cis</span> <span class="op">&lt;-</span> <span class="fu">CIS_grubbs</span><span class="op">(</span><span class="va">integration_matrices</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Warning: missing genes in refgenes table</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ</span> A total of 5 genes were found in the input data but not in the refgene table. This may be caused by a mismatch in the annotation phase of the matrix. Here is a summary: </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>     GeneName GeneStrand chr</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> 1:    CRELD2          -  15</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> 2:      HTR4          +   6</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> 3:    UBE2D2          +  16</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> 4: LINC01133          +  19</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> 5:  PLEKHG4B          -  14</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ</span> NOTE: missing genes will be removed from the final output! Review results carefully</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">ℹ</span> A total of 25 IS will be removed because of missing genes ( 1.48 % of total IS in input)</span>
<span class="r-in"><span><span class="va">cis</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $missing_genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     GeneName GeneStrand chr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:    CRELD2          -  15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:      HTR4          +   6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:    UBE2D2          +  16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4: LINC01133          +  19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:  PLEKHG4B          -  14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $missing_is</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $missing_is$absolute</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 25</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $missing_is$perc</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1.48</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $cis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 501 × 37</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    GeneName GeneStrand chr       n       mean    sd  median trimmed   mad    min</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> ABHD16A  -          6         3  31<span style="text-decoration: underline;">654</span>876     0   3.17<span style="color: #949494;">e</span>7  3.17<span style="color: #949494;">e</span>7    0  3.17<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> ABHD2    +          15        3  89<span style="text-decoration: underline;">642</span>998     0   8.96<span style="color: #949494;">e</span>7  8.96<span style="color: #949494;">e</span>7    0  8.96<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> ACAP2    -          3         2 195<span style="text-decoration: underline;">107</span>772     0   1.95<span style="color: #949494;">e</span>8  1.95<span style="color: #949494;">e</span>8    0  1.95<span style="color: #949494;">e</span>8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> ACOX1    -          17       10  73<span style="text-decoration: underline;">960</span>093. <span style="text-decoration: underline;">8</span>844.  7.40<span style="color: #949494;">e</span>7  7.40<span style="color: #949494;">e</span>7 <span style="text-decoration: underline;">4</span>236. 7.40<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> ACSM3    +          16        4  20<span style="text-decoration: underline;">777</span>188     0   2.08<span style="color: #949494;">e</span>7  2.08<span style="color: #949494;">e</span>7    0  2.08<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> ADD1     +          4         8   2<span style="text-decoration: underline;">865</span>706. <span style="text-decoration: underline;">6</span>694.  2.87<span style="color: #949494;">e</span>6  2.87<span style="color: #949494;">e</span>6 <span style="text-decoration: underline;">8</span>804. 2.86<span style="color: #949494;">e</span>6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> ADGRA3   -          4         2  22<span style="text-decoration: underline;">505</span>019     0   2.25<span style="color: #949494;">e</span>7  2.25<span style="color: #949494;">e</span>7    0  2.25<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> ADGRB3   +          6         4  69<span style="text-decoration: underline;">397</span>949     0   6.94<span style="color: #949494;">e</span>7  6.94<span style="color: #949494;">e</span>7    0  6.94<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> ADGRG3   +          16        3  57<span style="text-decoration: underline;">716</span>345     0   5.77<span style="color: #949494;">e</span>7  5.77<span style="color: #949494;">e</span>7    0  5.77<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> AHR      +          7         2  17<span style="text-decoration: underline;">227</span>687     0   1.72<span style="color: #949494;">e</span>7  1.72<span style="color: #949494;">e</span>7    0  1.72<span style="color: #949494;">e</span>7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 491 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 27 more variables: max &lt;dbl&gt;, range &lt;dbl&gt;, skew &lt;dbl&gt;, kurtosis &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_IS_perGene &lt;int&gt;, min_bp_integration_locus &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   max_bp_integration_locus &lt;dbl&gt;, IS_span_bp &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   avg_bp_integration_locus &lt;dbl&gt;, median_bp_integration_locus &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   distinct_orientations &lt;int&gt;, average_TxLen &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   raw_gene_integration_frequency &lt;dbl&gt;, …</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Giulia Pais, Andrea Calabria, Giulio Spinozzi.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

