<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Fetches the look-up tables for different categories of dynamic
vars.
For more details, refer to the dedicated vignette
vignette("workflow_start", package="ISAnalytics").
mandatory_IS_vars returns the look-up table of variables that are
used to uniquely identify integration events



annotation_IS_vars() returns the look-up table of variables that
contain genomic annotations



association_file_columns() returns the look-up table of variables that
contains information on how  metadata is structured



iss_stats_specs() returns the look-up table of variables that
contains information on the format of pool statistics files produced
automatically by VISPA2



matrix_file_suffixes() returns the look-up table of variables that
contains all default file names for each quantification type and it is
used by automated import functions

'><title>Current dynamic vars specifications getters. — mandatory_IS_vars • ISAnalytics</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Current dynamic vars specifications getters. — mandatory_IS_vars"><meta property="og:description" content='Fetches the look-up tables for different categories of dynamic
vars.
For more details, refer to the dedicated vignette
vignette("workflow_start", package="ISAnalytics").
mandatory_IS_vars returns the look-up table of variables that are
used to uniquely identify integration events



annotation_IS_vars() returns the look-up table of variables that
contain genomic annotations



association_file_columns() returns the look-up table of variables that
contains information on how  metadata is structured



iss_stats_specs() returns the look-up table of variables that
contains information on the format of pool statistics files produced
automatically by VISPA2



matrix_file_suffixes() returns the look-up table of variables that
contains all default file names for each quantification type and it is
used by automated import functions

'><meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ISAnalytics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.10.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/ISAnalytics.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/workflow_start.html">Setting up the workflow and first steps</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="nav-link" href="https://calabrialab.github.io/ISAnalytics/dev/">
    <span class="fas fa fas fa-code-branch"></span>
     
    DEVEL
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Current dynamic vars specifications getters.</h1>
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/HEAD/R/utility-functions.R" class="external-link"><code>R/utility-functions.R</code></a></small>
      <div class="d-none name"><code>mandatory_IS_vars.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Fetches the look-up tables for different categories of dynamic
vars.
For more details, refer to the dedicated vignette
<code><a href="../articles/workflow_start.html">vignette("workflow_start", package="ISAnalytics")</a></code>.</p><ul><li><p><code>mandatory_IS_vars</code> returns the look-up table of variables that are
used to uniquely identify integration events</p></li>
</ul><ul><li><p><code>annotation_IS_vars()</code> returns the look-up table of variables that
contain genomic annotations</p></li>
</ul><ul><li><p><code>association_file_columns()</code> returns the look-up table of variables that
contains information on how  metadata is structured</p></li>
</ul><ul><li><p><code>iss_stats_specs()</code> returns the look-up table of variables that
contains information on the format of pool statistics files produced
automatically by VISPA2</p></li>
</ul><ul><li><p><code>matrix_file_suffixes()</code> returns the look-up table of variables that
contains all default file names for each quantification type and it is
used by automated import functions</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mandatory_IS_vars</span><span class="op">(</span>include_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">annotation_IS_vars</span><span class="op">(</span>include_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">association_file_columns</span><span class="op">(</span>include_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">iss_stats_specs</span><span class="op">(</span>include_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">matrix_file_suffixes</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>include_types</dt>
<dd><p>If set to <code>TRUE</code> returns both the names and the types
associated, otherwise returns only a character vector of names</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A character vector or a data frame</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other dynamic vars: 
<code><a href="inspect_tags.html">inspect_tags</a>()</code>,
<code><a href="pcr_id_column.html">pcr_id_column</a>()</code>,
<code><a href="reset_mandatory_IS_vars.html">reset_mandatory_IS_vars</a>()</code>,
<code><a href="set_mandatory_IS_vars.html">set_mandatory_IS_vars</a>()</code>,
<code><a href="set_matrix_file_suffixes.html">set_matrix_file_suffixes</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Names only</span></span></span>
<span class="r-in"><span><span class="fu">mandatory_IS_vars</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr"               "integration_locus" "strand"           </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names and types</span></span></span>
<span class="r-in"><span><span class="fu">mandatory_IS_vars</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   names             types transform flag     tag       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> chr               char  <span style="color: #949494;">&lt;NULL&gt;</span>    required chromosome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> integration_locus int   <span style="color: #949494;">&lt;NULL&gt;</span>    required locus     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> strand            char  <span style="color: #949494;">&lt;NULL&gt;</span>    required is_strand </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names only</span></span></span>
<span class="r-in"><span><span class="fu">annotation_IS_vars</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "GeneName"   "GeneStrand"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names and types</span></span></span>
<span class="r-in"><span><span class="fu">annotation_IS_vars</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 2 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   names      types transform flag     tag        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> GeneName   char  <span style="color: #949494;">&lt;NULL&gt;</span>    required gene_symbol</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> GeneStrand char  <span style="color: #949494;">&lt;NULL&gt;</span>    required gene_strand</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names only</span></span></span>
<span class="r-in"><span><span class="fu">association_file_columns</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "ProjectID"                           "FUSIONID"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "PoolID"                              "TagSequence"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "SubjectID"                           "VectorType"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "VectorID"                            "ExperimentID"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "Tissue"                              "TimePoint"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "DNAFragmentation"                    "PCRMethod"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "TagIDextended"                       "Keywords"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15] "CellMarker"                          "TagID"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17] "NGSProvider"                         "NGSTechnology"                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] "ConverrtedFilesDir"                  "ConverrtedFilesName"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [21] "SourceFileFolder"                    "SourceFileNameR1"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [23] "SourceFileNameR2"                    "DNAnumber"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25] "ReplicateNumber"                     "DNAextractionDate"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [27] "DNAngUsed"                           "LinearPCRID"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [29] "LinearPCRDate"                       "SonicationDate"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31] "LigationDate"                        "1stExpoPCRID"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [33] "1stExpoPCRDate"                      "2ndExpoID"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [35] "2ndExpoDate"                         "FusionPrimerPCRID"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [37] "FusionPrimerPCRDate"                 "PoolDate"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [39] "SequencingDate"                      "VCN"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [41] "Genome"                              "SequencingRound"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [43] "Genotype"                            "TestGroup"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [45] "MOI"                                 "Engraftment"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [47] "Transduction"                        "Notes"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [49] "AddedField1"                         "AddedField2"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [51] "AddedField3"                         "AddedField4"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [53] "concatenatePoolIDSeqRun"             "AddedField6_RelativeBloodPercentage"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [55] "AddedField7_PurityTestFeasibility"   "AddedField8_FacsSeparationPurity"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [57] "Kapa"                                "ulForPool"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [59] "CompleteAmplificationID"             "UniqueID"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [61] "StudyTestID"                         "StudyTestGroup"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [63] "MouseID"                             "Tigroup"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [65] "Tisource"                            "PathToFolderProjectID"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [67] "SamplesNameCheck"                    "TimepointDays"                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [69] "TimepointMonths"                     "TimepointYears"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [71] "ng DNA corrected"                   </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names and types</span></span></span>
<span class="r-in"><span><span class="fu">association_file_columns</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 71 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    names        types transform flag     tag       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> ProjectID    char  <span style="color: #949494;">&lt;NULL&gt;</span>    required project_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> FUSIONID     char  <span style="color: #949494;">&lt;NULL&gt;</span>    optional fusion_id </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> PoolID       char  <span style="color: #949494;">&lt;NULL&gt;</span>    required pool_id   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> TagSequence  char  <span style="color: #949494;">&lt;NULL&gt;</span>    required tag_seq   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> SubjectID    char  <span style="color: #949494;">&lt;NULL&gt;</span>    required subject   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> VectorType   char  <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> VectorID     char  <span style="color: #949494;">&lt;NULL&gt;</span>    required vector_id </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> ExperimentID char  <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Tissue       char  <span style="color: #949494;">&lt;NULL&gt;</span>    required tissue    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> TimePoint    char  <span style="color: #949494;">&lt;formula&gt;</span> required tp_days   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 61 more rows</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names only</span></span></span>
<span class="r-in"><span><span class="fu">iss_stats_specs</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "RUN_NAME"              "POOL"                  "TAG"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "RAW_READS"             "QUALITY_PASSED"        "PHIX_MAPPING"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "PLASMID_MAPPED_BYPOOL" "BARCODE_MUX"           "LTR_IDENTIFIED"       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "TRIMMING_FINAL_LTRLC"  "LV_MAPPED"             "BWA_MAPPED_OVERALL"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "ISS_MAPPED_OVERALL"    "ISS_MAPPED_PP"        </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names and types</span></span></span>
<span class="r-in"><span><span class="fu">iss_stats_specs</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 14 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    names                 types transform flag     tag              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> RUN_NAME              char  <span style="color: #949494;">&lt;NULL&gt;</span>    required <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> POOL                  char  <span style="color: #949494;">&lt;NULL&gt;</span>    required vispa_concatenate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> TAG                   char  <span style="color: #949494;">&lt;formula&gt;</span> required tag_seq          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> RAW_READS             int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> QUALITY_PASSED        int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> PHIX_MAPPING          int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> PLASMID_MAPPED_BYPOOL int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> BARCODE_MUX           int   <span style="color: #949494;">&lt;NULL&gt;</span>    required <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> LTR_IDENTIFIED        int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> TRIMMING_FINAL_LTRLC  int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> LV_MAPPED             int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> BWA_MAPPED_OVERALL    int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> ISS_MAPPED_OVERALL    int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> ISS_MAPPED_PP         int   <span style="color: #949494;">&lt;NULL&gt;</span>    optional <span style="color: #BB0000;">NA</span>               </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Names only</span></span></span>
<span class="r-in"><span><span class="fu">matrix_file_suffixes</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 10 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    quantification   matrix_type   file_suffix                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> seqCount         annotated     seqCount_matrix.no0.annotated.tsv.gz        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> seqCount         not_annotated seqCount_matrix.tsv.gz                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> fragmentEstimate annotated     fragmentEstimate_matrix.no0.annotated.tsv.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> fragmentEstimate not_annotated fragmentEstimate_matrix.tsv.gz              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> barcodeCount     annotated     barcodeCount_matrix.no0.annotated.tsv.gz    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> barcodeCount     not_annotated barcodeCount_matrix.tsv.gz                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> cellCount        annotated     cellCount_matrix.no0.annotated.tsv.gz       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> cellCount        not_annotated cellCount_matrix.tsv.gz                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> ShsCount         annotated     ShsCount_matrix.no0.annotated.tsv.gz        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> ShsCount         not_annotated ShsCount_matrix.tsv.gz                      </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Giulia Pais, Andrea Calabria, Giulio Spinozzi.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

