<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ ISAnalytics</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png"><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ISAnalytics</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.5.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/ISAnalytics.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/workflow_start.html">Setting up the workflow and first steps</a>
    <a class="dropdown-item" href="../articles/sharing_analyses.html">Sharing analyses with ISAnalytics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="nav-link" href="https://calabrialab.github.io/ISAnalytics/">
    <span class="fas fa fas fa-code-branch"></span>
     
    RELEASE
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="import-functions">Import functions<a class="anchor" aria-label="anchor" href="#import-functions"></a></h2>
      
      <p class="section-desc"></p><p>Importing integration matrices and metadata from files</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="import_association_file.html">import_association_file()</a></code> 
        </dt>
        <dd>Import the association file from disk</dd>
      </dl><dl><dt>
          
          <code><a href="import_Vispa2_stats.html">import_Vispa2_stats()</a></code> 
        </dt>
        <dd>Import Vispa2 stats given the aligned association file.</dd>
      </dl><dl><dt>
          
          <code><a href="import_single_Vispa2Matrix.html">import_single_Vispa2Matrix()</a></code> 
        </dt>
        <dd>Import a single integration matrix from file</dd>
      </dl><dl><dt>
          
          <code><a href="import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices()</a></code> 
        </dt>
        <dd>Import integration matrices from paths in the association file.</dd>
      </dl></div><div class="section level2">
      <h2 id="data-cleaning-and-pre-processing">Data cleaning and pre-processing<a class="anchor" aria-label="anchor" href="#data-cleaning-and-pre-processing"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="annotation_issues.html">annotation_issues()</a></code> 
        </dt>
        <dd>Check for genomic annotation problems in IS matrices.</dd>
      </dl><dl><dt>
          
          <code><a href="compute_near_integrations.html">compute_near_integrations()</a></code> 
        </dt>
        <dd>Scans input matrix to find and merge near integration sites.</dd>
      </dl><dl><dt>
          
          <code><a href="remove_collisions.html">remove_collisions()</a></code> 
        </dt>
        <dd>Identifies and removes collisions.</dd>
      </dl><dl><dt>
          
          <code><a href="realign_after_collisions.html">realign_after_collisions()</a></code> 
        </dt>
        <dd>Re-aligns matrices of other quantification types based on the processed
sequence count matrix.</dd>
      </dl><dl><dt>
          
          <code><a href="outlier_filter.html">outlier_filter()</a></code> 
        </dt>
        <dd>Filter out outliers in metadata, identified by the chosen outlier test.</dd>
      </dl><dl><dt>
          
          <code><a href="outliers_by_pool_fragments.html">outliers_by_pool_fragments()</a></code> 
        </dt>
        <dd>Identify and flag outliers based on pool fragments.</dd>
      </dl><dl><dt>
          
          <code><a href="aggregate_metadata.html">aggregate_metadata()</a></code> 
        </dt>
        <dd>Performs aggregation on metadata contained in the association file.</dd>
      </dl><dl><dt>
          
          <code><a href="aggregate_values_by_key.html">aggregate_values_by_key()</a></code> 
        </dt>
        <dd>Aggregates matrices values based on specified key.</dd>
      </dl><dl><dt>
          
          <code><a href="threshold_filter.html">threshold_filter()</a></code> 
        </dt>
        <dd>Filter data frames with custom predicates</dd>
      </dl></div><div class="section level2">
      <h2 id="analysis-functions">Analysis functions<a class="anchor" aria-label="anchor" href="#analysis-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="compute_abundance.html">compute_abundance()</a></code> 
        </dt>
        <dd>Computes the abundance for every integration event in the input data frame.</dd>
      </dl><dl><dt>
          
          <code><a href="top_integrations.html">top_integrations()</a></code> 
        </dt>
        <dd>Sorts and keeps the top n integration sites based on the values
in a given column.</dd>
      </dl><dl><dt>
          
          <code><a href="top_targeted_genes.html">top_targeted_genes()</a></code> 
        </dt>
        <dd>Top n targeted genes based on number of IS.</dd>
      </dl><dl><dt>
          
          <code><a href="sample_statistics.html">sample_statistics()</a></code> 
        </dt>
        <dd>Computes user specified functions on numerical columns and updates
the metadata data frame accordingly.</dd>
      </dl><dl><dt>
          
          <code><a href="CIS_grubbs.html">CIS_grubbs()</a></code> 
        </dt>
        <dd>Grubbs test for Common Insertion Sites (CIS).</dd>
      </dl><dl><dt>
          
          <code><a href="gene_frequency_fisher.html">gene_frequency_fisher()</a></code> 
        </dt>
        <dd>Compute Fisher's exact test on gene frequencies.</dd>
      </dl><dl><dt>
          
          <code><a href="cumulative_is.html">cumulative_is()</a></code> 
        </dt>
        <dd>Expands integration matrix with the cumulative IS union over time.</dd>
      </dl><dl><dt>
          
          <code><a href="is_sharing.html">is_sharing()</a></code> 
        </dt>
        <dd>Sharing of integration sites between given groups.</dd>
      </dl><dl><dt>
          
          <code><a href="purity_filter.html">purity_filter()</a></code> 
        </dt>
        <dd>Filter integration sites based on purity.</dd>
      </dl><dl><dt>
          
          <code><a href="iss_source.html">iss_source()</a></code> 
        </dt>
        <dd>Find the source of IS by evaluating sharing.</dd>
      </dl><dl><dt>
          
          <code><a href="HSC_population_size_estimate.html">HSC_population_size_estimate()</a></code> 
        </dt>
        <dd>Hematopoietic stem cells population size estimate.</dd>
      </dl></div><div class="section level2">
      <h2 id="plotting-functions">Plotting functions<a class="anchor" aria-label="anchor" href="#plotting-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="CIS_volcano_plot.html">CIS_volcano_plot()</a></code> 
        </dt>
        <dd>Trace volcano plot for computed CIS data.</dd>
      </dl><dl><dt>
          
          <code><a href="HSC_population_plot.html">HSC_population_plot()</a></code> 
        </dt>
        <dd>Plot of the estimated HSC population size for each patient.</dd>
      </dl><dl><dt>
          
          <code><a href="integration_alluvial_plot.html">integration_alluvial_plot()</a></code> 
        </dt>
        <dd>Alluvial plots for IS distribution in time.</dd>
      </dl><dl><dt>
          
          <code><a href="top_abund_tableGrob.html">top_abund_tableGrob()</a></code> 
        </dt>
        <dd>Summary top abundant tableGrobs for plots.</dd>
      </dl><dl><dt>
          
          <code><a href="sharing_heatmap.html">sharing_heatmap()</a></code> 
        </dt>
        <dd>Plot IS sharing heatmaps.</dd>
      </dl><dl><dt>
          
          <code><a href="sharing_venn.html">sharing_venn()</a></code> 
        </dt>
        <dd>Produce tables to plot sharing venn or euler diagrams.</dd>
      </dl><dl><dt>
          
          <code><a href="circos_genomic_density.html">circos_genomic_density()</a></code> 
        </dt>
        <dd>Trace a circos plot of genomic densities.</dd>
      </dl><dl><dt>
          
          <code><a href="fisher_scatterplot.html">fisher_scatterplot()</a></code> 
        </dt>
        <dd>Plot results of gene frequency Fisher's exact test.</dd>
      </dl></div><div class="section level2">
      <h2 id="utility-functions">Utility functions<a class="anchor" aria-label="anchor" href="#utility-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="as_sparse_matrix.html">as_sparse_matrix()</a></code> 
        </dt>
        <dd>Converts tidy integration matrices in the original sparse matrix
form.</dd>
      </dl><dl><dt>
          
          <code><a href="generate_blank_association_file.html">generate_blank_association_file()</a></code> 
        </dt>
        <dd>Create a blank association file.</dd>
      </dl><dl><dt>
          
          <code><a href="generate_Vispa2_launch_AF.html">generate_Vispa2_launch_AF()</a></code> 
        </dt>
        <dd>Creates a reduced association file for a VISPA2 run,
given project and pool</dd>
      </dl><dl><dt>
          
          <code><a href="inspect_tags.html">inspect_tags()</a></code> 
        </dt>
        <dd>Retrieve description of a tag by name.</dd>
      </dl><dl><dt>
          
          <code><a href="available_tags.html">available_tags()</a></code> 
        </dt>
        <dd>All available tags for dynamic vars look-up tables.</dd>
      </dl><dl><dt>
          
          <code><a href="set_mandatory_IS_vars.html">set_mandatory_IS_vars()</a></code> <code><a href="set_mandatory_IS_vars.html">set_annotation_IS_vars()</a></code> <code><a href="set_mandatory_IS_vars.html">set_af_columns_def()</a></code> <code><a href="set_mandatory_IS_vars.html">set_iss_stats_specs()</a></code> 
        </dt>
        <dd>Define custom dynamic vars.</dd>
      </dl><dl><dt>
          
          <code><a href="reset_mandatory_IS_vars.html">reset_mandatory_IS_vars()</a></code> <code><a href="reset_mandatory_IS_vars.html">reset_annotation_IS_vars()</a></code> <code><a href="reset_mandatory_IS_vars.html">reset_af_columns_def()</a></code> <code><a href="reset_mandatory_IS_vars.html">reset_iss_stats_specs()</a></code> <code><a href="reset_mandatory_IS_vars.html">reset_matrix_file_suffixes()</a></code> <code><a href="reset_mandatory_IS_vars.html">reset_dyn_vars_config()</a></code> 
        </dt>
        <dd>Resets dynamic vars to the default values.</dd>
      </dl><dl><dt>
          
          <code><a href="mandatory_IS_vars.html">mandatory_IS_vars()</a></code> <code><a href="mandatory_IS_vars.html">annotation_IS_vars()</a></code> <code><a href="mandatory_IS_vars.html">association_file_columns()</a></code> <code><a href="mandatory_IS_vars.html">iss_stats_specs()</a></code> <code><a href="mandatory_IS_vars.html">matrix_file_suffixes()</a></code> 
        </dt>
        <dd>Current dynamic vars specifications getters.</dd>
      </dl><dl><dt>
          
          <code><a href="set_matrix_file_suffixes.html">set_matrix_file_suffixes()</a></code> 
        </dt>
        <dd>Sets the look-up table for matrix file suffixes.</dd>
      </dl><dl><dt>
          
          <code><a href="pcr_id_column.html">pcr_id_column()</a></code> 
        </dt>
        <dd>Easily retrieve the name of the pcr id column.</dd>
      </dl><dl><dt>
          
          <code><a href="transform_columns.html">transform_columns()</a></code> 
        </dt>
        <dd>Apply transformations to an arbitrary number of columns.</dd>
      </dl><dl><dt>
          
          <code><a href="comparison_matrix.html">comparison_matrix()</a></code> 
        </dt>
        <dd>Obtain a single integration matrix from individual quantification
matrices.</dd>
      </dl><dl><dt>
          
          <code><a href="separate_quant_matrices.html">separate_quant_matrices()</a></code> 
        </dt>
        <dd>Separate a multiple-quantification matrix into single quantification
matrices.</dd>
      </dl><dl><dt>
          
          <code><a href="generate_default_folder_structure.html">generate_default_folder_structure()</a></code> 
        </dt>
        <dd>Generate a default folder structure, following VISPA2 standards</dd>
      </dl><dl><dt>
          
          <code><a href="NGSdataExplorer.html">NGSdataExplorer()</a></code> 
        </dt>
        <dd>Launch the shiny application NGSdataExplorer.</dd>
      </dl></div><div class="section level2">
      <h2 id="exported-variables-and-helpers">Exported variables and helpers<a class="anchor" aria-label="anchor" href="#exported-variables-and-helpers"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="date_formats.html">date_formats()</a></code> 
        </dt>
        <dd>Possible choices for the <code>dates_format</code> parameter in
<code>import_association_file</code>,
<code>import_parallel_vispa2Matrices_interactive</code> and
<code>import_parallel_vispa2Matrices_auto</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="matching_options.html">matching_options()</a></code> 
        </dt>
        <dd>Possible choices for the <code>matching_opt</code> parameter.</dd>
      </dl><dl><dt>
          
          <code><a href="quantification_types.html">quantification_types()</a></code> 
        </dt>
        <dd>Possible choices for the <code>quantification_type</code> parameter.</dd>
      </dl><dl><dt>
          
          <code><a href="reduced_AF_columns.html">reduced_AF_columns()</a></code> 
        </dt>
        <dd>Names of the columns of the association file to consider for
Vispa2 launch.</dd>
      </dl><dl><dt>
          
          <code><a href="default_stats.html">default_stats()</a></code> 
        </dt>
        <dd>A set of pre-defined functions for <code>sample_statistics</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="clinical_relevant_suspicious_genes.html">clinical_relevant_suspicious_genes()</a></code> 
        </dt>
        <dd>Clinical relevant suspicious genes (for mouse and human).</dd>
      </dl><dl><dt>
          
          <code><a href="known_clinical_oncogenes.html">known_clinical_oncogenes()</a></code> 
        </dt>
        <dd>Known clinical oncogenes (for mouse and human).</dd>
      </dl><dl><dt>
          
          <code><a href="available_outlier_tests.html">available_outlier_tests()</a></code> 
        </dt>
        <dd>A character vector containing all the names of the currently supported
outliers tests that can be called in the function outlier_filter.</dd>
      </dl><dl><dt>
          
          <code><a href="blood_lineages_default.html">blood_lineages_default()</a></code> 
        </dt>
        <dd>Default blood lineages info</dd>
      </dl><dl><dt>
          
          <code><a href="default_iss_file_prefixes.html">default_iss_file_prefixes()</a></code> 
        </dt>
        <dd>Default regex prefixes for Vispa2 stats files.</dd>
      </dl><dl><dt>
          
          <code><a href="default_meta_agg.html">default_meta_agg()</a></code> 
        </dt>
        <dd>Default metadata aggregation function table</dd>
      </dl><dl><dt>
          
          <code><a href="default_report_path.html">default_report_path()</a></code> 
        </dt>
        <dd>Default folder for saving ISAnalytics reports. Supplied as default
argument for several functions.</dd>
      </dl><dl><dt>
          
          <code><a href="refGene_table_cols.html">refGene_table_cols()</a></code> 
        </dt>
        <dd>Required columns for refGene file.</dd>
      </dl><dl><dt>
          
          <code><a href="default_af_transform.html">default_af_transform()</a></code> 
        </dt>
        <dd>Default transformations to apply to association file columns.</dd>
      </dl><dl><dt>
          
          <code><a href="default_rec_agg_lambdas.html">default_rec_agg_lambdas()</a></code> 
        </dt>
        <dd>Defaults for column aggregations in <code><a href="../reference/compute_near_integrations.html">compute_near_integrations()</a></code>.</dd>
      </dl></div><div class="section level2">
      <h2 id="deprecated">Deprecated<a class="anchor" aria-label="anchor" href="#deprecated"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="import_parallel_Vispa2Matrices_interactive.html">import_parallel_Vispa2Matrices_interactive()</a></code> 
        </dt>
        <dd>Import integration matrices from association file.</dd>
      </dl><dl><dt>
          
          <code><a href="import_parallel_Vispa2Matrices_auto.html">import_parallel_Vispa2Matrices_auto()</a></code> 
        </dt>
        <dd>Import integration matrices from association file.</dd>
      </dl><dl><dt>
          
          <code><a href="cumulative_count_union.html">cumulative_count_union()</a></code> 
        </dt>
        <dd>Integrations cumulative count in time by sample</dd>
      </dl><dl><dt>
          
          <code><a href="unzip_file_system.html">unzip_file_system()</a></code> 
        </dt>
        <dd>A utility function to unzip and use example file systems included in the
package</dd>
      </dl></div><div class="section level2">
      <h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="association_file.html">association_file</a></code> 
        </dt>
        <dd>Example of association file.</dd>
      </dl><dl><dt>
          
          <code><a href="integration_matrices.html">integration_matrices</a></code> 
        </dt>
        <dd>Example of imported multi-quantification integration matrices.</dd>
      </dl><dl><dt>
          
          <code><a href="proto_oncogenes.html">proto_oncogenes</a></code> <code><a href="proto_oncogenes.html">tumor_suppressors</a></code> 
        </dt>
        <dd>Data frames for proto-oncogenes (human and mouse)
amd tumor-suppressor genes from UniProt.</dd>
      </dl><dl><dt>
          
          <code><a href="refGenes_hg19.html">refGenes_hg19</a></code> <code><a href="refGenes_hg19.html">refGenes_mm9</a></code> 
        </dt>
        <dd>Gene annotation files for hg19, mm9 and mm10.</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

