<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>ISAnalytics: Analyze gene therapy vector insertion sites data
identified from genomics next generation sequencing reads for
clonal tracking studies — ISAnalytics • ISAnalytics</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="ISAnalytics: Analyze gene therapy vector insertion sites data
identified from genomics next generation sequencing reads for
clonal tracking studies — ISAnalytics" />
<meta property="og:description" content="

In gene therapy, stem cells are modified using viral
vectors to deliver the therapeutic transgene and replace functional
properties since the genetic modification is stable and inherited in
all cell progeny. The retrieval and mapping of the sequences flanking
the virus-host DNA junctions allows the identification of insertion
sites (IS), essential for monitoring the evolution of genetically
modified cells in vivo. A comprehensive toolkit for the analysis of
IS is required to foster clonal trackign studies and supporting the
assessment of safety and long term efficacy in vivo. This package
is aimed at (1) supporting automation of IS workflow, (2) performing
base and advance analysis for IS tracking (clonal abundance, clonal
expansions and statistics for insertional mutagenesis, etc.),
(3) providing basic biology insights of transduced stem cells in vivo." />
<meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png" />


<meta name="robots" content="noindex">

<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISAnalytics</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">1.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/aggregate_function_usage.html">Working with aggregate functions</a>
    </li>
    <li>
      <a href="../articles/collision_removal.html">Collision removal functionality</a>
    </li>
    <li>
      <a href="../articles/how_to_import_functions.html">How to use import functions</a>
    </li>
    <li>
      <a href="../articles/sharing_analyses.html">Sharing analyses with ISAnalytics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://calabrialab.github.io/ISAnalytics/">
    <span class="fas fa-code-branch"></span>
     
    RELEASE
  </a>
</li>
<li>
  <a href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ISAnalytics: Analyze gene therapy vector insertion sites data
identified from genomics next generation sequencing reads for
clonal tracking studies</h1>
    <small class="dont-index">Source: <a href='https://github.com/calabrialab/ISAnalytics/blob/master/R/ISAnalytics-package.R'><code>R/ISAnalytics-package.R</code></a></small>
    <div class="hidden name"><code>ISAnalytics.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><a href='https://lifecycle.r-lib.org/articles/stages.html#stable'><img src='figures/lifecycle-stable.svg' alt='[Stable]' /></a>

In gene therapy, stem cells are modified using viral
vectors to deliver the therapeutic transgene and replace functional
properties since the genetic modification is stable and inherited in
all cell progeny. The retrieval and mapping of the sequences flanking
the virus-host DNA junctions allows the identification of insertion
sites (IS), essential for monitoring the evolution of genetically
modified cells in vivo. A comprehensive toolkit for the analysis of
IS is required to foster clonal trackign studies and supporting the
assessment of safety and long term efficacy in vivo. This package
is aimed at (1) supporting automation of IS workflow, (2) performing
base and advance analysis for IS tracking (clonal abundance, clonal
expansions and statistics for insertional mutagenesis, etc.),
(3) providing basic biology insights of transduced stem cells in vivo.</p>
    </div>



    <h2 class="hasAnchor" id="useful-resources"><a class="anchor" href="#useful-resources"></a>Useful resources</h2>

    

<ul>
<li><p><a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5702242/'>VISPA2:
A Scalable Pipeline for High-Throughput Identification
and Annotation of Vector Integration Sites</a></p></li>
</ul>

    <h2 class="hasAnchor" id="isanalytics-function-families"><a class="anchor" href="#isanalytics-function-families"></a>ISAnalytics function families</h2>

    

<ul>
<li><p>Import functions:</p><ul>
<li><p><code><a href='import_single_Vispa2Matrix.html'>import_single_Vispa2Matrix</a></code></p></li>
<li><p><code><a href='import_association_file.html'>import_association_file</a></code></p></li>
<li><p><code><a href='import_Vispa2_stats.html'>import_Vispa2_stats</a></code></p></li>
<li><p><code><a href='import_parallel_Vispa2Matrices.html'>import_parallel_Vispa2Matrices</a></code></p></li>
</ul></li>
<li><p>Aggregation functions:</p><ul>
<li><p><code><a href='aggregate_metadata.html'>aggregate_metadata</a></code></p></li>
<li><p><code><a href='aggregate_values_by_key.html'>aggregate_values_by_key</a></code></p></li>
</ul></li>
<li><p>Collision removal functions:</p><ul>
<li><p><code><a href='remove_collisions.html'>remove_collisions</a></code></p></li>
<li><p><code><a href='realign_after_collisions.html'>realign_after_collisions</a></code></p></li>
</ul></li>
<li><p>Removal of outliers from raw reads</p><ul>
<li><p><code><a href='outlier_filter.html'>outlier_filter</a></code></p></li>
<li><p><code><a href='outliers_by_pool_fragments.html'>outliers_by_pool_fragments</a></code></p></li>
</ul></li>
<li><p>Recalibration functions:</p><ul>
<li><p><code><a href='compute_near_integrations.html'>compute_near_integrations</a></code></p></li>
</ul></li>
<li><p>Analysis functions:</p><ul>
<li><p><code><a href='compute_abundance.html'>compute_abundance</a></code></p></li>
<li><p><code><a href='comparison_matrix.html'>comparison_matrix</a></code></p></li>
<li><p><code><a href='separate_quant_matrices.html'>separate_quant_matrices</a></code></p></li>
<li><p><code><a href='threshold_filter.html'>threshold_filter</a></code></p></li>
<li><p><code><a href='top_integrations.html'>top_integrations</a></code></p></li>
<li><p><code><a href='sample_statistics.html'>sample_statistics</a></code></p></li>
<li><p><code><a href='CIS_grubbs.html'>CIS_grubbs</a></code></p></li>
<li><p><code><a href='cumulative_count_union.html'>cumulative_count_union</a></code></p></li>
<li><p><code><a href='is_sharing.html'>is_sharing</a></code></p></li>
<li><p><code><a href='cumulative_is.html'>cumulative_is</a></code></p></li>
</ul></li>
<li><p>HSC population size estimate:</p><ul>
<li><p><code><a href='HSC_population_size_estimate.html'>HSC_population_size_estimate</a></code></p></li>
</ul></li>
<li><p>Plotting functions:</p><ul>
<li><p><code><a href='CIS_volcano_plot.html'>CIS_volcano_plot</a></code></p></li>
<li><p><code><a href='HSC_population_plot.html'>HSC_population_plot</a></code></p></li>
<li><p><code><a href='sharing_heatmap.html'>sharing_heatmap</a></code></p></li>
<li><p><code><a href='integration_alluvial_plot.html'>integration_alluvial_plot</a></code></p></li>
<li><p><code><a href='top_abund_tableGrob.html'>top_abund_tableGrob</a></code></p></li>
<li><p><code><a href='circos_genomic_density.html'>circos_genomic_density</a></code></p></li>
<li><p><code><a href='sharing_venn.html'>sharing_venn</a></code></p></li>
</ul></li>
<li><p>Utility functions:</p><ul>
<li><p><code><a href='generate_blank_association_file.html'>generate_blank_association_file</a></code></p></li>
<li><p><code><a href='generate_Vispa2_launch_AF.html'>generate_Vispa2_launch_AF</a></code></p></li>
<li><p><code><a href='unzip_file_system.html'>unzip_file_system</a></code></p></li>
<li><p><code><a href='as_sparse_matrix.html'>as_sparse_matrix</a></code></p></li>
</ul></li>
</ul>

    <h2 class="hasAnchor" id="vignettes"><a class="anchor" href="#vignettes"></a>Vignettes</h2>

    

<ul>
<li><p><code><a href='../articles/how_to_import_functions.html'>vignette("how_to_import_functions", package = "ISAnalytics")</a></code></p></li>
<li><p><code><a href='../articles/collision_removal.html'>vignette("collision_removal",
package = "ISAnalytics")</a></code></p></li>
<li><p><code><a href='../articles/aggregate_function_usage.html'>vignette("aggregate_function_usage",
package = "ISAnalytics")</a></code></p></li>
<li><p><code><a href='../articles/sharing_analyses.html'>vignette("sharing_analyses",
package = "ISAnalytics")</a></code></p></li>
</ul>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


