<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="
Identify and flag outliers based on expected number of raw reads per pool."><title>Identify and flag outliers based on pool fragments. — outliers_by_pool_fragments • ISAnalytics</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identify and flag outliers based on pool fragments. — outliers_by_pool_fragments"><meta property="og:description" content="
Identify and flag outliers based on expected number of raw reads per pool."><meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png"><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ISAnalytics</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.9.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/ISAnalytics.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/workflow_start.html">Setting up the workflow and first steps</a>
    <a class="dropdown-item" href="../articles/sharing_analyses.html">Sharing analyses with ISAnalytics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="nav-link" href="https://calabrialab.github.io/ISAnalytics/">
    <span class="fas fa fas fa-code-branch"></span>
     
    RELEASE
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Identify and flag outliers based on pool fragments.</h1>
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/HEAD/R/outlier-filtering.R" class="external-link"><code>R/outlier-filtering.R</code></a></small>
      <div class="d-none name"><code>outliers_by_pool_fragments.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="figures/lifecycle-stable.svg" alt="[Stable]"></a>
Identify and flag outliers based on expected number of raw reads per pool.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">outliers_by_pool_fragments</span><span class="op">(</span></span>
<span>  <span class="va">metadata</span>,</span>
<span>  key <span class="op">=</span> <span class="st">"BARCODE_MUX"</span>,</span>
<span>  outlier_p_value_threshold <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  normality_test <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  normality_p_value_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  transform_log2 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  per_pool_test <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pool_col <span class="op">=</span> <span class="st">"PoolID"</span>,</span>
<span>  min_samples_per_pool <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  flag_logic <span class="op">=</span> <span class="st">"AND"</span>,</span>
<span>  keep_calc_cols <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  report_path <span class="op">=</span> <span class="fu"><a href="default_report_path.html">default_report_path</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>metadata</dt>
<dd><p>The metadata data frame</p></dd>


<dt>key</dt>
<dd><p>A character vector of numeric column names</p></dd>


<dt>outlier_p_value_threshold</dt>
<dd><p>The p value threshold for a read to be
considered an outlier</p></dd>


<dt>normality_test</dt>
<dd><p>Perform normality test? Normality is assessed for
each column in the key using Shapiro-Wilk test and if the values do not
follow a normal distribution, other calculations are skipped</p></dd>


<dt>normality_p_value_threshold</dt>
<dd><p>Normality threshold</p></dd>


<dt>transform_log2</dt>
<dd><p>Perform a log2 trasformation on values prior the
actual calculations?</p></dd>


<dt>per_pool_test</dt>
<dd><p>Perform the test for each pool?</p></dd>


<dt>pool_col</dt>
<dd><p>A character vector of the names of the columns that
uniquely identify a pool</p></dd>


<dt>min_samples_per_pool</dt>
<dd><p>The minimum number of samples that a pool
needs to contain in order to be processed - relevant only if
<code>per_pool_test = TRUE</code></p></dd>


<dt>flag_logic</dt>
<dd><p>A character vector of logic operators to obtain a
global flag formula - only relevant if the key is longer than one.
All operators must be chosen between:
AND, OR, XOR, NAND, NOR, XNOR</p></dd>


<dt>keep_calc_cols</dt>
<dd><p>Keep the calculation columns in the output data frame?</p></dd>


<dt>report_path</dt>
<dd><p>The path where the report file should be saved.
Can be a folder, a file or NULL if no report should be produced.
Defaults to <code>{user_home}/ISAnalytics_reports</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data frame of metadata with the column <code>to_remove</code></p>


    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    
<div class="section">
<h3 id="modular-structure">Modular structure<a class="anchor" aria-label="anchor" href="#modular-structure"></a></h3>


<p>The outlier filtering functions are structured in a modular fashion.
There are 2 kind of functions:</p><ul><li><p>Outlier tests - Functions that perform some kind of calculation based
on inputs and flags metadata</p></li>
<li><p>Outlier filter - A function that takes one or more outlier tests,
combines all the flags with a given logic and filters out rows that
are flagged as outliers</p></li>
</ul><p>This function is an outlier test, and calculates for each column in the key</p><ul><li><p>The zscore of the values</p></li>
<li><p>The tstudent of the values</p></li>
<li><p>The the associated p-value (tdist)</p></li>
</ul><p>Optionally the test can be performed for each pool and a normality test
can be run prior the actual calculations.
Samples are flagged if this condition is respected:</p><ul><li><p>tdist &lt; outlier_p_value_threshold &amp; zscore &lt; 0</p></li>
</ul><p>If the key contains more than one column an additional flag logic can be
specified for combining the results.
Example:
let's suppose the key contains the names of two columns, X and Y
<code>key = c("X", "Y")</code>
if we specify the the argument <code>flag_logic = "AND"</code> then the reads will
be flagged based on this global condition:
(tdist_X &lt; outlier_p_value_threshold &amp; zscore_X &lt; 0) AND
(tdist_Y &lt; outlier_p_value_threshold &amp; zscore_Y &lt; 0)</p>
<p>The user can specify one or more logical operators that will be applied
in sequence.</p>
</div>

    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other Data cleaning and pre-processing: 
<code><a href="aggregate_metadata.html">aggregate_metadata</a>()</code>,
<code><a href="aggregate_values_by_key.html">aggregate_values_by_key</a>()</code>,
<code><a href="compute_near_integrations.html">compute_near_integrations</a>()</code>,
<code><a href="default_meta_agg.html">default_meta_agg</a>()</code>,
<code><a href="outlier_filter.html">outlier_filter</a>()</code>,
<code><a href="purity_filter.html">purity_filter</a>()</code>,
<code><a href="realign_after_collisions.html">realign_after_collisions</a>()</code>,
<code><a href="remove_collisions.html">remove_collisions</a>()</code>,
<code><a href="threshold_filter.html">threshold_filter</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"association_file"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">flagged</span> <span class="op">&lt;-</span> <span class="fu">outliers_by_pool_fragments</span><span class="op">(</span><span class="va">association_file</span>,</span></span>
<span class="r-in"><span>    report_path <span class="op">=</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing NAs from data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Log2 transformation, removing values &lt;= 0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">flagged</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ProjectID  FUSIONID PoolID TagSequence SubjectID VectorType VectorID</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:      PJ01 ET#382.46 POOL01   LTR75LC38     PT001      lenti    GLOBE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:      PJ01 ET#381.40 POOL01   LTR53LC32     PT001      lenti    GLOBE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:      PJ01  ET#381.9 POOL01   LTR83LC66     PT001      lenti    GLOBE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:      PJ01 ET#381.71 POOL01   LTR27LC94     PT001      lenti    GLOBE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:      PJ01  ET#381.2 POOL01   LTR69LC52     PT001      lenti    GLOBE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:      PJ01 ET#382.28 POOL01    LTR37LC2     PT001      lenti    GLOBE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ExperimentID Tissue TimePoint DNAFragmentation PCRMethod TagIDextended</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:         &lt;NA&gt;     PB      0060            SONIC      SLiM     LTR75LC38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:         &lt;NA&gt;     BM      0180            SONIC      SLiM     LTR53LC32</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:         &lt;NA&gt;     BM      0180            SONIC      SLiM     LTR83LC66</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:         &lt;NA&gt;     BM      0180            SONIC      SLiM     LTR27LC94</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:         &lt;NA&gt;     PB      0180            SONIC      SLiM     LTR69LC52</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:         &lt;NA&gt;     BM      0060            SONIC      SLiM      LTR37LC2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Keywords CellMarker      TagID NGSProvider NGSTechnology ConverrtedFilesDir</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:     &lt;NA&gt;        MNC LTR75.LC38        &lt;NA&gt;         HiSeq               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:     &lt;NA&gt;        MNC LTR53.LC32        &lt;NA&gt;         HiSeq               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:     &lt;NA&gt;        MNC LTR83.LC66        &lt;NA&gt;         HiSeq               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:     &lt;NA&gt;        MNC LTR27.LC94        &lt;NA&gt;         HiSeq               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:     &lt;NA&gt;        MNC LTR69.LC52        &lt;NA&gt;         HiSeq               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:     &lt;NA&gt;        MNC  LTR37.LC2        &lt;NA&gt;         HiSeq               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ConverrtedFilesName SourceFileFolder SourceFileNameR1 SourceFileNameR2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:                &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:                &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:                &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:                &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:                &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:                &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    DNAnumber ReplicateNumber DNAextractionDate DNAngUsed LinearPCRID</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1: PT001-103               3        2016-03-16    23.184        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:  PT001-81               2        2016-07-15   181.440        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:  PT001-81               1        2016-07-15   181.440        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:  PT001-81               3        2016-07-15   181.440        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:  PT001-74               1        2016-07-15    23.058        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6: PT001-107               2        2016-03-16   171.360        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    LinearPCRDate SonicationDate LigationDate 1stExpoPCRID 1stExpoPCRDate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:          &lt;NA&gt;     2016-11-02   2016-11-02    ET#380.46     2016-11-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:          &lt;NA&gt;     2016-11-02   2016-11-02    ET#379.40     2016-11-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:          &lt;NA&gt;     2016-11-02   2016-11-02     ET#379.9     2016-11-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:          &lt;NA&gt;     2016-11-02   2016-11-02    ET#379.71     2016-11-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:          &lt;NA&gt;     2016-11-02   2016-11-02     ET#379.2     2016-11-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:          &lt;NA&gt;     2016-11-02   2016-11-02    ET#380.28     2016-11-02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    2ndExpoID 2ndExpoDate FusionPrimerPCRID FusionPrimerPCRDate   PoolDate</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:      &lt;NA&gt;        &lt;NA&gt;         ET#382.46          2016-11-03 2016-11-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:      &lt;NA&gt;        &lt;NA&gt;         ET#381.40          2016-11-03 2016-11-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:      &lt;NA&gt;        &lt;NA&gt;          ET#381.9          2016-11-03 2016-11-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:      &lt;NA&gt;        &lt;NA&gt;         ET#381.71          2016-11-03 2016-11-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:      &lt;NA&gt;        &lt;NA&gt;          ET#381.2          2016-11-03 2016-11-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:      &lt;NA&gt;        &lt;NA&gt;         ET#382.28          2016-11-03 2016-11-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    SequencingDate  VCN Genome SequencingRound Genotype TestGroup  MOI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:     2016-11-15 0.30   hg19               1     &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:     2016-11-15 0.27   hg19               1     &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:     2016-11-15 0.27   hg19               1     &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:     2016-11-15 0.27   hg19               1     &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:     2016-11-15 0.24   hg19               1     &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:     2016-11-15 0.42   hg19               1     &lt;NA&gt;      &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Engraftment Transduction Notes AddedField1 AddedField2 AddedField3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:          NA           NA  &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:          NA           NA  &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:          NA           NA  &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:          NA           NA  &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:          NA           NA  &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:          NA           NA  &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AddedField4 concatenatePoolIDSeqRun AddedField6_RelativeBloodPercentage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        &lt;NA&gt;                POOL01-1                                &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        &lt;NA&gt;                POOL01-1                                &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        &lt;NA&gt;                POOL01-1                                &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        &lt;NA&gt;                POOL01-1                                &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:        &lt;NA&gt;                POOL01-1                                &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:        &lt;NA&gt;                POOL01-1                                &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    AddedField7_PurityTestFeasibility AddedField8_FacsSeparationPurity Kapa</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:                                NA                               NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:                                NA                               NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:                                NA                               NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:                                NA                               NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:                                NA                               NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:                                NA                               NA   NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ulForPool</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                 CompleteAmplificationID</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1: PJ01_POOL01_LTR75LC38_PT001_PT001-103_lenti_GLOBE_PB_1_SLiM_0060_MNC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:  PJ01_POOL01_LTR53LC32_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:  PJ01_POOL01_LTR83LC66_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:  PJ01_POOL01_LTR27LC94_PT001_PT001-81_lenti_GLOBE_BM_1_SLiM_0180_MNC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:  PJ01_POOL01_LTR69LC52_PT001_PT001-74_lenti_GLOBE_PB_1_SLiM_0180_MNC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:  PJ01_POOL01_LTR37LC2_PT001_PT001-107_lenti_GLOBE_BM_1_SLiM_0060_MNC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  UniqueID StudyTestID StudyTestGroup MouseID Tigroup Tisource</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1: ID00000000000000007433        &lt;NA&gt;             NA      NA    &lt;NA&gt;     &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2: ID00000000000000007340        &lt;NA&gt;             NA      NA    &lt;NA&gt;     &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3: ID00000000000000007310        &lt;NA&gt;             NA      NA    &lt;NA&gt;     &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4: ID00000000000000007370        &lt;NA&gt;             NA      NA    &lt;NA&gt;     &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5: ID00000000000000007303        &lt;NA&gt;             NA      NA    &lt;NA&gt;     &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6: ID00000000000000007417        &lt;NA&gt;             NA      NA    &lt;NA&gt;     &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    PathToFolderProjectID SamplesNameCheck TimepointDays TimepointMonths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:                 /PJ01             &lt;NA&gt;          0060              02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:                 /PJ01             &lt;NA&gt;          0180              06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:                 /PJ01             &lt;NA&gt;          0180              06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:                 /PJ01             &lt;NA&gt;          0180              06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:                 /PJ01             &lt;NA&gt;          0180              06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:                 /PJ01             &lt;NA&gt;          0060              02</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    TimepointYears ng DNA corrected      RUN_NAME PHIX_MAPPING</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:             01            23.18 PJ01|POOL01-1     43586699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:             01           181.44 PJ01|POOL01-1     43586699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:             01           181.44 PJ01|POOL01-1     43586699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:             01           181.44 PJ01|POOL01-1     43586699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:             01            23.06 PJ01|POOL01-1     43586699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:             01           171.36 PJ01|POOL01-1     43586699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    PLASMID_MAPPED_BYPOOL BARCODE_MUX LTR_IDENTIFIED TRIMMING_FINAL_LTRLC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:               2256176      645026         645026               630965</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:               2256176      652208         652177               649044</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:               2256176      451519         451512               449669</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:               2256176      426500         426499               425666</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:               2256176       18300          18300                18290</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:               2256176      729327         729327               727219</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    LV_MAPPED BWA_MAPPED_OVERALL ISS_MAPPED_OVERALL RAW_READS QUALITY_PASSED</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:    211757             402477             219452        NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:    303300             322086             222646        NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:    204810             227275             149385        NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:    185752             223915             143283        NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:      6962              10487               5907        NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:    318653             369117             235640        NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ISS_MAPPED_PP log2_BARCODE_MUX normality_BARCODE_MUX zscore_BARCODE_MUX</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:            NA         19.29900                    NA          0.7338238</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:            NA         19.31497                    NA          0.7411949</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:            NA         18.78443                    NA          0.4963905</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:            NA         18.70219                    NA          0.4584429</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:            NA         14.15956                    NA         -1.6376186</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:            NA         19.47621                    NA          0.8155914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    tstudent_BARCODE_MUX tdist_BARCODE_MUX processed flag_BARCODE_MUX to_remove</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:            0.7422529        0.29678647      TRUE            FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:            0.7498983        0.29506532      TRUE            FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:            0.4987165        0.34660114      TRUE            FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:            0.4602087        0.35328569      TRUE            FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:            1.7457131        0.08872463      TRUE            FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:            0.8274007        0.27728811      TRUE            FALSE     FALSE</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

