<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Working with aggregate functions • ISAnalytics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/slate/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Working with aggregate functions">
<meta property="og:description" content="ISAnalytics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISAnalytics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/aggregate_function_usage.html">Working with aggregate functions</a>
    </li>
    <li>
      <a href="../articles/collision_removal.html">Collision removal functionality</a>
    </li>
    <li>
      <a href="../articles/how_to_import_functions.html">How to use import functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="aggregate_function_usage_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Working with aggregate functions</h1>
                        <h4 class="author">Giulia Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget, Via Olgettina 60, 20132 Milano - Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 class="date">21 September 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/master/vignettes/aggregate_function_usage.Rmd"><code>vignettes/aggregate_function_usage.Rmd</code></a></small>
      <div class="hidden name"><code>aggregate_function_usage.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>In this vignette we’re going to explain in detail how to use functions of the aggregate family, namely:</p>
<ol style="list-style-type: decimal">
<li><code>aggregate_metadata</code></li>
<li><code>aggregate_values_by_key</code></li>
</ol>
<div id="how-to-install-isanalytics" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-install-isanalytics" class="anchor"></a>How to install ISAnalytics</h2>
<p>To install the package run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, quietly = <span class="fl">TRUE</span>)) {
      <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
  }
<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"ISAnalytics"</span>)
</pre></div>
<p>To install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="kw"><a href="https://devtools.r-lib.org">devtools</a></span>)) {
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
}
<span class="kw">devtools</span>::<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"calabrialab/ISAnalytics"</span>, build_vignettes = <span class="fl">TRUE</span>)
</pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com//calabrialab/isanalytics">ISAnalytics</a></span>)
<span class="co">#&gt; Loading required package: magrittr</span>
</pre></div>
</div>
<div id="setting-options" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-options" class="anchor"></a>Setting options</h2>
<p><code>ISAnalytics</code> has a verbose option that allows some functions to print additional information to the console while they’re executing. To disable this feature do:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># DISABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.verbose"</span> = <span class="fl">FALSE</span>)

<span class="co"># ENABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.verbose"</span> = <span class="fl">TRUE</span>)
</pre></div>
<p>Some functions also produce report in a user-friendly HTML format, to set this feature:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># DISABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.widgets"</span> = <span class="fl">FALSE</span>)

<span class="co"># ENABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.widgets"</span> = <span class="fl">TRUE</span>)
</pre></div>
</div>
</div>
<div id="aggregating-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#aggregating-metadata" class="anchor"></a>Aggregating metadata</h1>
<p>We refer to information contained in the association file as “metadata”: sometimes it’s useful to obtain collective information based on a certain group of variables we’re interested in. The function <code>aggregate_metadata</code> does just that, according to the grouping variables, meaning the names of the columns in the association file to perform a <code>group_by</code> operation with, creates a summary which includes:</p>
<ul>
<li>
<code>FusionPrimerPCRDate</code> - The minimum date in the group for this variable</li>
<li>
<code>LinearPCRDate</code> - The minimum date in the group for this variable</li>
<li>
<code>VCN</code> - The mean of “VCN” column for each group</li>
<li>
<code>Avg_DNAngUsed</code> - The mean of “DNAngUsed”column for each group</li>
<li>
<code>Kapa</code> - The mean of “Kapa” column for each group</li>
<li>
<code>DNAngUsed</code> - The sum of “DNAngUsed” column for each group</li>
<li>
<code>ulForPool</code> - The sum of “ulForPool” column for each group</li>
<li>
<code>AggregateMeta</code> - A string obtained by concatenation of all of the grouping variables separated by "_"</li>
</ul>
<div id="typical-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#typical-workflow" class="anchor"></a>Typical workflow</h2>
<p>Import the association file via <code>import_assocition_file</code>. If you need more information on import function please view the vignette “How to use import functions”.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(ISAnalytics.widgets = <span class="fl">FALSE</span>)
<span class="kw">path_AF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ex_association_file.tsv"</span>,
    package = <span class="st">"ISAnalytics"</span>
)

<span class="kw">root_correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"fs.zip"</span>,
    package = <span class="st">"ISAnalytics"</span>
)
<span class="kw">root_correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unzip_file_system.html">unzip_file_system</a></span>(<span class="kw">root_correct</span>, <span class="st">"fs"</span>)
<span class="kw">association_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_association_file.html">import_association_file</a></span>(<span class="kw">path_AF</span>, <span class="kw">root_correct</span>)
</pre></div>
<p>Perform aggregation:</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">aggregated_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_metadata.html">aggregate_metadata</a></span>(<span class="kw">association_file</span>,
    grouping_keys = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
        <span class="st">"SubjectID"</span>,
        <span class="st">"CellMarker"</span>,
        <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span>
    ),
    import_stats = <span class="fl">FALSE</span>
)
</pre></div>
<table class="table">
<colgroup>
<col width="18%">
<col width="8%">
<col width="7%">
<col width="4%">
<col width="6%">
<col width="13%">
<col width="9%">
<col width="3%">
<col width="9%">
<col width="5%">
<col width="6%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th align="left">AggregateMeta</th>
<th align="left">SubjectID</th>
<th align="left">CellMarker</th>
<th align="left">Tissue</th>
<th align="left">TimePoint</th>
<th align="left">FusionPrimerPCRDate</th>
<th align="left">LinearPCRDate</th>
<th align="right">VCN</th>
<th align="right">Avg_DNAngUsed</th>
<th align="right">Kapa</th>
<th align="right">DNAngUsed</th>
<th align="right">ulForPool</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">VA2020-mix10_NONE_NONE_0000</td>
<td align="left">VA2020-mix10</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">0.70</td>
<td align="right">100</td>
<td align="right">26.54667</td>
<td align="right">300</td>
<td align="right">25.70</td>
</tr>
<tr class="even">
<td align="left">VA2020-mix11_NONE_NONE_0000</td>
<td align="left">VA2020-mix11</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">2.80</td>
<td align="right">100</td>
<td align="right">59.81000</td>
<td align="right">300</td>
<td align="right">12.29</td>
</tr>
<tr class="odd">
<td align="left">VA2020-mix12_NONE_NONE_0000</td>
<td align="left">VA2020-mix12</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">4.30</td>
<td align="right">100</td>
<td align="right">79.80333</td>
<td align="right">300</td>
<td align="right">7.98</td>
</tr>
<tr class="even">
<td align="left">VA2020-mix13_NONE_NONE_0000</td>
<td align="left">VA2020-mix13</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">9.89</td>
<td align="right">100</td>
<td align="right">85.87000</td>
<td align="right">300</td>
<td align="right">7.46</td>
</tr>
<tr class="odd">
<td align="left">VA2020-mix14_NONE_NONE_0000</td>
<td align="left">VA2020-mix14</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
<td align="right">0.00</td>
</tr>
</tbody>
</table>
<p>As you can see there is an additional parameter you can set, <code>import_stats</code>: if set to <code>TRUE</code>, the function will automatically look into the file system you provided as root when you imported the association file and will try to locate first the <code>iss</code> folder for each project and then all Vispa2 “stats” files.<br>
Vispa2 stats contain useful information that is not included in the association file and is linked to the single Vispa2 run. In some cases it’s useful to perform aggregation on those info too. If you set the parameter to <code>TRUE</code>, besides the columns mentioned before you will also have:</p>
<ul>
<li>
<code>BARCODE_MUX</code> - The sum of the “BARCODE_MUX” column for the group</li>
<li>
<code>TRIMMING_FINAL_LTRLC</code> - The sum of the “TRIMMING_FINAL_LTRLC” column for the group</li>
<li>
<code>LV_MAPPED</code> - The sum of the “LV_MAPPED” column for the group</li>
<li>
<code>BWA_MAPPED_OVERALL</code> - The sum of the “BWA_MAPPED_OVERALL” column for the group</li>
<li>
<code>ISS_MAPPED_PP</code> - The sum of the “ISS_MAPPED_PP” column for the group</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">aggregated_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_metadata.html">aggregate_metadata</a></span>(<span class="kw">association_file</span>,
    grouping_keys = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
        <span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>,
        <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span>
    ),
    import_stats = <span class="fl">TRUE</span>
)
</pre></div>
<table class="table">
<colgroup>
<col width="12%">
<col width="5%">
<col width="4%">
<col width="3%">
<col width="4%">
<col width="8%">
<col width="6%">
<col width="2%">
<col width="6%">
<col width="3%">
<col width="4%">
<col width="4%">
<col width="5%">
<col width="9%">
<col width="4%">
<col width="8%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th align="left">AggregateMeta</th>
<th align="left">SubjectID</th>
<th align="left">CellMarker</th>
<th align="left">Tissue</th>
<th align="left">TimePoint</th>
<th align="left">FusionPrimerPCRDate</th>
<th align="left">LinearPCRDate</th>
<th align="right">VCN</th>
<th align="right">Avg_DNAngUsed</th>
<th align="right">Kapa</th>
<th align="right">DNAngUsed</th>
<th align="right">ulForPool</th>
<th align="right">BARCODE_MUX</th>
<th align="right">TRIMMING_FINAL_LTRLC</th>
<th align="right">LV_MAPPED</th>
<th align="right">BWA_MAPPED_OVERALL</th>
<th align="right">ISS_MAPPED_PP</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">VA2020-mix10_NONE_NONE_0000</td>
<td align="left">VA2020-mix10</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">0.70</td>
<td align="right">100</td>
<td align="right">26.54667</td>
<td align="right">300</td>
<td align="right">25.70</td>
<td align="right">19489792</td>
<td align="right">19455728</td>
<td align="right">8457580</td>
<td align="right">10476016</td>
<td align="right">9544692</td>
</tr>
<tr class="even">
<td align="left">VA2020-mix11_NONE_NONE_0000</td>
<td align="left">VA2020-mix11</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">2.80</td>
<td align="right">100</td>
<td align="right">59.81000</td>
<td align="right">300</td>
<td align="right">12.29</td>
<td align="right">34294696</td>
<td align="right">34199968</td>
<td align="right">13941856</td>
<td align="right">19007548</td>
<td align="right">14986620</td>
</tr>
<tr class="odd">
<td align="left">VA2020-mix12_NONE_NONE_0000</td>
<td align="left">VA2020-mix12</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">4.30</td>
<td align="right">100</td>
<td align="right">79.80333</td>
<td align="right">300</td>
<td align="right">7.98</td>
<td align="right">23364272</td>
<td align="right">23274392</td>
<td align="right">11694244</td>
<td align="right">10524788</td>
<td align="right">7752560</td>
</tr>
<tr class="even">
<td align="left">VA2020-mix13_NONE_NONE_0000</td>
<td align="left">VA2020-mix13</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">2020-04-16</td>
<td align="left">NA</td>
<td align="right">9.89</td>
<td align="right">100</td>
<td align="right">85.87000</td>
<td align="right">300</td>
<td align="right">7.46</td>
<td align="right">24345956</td>
<td align="right">24249196</td>
<td align="right">9936136</td>
<td align="right">13188028</td>
<td align="right">9321144</td>
</tr>
<tr class="odd">
<td align="left">VA2020-mix14_NONE_NONE_0000</td>
<td align="left">VA2020-mix14</td>
<td align="left">NONE</td>
<td align="left">NONE</td>
<td align="left">0000</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
<td align="right">0.00</td>
<td align="right">1161066</td>
<td align="right">1156914</td>
<td align="right">484480</td>
<td align="right">620925</td>
<td align="right">438088</td>
</tr>
</tbody>
</table>
<p>If you have the option <code>ISAnalytics.widgets</code> set to TRUE, this will produce a report in HTML format that tells you which stats files were imported. To avoid this, you can set the option to FALSE.</p>
</div>
</div>
<div id="aggregation-of-values-by-key" class="section level1">
<h1 class="hasAnchor">
<a href="#aggregation-of-values-by-key" class="anchor"></a>Aggregation of values by key</h1>
<p><code>ISAnalytics</code> contains useful functions to aggregate the values contained in your imported matrices based on a key, aka a single column or a combination of columns contained in the association file that are related to the samples.</p>
<div id="typical-workflow-1" class="section level2">
<h2 class="hasAnchor">
<a href="#typical-workflow-1" class="anchor"></a>Typical workflow</h2>
<p>Import your association file (see previous section) and then import your matrices:</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices_auto.html">import_parallel_Vispa2Matrices_auto</a></span>(
    association_file = <span class="kw">association_file</span>, root = <span class="kw">NULL</span>,
    quantification_type = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"fragmentEstimate"</span>, <span class="st">"seqCount"</span>),
    matrix_type = <span class="st">"annotated"</span>, workers = <span class="fl">2</span>, patterns = <span class="kw">NULL</span>,
    matching_opt = <span class="st">"ANY"</span>
)
</pre></div>
<p>The function <code>aggregate_values_by_key</code> can perform the aggregation both on the list of matrices and a single matrix.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co"># Takes the whole list and produces a list in output</span>
<span class="kw">aggregated_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(<span class="kw">matrices</span>, <span class="kw">association_file</span>)

<span class="co"># Takes a single matrix and produces a single matrix as output</span>
<span class="kw">aggregated_matrices_single</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(<span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>, 
                                                      <span class="kw">association_file</span>)
</pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">SubjectID</th>
<th align="right">Aggregated_value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">8884086</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix11</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">11322005</td>
<td align="left">-</td>
<td align="left">MTOR</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">11772847</td>
<td align="left">+</td>
<td align="left">DRAXIN</td>
<td align="left">+</td>
<td align="left">VA2020-mix11</td>
<td align="right">52</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">22453532</td>
<td align="left">+</td>
<td align="left">WNT4</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">8</td>
</tr>
</tbody>
</table>
<div id="changing-parameters-to-obtain-different-results" class="section level3">
<h3 class="hasAnchor">
<a href="#changing-parameters-to-obtain-different-results" class="anchor"></a>Changing parameters to obtain different results</h3>
<p>The function has several different parameters that have default values that can be changed according to user preference.</p>
<ol style="list-style-type: decimal">
<li>
<strong>Changing the <code>key</code> value</strong><br>
You can change the value of the parameter key as you see fit. This parameter should contain one or multiple columns of the association file that you want to include in the grouping when performing the aggregation. The default value is set to “SubjectID”.</li>
</ol>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">agg1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(
    x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>,
    association_file = <span class="kw">association_file</span>,
    key = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SubjectID"</span>, <span class="st">"ProjectID"</span>)
)
</pre></div>
<table class="table">
<colgroup>
<col width="4%">
<col width="19%">
<col width="7%">
<col width="10%">
<col width="11%">
<col width="14%">
<col width="13%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">SubjectID</th>
<th align="left">ProjectID</th>
<th align="right">Aggregated_value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="left">CLOEXP</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="left">PROJECT1100</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="left">PROJECT1101</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="left">CLOEXP</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="left">PROJECT1100</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="left">PROJECT1101</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>Changing the <code>lambda</code> and <code>args</code> values</strong><br>
The <code>lambda</code> parameter indicates the function to be applied to the values for aggregation. By default this is set to “sum”, but you can change it (together with other parameters also), to apply other functions, both exported from packages or fully customized to your needs.<br>
The only important note is that this function should perform some kind of aggregation on numeric values, this means in practical terms it has to accept a vector of numeric/integer values as input and produce a SINGLE value as output. Valid options for this purpose might be: <code>sum</code>, <code>mean</code>, <code>median</code>, <code>min</code>, <code>max</code> and so on.</li>
</ol>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">agg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(
    x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>,
    association_file = <span class="kw">association_file</span>,
    key = <span class="st">"SubjectID"</span>,
    lambda = <span class="st">"mean"</span>
)
</pre></div>
<p>If the function you’re using accepts other parameters than the input vector, you can specify them in the <code>args</code> parameter as a named list:</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">agg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(
    x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>,
    association_file = <span class="kw">association_file</span>,
    key = <span class="st">"SubjectID"</span>,
    lambda = <span class="st">"mean"</span>,
    args = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(na.rm = <span class="fl">TRUE</span>)
)
</pre></div>
<p>In this case the <code>mean</code> function accepts an additional parameter <code>na.rm</code> that we want to set to TRUE. Note that the arguments must be in a named list, where names are the names of the parameters to pass on to the function.</p>
<table class="table">
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">SubjectID</th>
<th align="right">Aggregated_value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">8884086</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix11</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">11322005</td>
<td align="left">-</td>
<td align="left">MTOR</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">11772847</td>
<td align="left">+</td>
<td align="left">DRAXIN</td>
<td align="left">+</td>
<td align="left">VA2020-mix11</td>
<td align="right">13</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">22453532</td>
<td align="left">+</td>
<td align="left">WNT4</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Changing the <code>group</code> value</strong><br>
The <code>group</code> parameter should contain all other variables to include in the grouping besides <code>key</code>. By default this contains <code><a href="https://rdrr.io/r/base/c.html">c("chr", "integration_locus", "strand", "GeneName", "GeneStrand")</a></code>. You can change this grouping as you see fit, if you don’t want to add any other variable to the key, just set it to NULL.</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">agg3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(
    x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>,
    association_file = <span class="kw">association_file</span>,
    key = <span class="st">"SubjectID"</span>,
    lambda = <span class="st">"mean"</span>,
    args = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(na.rm = <span class="fl">TRUE</span>),
    group = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chr"</span>, <span class="st">"integration_locus"</span>, <span class="st">"strand"</span>)
)
</pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">SubjectID</th>
<th align="right">Aggregated_value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">8884086</td>
<td align="left">-</td>
<td align="left">VA2020-mix11</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">11322005</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">11772847</td>
<td align="left">+</td>
<td align="left">VA2020-mix11</td>
<td align="right">13</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">22453532</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<ol start="4" style="list-style-type: decimal">
<li>
<strong>Changing <code>namespace</code> and <code>env</code></strong><br>
As stated before, you can specify several different functions, for them to be exported by the base package is not mandatory. If the function you’re interested in applying is exported by an installed package, you should specify the name of the package in the <code>namespace</code> parameter.<br>
It is also possible to apply a fully custom function (provided it respects the restrictions stated before): this means you can write your own function in some environment (most commonly the global environment) and pass it in the function as <code>lambda</code>. In this case you have to specify the environment in which the function is defined and set <code>namespace</code> as NULL.</li>
</ol>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">foo</span> <span class="op">&lt;-</span> <span class="fu">function</span>(<span class="kw">x</span>) {
    <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw">x</span>)
}
<span class="kw">env</span> <span class="op">&lt;-</span> <span class="kw">rlang</span>::<span class="fu"><a href="https://rlang.r-lib.org/reference/caller_env.html">current_env</a></span>()

<span class="kw">agg4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(
    x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>,
    association_file = <span class="kw">association_file</span>,
    key = <span class="st">"SubjectID"</span>,
    lambda = <span class="st">"foo"</span>,
    args = <span class="kw">NULL</span>,
    namespace = <span class="kw">NULL</span>,
    env = <span class="kw">env</span>
)
</pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">SubjectID</th>
<th align="right">Aggregated_value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">8884086</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix11</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">11322005</td>
<td align="left">-</td>
<td align="left">MTOR</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">11772847</td>
<td align="left">+</td>
<td align="left">DRAXIN</td>
<td align="left">+</td>
<td align="left">VA2020-mix11</td>
<td align="right">52</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">22453532</td>
<td align="left">+</td>
<td align="left">WNT4</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">8</td>
</tr>
</tbody>
</table>
<p>NOTE: when specifying <code>env</code>, note that also all other symbols (including those passed in <code>args</code>) will be evaluated in <code>env</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">foo</span> <span class="op">&lt;-</span> <span class="fu">function</span>(<span class="kw">x</span>, <span class="kw">y</span>) {
    <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw">x</span>, <span class="kw">y</span>)
}
<span class="kw">c</span> <span class="op">&lt;-</span> <span class="fl">2</span>
<span class="kw">env</span> <span class="op">&lt;-</span> <span class="kw">rlang</span>::<span class="fu"><a href="https://rlang.r-lib.org/reference/caller_env.html">current_env</a></span>()

<span class="kw">agg4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(
    x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>,
    association_file = <span class="kw">association_file</span>,
    key = <span class="st">"SubjectID"</span>,
    lambda = <span class="st">"foo"</span>,
    args = <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(y = <span class="kw">c</span>),
    namespace = <span class="kw">NULL</span>,
    env = <span class="kw">env</span>
)
</pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">SubjectID</th>
<th align="right">Aggregated_value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="right">10</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">8884086</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix11</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">11322005</td>
<td align="left">-</td>
<td align="left">MTOR</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">11772847</td>
<td align="left">+</td>
<td align="left">DRAXIN</td>
<td align="left">+</td>
<td align="left">VA2020-mix11</td>
<td align="right">54</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">22453532</td>
<td align="left">+</td>
<td align="left">WNT4</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">10</td>
</tr>
</tbody>
</table>
</div>
<div id="using-the-function-to-obtain-summaries-for-each-group-based-on-values" class="section level3">
<h3 class="hasAnchor">
<a href="#using-the-function-to-obtain-summaries-for-each-group-based-on-values" class="anchor"></a>Using the function to obtain summaries for each group based on values</h3>
<p>As we’ve seen in the previous section, we can pass any lambda function to <code>aggregate_values_by_key</code>, provided it takes as input a vector of integer or numeric values and produces a single value as output. This is true also for other objects (S3/S4 etc): you can use this function for example to produce summaries and include them in the tibble. A function we recommend is <code><a href="https://rdrr.io/pkg/psych/man/describe.html">psych::describe</a></code>.</p>
<pre><code>#&gt; Loading required package: psych</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">agg5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span>(
    x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>,
    association_file = <span class="kw">association_file</span>,
    key = <span class="st">"SubjectID"</span>,
    lambda = <span class="st">"describe"</span>,
    args = <span class="kw">NULL</span>,
    namespace = <span class="st">"psych"</span>,
    env = <span class="kw">env</span>
)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">agg5</span>)
<span class="co">#&gt; # A tibble: 6 x 7</span>
<span class="co">#&gt;   chr   integration_locus strand GeneName GeneStrand SubjectID  Aggregated_value</span>
<span class="co">#&gt;   &lt;chr&gt;             &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;      &lt;list&gt;          </span>
<span class="co">#&gt; 1 1               4519734 +      LINC017… +          VA2020-mi… &lt;tibble [1 × 13…</span>
<span class="co">#&gt; 2 1               8426336 -      RERE     -          VA2020-mi… &lt;tibble [1 × 13…</span>
<span class="co">#&gt; 3 1               8884086 -      RERE     -          VA2020-mi… &lt;tibble [1 × 13…</span>
<span class="co">#&gt; 4 1              11322005 -      MTOR     -          VA2020-mi… &lt;tibble [1 × 13…</span>
<span class="co">#&gt; 5 1              11772847 +      DRAXIN   +          VA2020-mi… &lt;tibble [1 × 13…</span>
<span class="co">#&gt; 6 1              22453532 +      WNT4     -          VA2020-mi… &lt;tibble [1 × 13…</span>
</pre></div>
<p>To see all of the contents of the lists, use <code><a href="https://tidyr.tidyverse.org/reference/nest.html">tidyr::unnest</a></code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="kw">agg5</span> <span class="op">&lt;-</span> <span class="kw">agg5</span> <span class="op">%&gt;%</span> <span class="kw">tidyr</span>::<span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span>(<span class="kw">.data</span><span class="op">$</span><span class="kw">Aggregated_value</span>)
</pre></div>
<table class="table">
<colgroup>
<col width="3%">
<col width="13%">
<col width="5%">
<col width="7%">
<col width="8%">
<col width="10%">
<col width="3%">
<col width="2%">
<col width="3%">
<col width="2%">
<col width="5%">
<col width="6%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="4%">
<col width="3%">
<col width="6%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">SubjectID</th>
<th align="right">vars</th>
<th align="right">n</th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">median</th>
<th align="right">trimmed</th>
<th align="right">mad</th>
<th align="right">min</th>
<th align="right">max</th>
<th align="right">range</th>
<th align="right">skew</th>
<th align="right">kurtosis</th>
<th align="right">se</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">4519734</td>
<td align="left">+</td>
<td align="left">LINC01777</td>
<td align="left">+</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8426336</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">8884086</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">VA2020-mix11</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">11322005</td>
<td align="left">-</td>
<td align="left">MTOR</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">11772847</td>
<td align="left">+</td>
<td align="left">DRAXIN</td>
<td align="left">+</td>
<td align="left">VA2020-mix11</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">13</td>
<td align="right">0</td>
<td align="right">13</td>
<td align="right">13</td>
<td align="right">0</td>
<td align="right">13</td>
<td align="right">13</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">22453532</td>
<td align="left">+</td>
<td align="left">WNT4</td>
<td align="left">-</td>
<td align="left">VA2020-mix10</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">NaN</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>The <em><a href="https://bioconductor.org/packages/3.12/ISAnalytics">ISAnalytics</a></em> package <a id="cite-calabrialab_2020"></a>(<a href="http://www.bioconductor.org/packages/ISAnalytics">calabrialab, 2020</a>) was made possible thanks to:</p>
<ul>
<li>R <a id="cite-R_Core_Team_2020"></a>(<a href="https://www.R-project.org/">R Core Team, 2020</a>)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> <a id="cite-Oles_2020"></a>(<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> <a id="cite-Boettiger_2019"></a>(<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> <a id="cite-Xie_2020"></a>(<a href="https://yihui.org/knitr/">Xie, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> <a id="cite-Allaire_2020"></a>(<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=sessioninfo">sessioninfo</a></em> <a id="cite-Csardi_2018"></a>(<a href="https://github.com/r-lib/sessioninfo#readme">Csárdi, core, Wickham, Chang, et al., 2018</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=testthat">testthat</a></em> <a id="cite-Wickham_2011"></a>(<a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">Wickham, 2011</a>)</li>
</ul>
<p>This package was developed using <em><a href="https://github.com/lcolladotor/biocthis">biocthis</a></em>.</p>
<p><code>R</code> session information.</p>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  setting  value                                      
#&gt;  version  R version 4.0.2 Patched (2020-09-17 r79226)
#&gt;  os       Ubuntu 20.04.1 LTS                         
#&gt;  system   x86_64, linux-gnu                          
#&gt;  ui       X11                                        
#&gt;  language (EN)                                       
#&gt;  collate  en_US.UTF-8                                
#&gt;  ctype    en_US.UTF-8                                
#&gt;  tz       UTC                                        
#&gt;  date     2020-09-21                                 
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  package       * version    date       lib source                        
#&gt;  assertthat      0.2.1      2019-03-21 [2] RSPM (R 4.0.0)                
#&gt;  backports       1.1.10     2020-09-15 [2] RSPM (R 4.0.2)                
#&gt;  bibtex          0.4.2.2    2020-01-02 [1] RSPM (R 4.0.0)                
#&gt;  BiocManager     1.30.10    2019-11-16 [2] CRAN (R 4.0.2)                
#&gt;  BiocParallel    1.23.2     2020-07-06 [1] Bioconductor                  
#&gt;  BiocStyle     * 2.17.0     2020-04-27 [1] Bioconductor                  
#&gt;  bookdown        0.20       2020-06-23 [1] RSPM (R 4.0.2)                
#&gt;  cli             2.0.2      2020-02-28 [2] RSPM (R 4.0.0)                
#&gt;  cpp11           0.2.1      2020-08-11 [1] RSPM (R 4.0.2)                
#&gt;  crayon          1.3.4      2017-09-16 [2] RSPM (R 4.0.0)                
#&gt;  desc            1.2.0      2018-05-01 [2] RSPM (R 4.0.0)                
#&gt;  digest          0.6.25     2020-02-23 [2] RSPM (R 4.0.0)                
#&gt;  dplyr           1.0.2      2020-08-18 [1] RSPM (R 4.0.2)                
#&gt;  ellipsis        0.3.1      2020-05-15 [2] RSPM (R 4.0.0)                
#&gt;  evaluate        0.14       2019-05-28 [2] RSPM (R 4.0.0)                
#&gt;  fansi           0.4.1      2020-01-08 [2] RSPM (R 4.0.0)                
#&gt;  forcats         0.5.0      2020-03-01 [1] RSPM (R 4.0.0)                
#&gt;  fs              1.5.0      2020-07-31 [2] RSPM (R 4.0.2)                
#&gt;  generics        0.0.2      2018-11-29 [1] RSPM (R 4.0.0)                
#&gt;  glue            1.4.2      2020-08-27 [2] RSPM (R 4.0.2)                
#&gt;  highr           0.8        2019-03-20 [2] RSPM (R 4.0.0)                
#&gt;  hms             0.5.3      2020-01-08 [1] RSPM (R 4.0.0)                
#&gt;  htmltools       0.5.0      2020-06-16 [2] RSPM (R 4.0.1)                
#&gt;  htmlwidgets     1.5.1      2019-10-08 [2] RSPM (R 4.0.0)                
#&gt;  httr            1.4.2      2020-07-20 [2] RSPM (R 4.0.2)                
#&gt;  ISAnalytics   * 0.99.11    2020-09-21 [1] Bioconductor                  
#&gt;  jsonlite        1.7.1      2020-09-07 [2] RSPM (R 4.0.2)                
#&gt;  knitcitations * 1.0.10     2019-09-15 [1] RSPM (R 4.0.0)                
#&gt;  knitr           1.29       2020-06-23 [2] RSPM (R 4.0.2)                
#&gt;  lattice         0.20-41    2020-04-02 [3] CRAN (R 4.0.2)                
#&gt;  lifecycle       0.2.0      2020-03-06 [2] RSPM (R 4.0.0)                
#&gt;  lubridate       1.7.9      2020-06-08 [1] RSPM (R 4.0.2)                
#&gt;  magrittr      * 1.5        2014-11-22 [2] RSPM (R 4.0.0)                
#&gt;  memoise         1.1.0      2017-04-21 [2] RSPM (R 4.0.0)                
#&gt;  mnormt          2.0.2      2020-09-01 [1] RSPM (R 4.0.2)                
#&gt;  nlme            3.1-149    2020-08-23 [3] CRAN (R 4.0.2)                
#&gt;  pillar          1.4.6      2020-07-10 [2] RSPM (R 4.0.2)                
#&gt;  pkgconfig       2.0.3      2019-09-22 [2] RSPM (R 4.0.0)                
#&gt;  pkgdown         1.6.1.9000 2020-09-21 [1] Github (r-lib/pkgdown@e1e46f2)
#&gt;  plyr            1.8.6      2020-03-03 [1] RSPM (R 4.0.2)                
#&gt;  psych         * 2.0.8      2020-09-04 [1] RSPM (R 4.0.2)                
#&gt;  purrr           0.3.4      2020-04-17 [2] RSPM (R 4.0.0)                
#&gt;  R6              2.4.1      2019-11-12 [2] RSPM (R 4.0.0)                
#&gt;  ragg            0.3.1      2020-07-03 [1] RSPM (R 4.0.2)                
#&gt;  Rcpp            1.0.5      2020-07-06 [2] RSPM (R 4.0.2)                
#&gt;  reactable       0.2.2      2020-09-13 [1] RSPM (R 4.0.2)                
#&gt;  readr           1.3.1      2018-12-21 [1] RSPM (R 4.0.2)                
#&gt;  RefManageR      1.2.12     2019-04-03 [1] RSPM (R 4.0.0)                
#&gt;  rlang           0.4.7      2020-07-09 [2] RSPM (R 4.0.2)                
#&gt;  rmarkdown       2.3        2020-06-18 [1] RSPM (R 4.0.1)                
#&gt;  rprojroot       1.3-2      2018-01-03 [2] RSPM (R 4.0.0)                
#&gt;  sessioninfo   * 1.1.1      2018-11-05 [2] RSPM (R 4.0.0)                
#&gt;  stringi         1.5.3      2020-09-09 [2] RSPM (R 4.0.2)                
#&gt;  stringr         1.4.0      2019-02-10 [2] RSPM (R 4.0.0)                
#&gt;  systemfonts     0.3.1      2020-09-08 [1] RSPM (R 4.0.2)                
#&gt;  tibble          3.0.3      2020-07-10 [2] RSPM (R 4.0.2)                
#&gt;  tidyr           1.1.2      2020-08-27 [1] RSPM (R 4.0.2)                
#&gt;  tidyselect      1.1.0      2020-05-11 [1] RSPM (R 4.0.0)                
#&gt;  tmvnsim         1.0-2      2016-12-15 [1] RSPM (R 4.0.0)                
#&gt;  utf8            1.1.4      2018-05-24 [2] RSPM (R 4.0.0)                
#&gt;  vctrs           0.3.4      2020-08-29 [2] RSPM (R 4.0.2)                
#&gt;  withr           2.2.0      2020-04-20 [2] RSPM (R 4.0.0)                
#&gt;  xfun            0.17       2020-09-09 [2] RSPM (R 4.0.2)                
#&gt;  xml2            1.3.2      2020-04-23 [2] RSPM (R 4.0.0)                
#&gt;  yaml            2.2.1      2020-02-01 [2] RSPM (R 4.0.0)                
#&gt;  zip             2.1.1      2020-08-27 [1] RSPM (R 4.0.2)                
#&gt; 
#&gt; [1] /usr/local/lib/R/host-site-library
#&gt; [2] /usr/local/lib/R/site-library
#&gt; [3] /usr/local/lib/R/library</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> (<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (<a href="https://yihui.org/knitr/">Xie, 2020</a>) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> (<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> (<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>).</p>
<p>[1] J. Allaire, Y. Xie, J. McPherson, J. Luraschi, et al. <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.3. 2020. &lt;URL: <a href="https://github.com/rstudio/rmarkdown" class="uri">https://github.com/rstudio/rmarkdown</a>&gt;.</p>
<p>[2] C. Boettiger. <em>knitcitations: Citations for ‘Knitr’ Markdown Files</em>. R package version 1.0.10. 2019. &lt;URL: <a href="https://github.com/cboettig/knitcitations" class="uri">https://github.com/cboettig/knitcitations</a>&gt;.</p>
<p>[3] calabrialab. <em>Analyze gene therapy vector insertion sites data identified from genomics next generation sequencing reads for clonal tracking studies</em>. <a href="https://github.com/calabrialab/ISAnalytics" class="uri">https://github.com/calabrialab/ISAnalytics</a> - R package version 0.99.11. 2020. DOI: 10.18129/B9.bioc.ISAnalytics. &lt;URL: <a href="http://www.bioconductor.org/packages/ISAnalytics" class="uri">http://www.bioconductor.org/packages/ISAnalytics</a>&gt;.</p>
<p>[4] G. Csárdi, R. core, H. Wickham, W. Chang, et al. <em>sessioninfo: R Session Information</em>. R package version 1.1.1. 2018. &lt;URL: <a href="https://github.com/r-lib/sessioninfo#readme" class="uri">https://github.com/r-lib/sessioninfo#readme</a>&gt;.</p>
<p>[5] A. Oleś, M. Morgan, and W. Huber. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.17.0. 2020. &lt;URL: <a href="https://github.com/Bioconductor/BiocStyle" class="uri">https://github.com/Bioconductor/BiocStyle</a>&gt;.</p>
<p>[6] R Core Team. <em>R: A Language and Environment for Statistical Computing</em>. R Foundation for Statistical Computing. Vienna, Austria, 2020. &lt;URL: <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>&gt;.</p>
<p>[7] H. Wickham. “testthat: Get Started with Testing”. In: <em>The R Journal</em> 3 (2011), pp. 5-10. &lt;URL: <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf" class="uri">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>&gt;.</p>
<p>[8] Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.29. 2020. &lt;URL: <a href="https://yihui.org/knitr/" class="uri">https://yihui.org/knitr/</a>&gt;.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
