<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Collision removal functionality • ISAnalytics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Collision removal functionality">
<meta property="og:description" content="ISAnalytics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISAnalytics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/aggregate_function_usage.html">Working with aggregate functions</a>
    </li>
    <li>
      <a href="../articles/collision_removal.html">Collision removal functionality</a>
    </li>
    <li>
      <a href="../articles/how_to_import_functions.html">How to use import functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="collision_removal_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Collision removal functionality</h1>
                        <h4 class="author">Giulia Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget, Via Olgettina 60, 20132 Milano - Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 class="date">14 September 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/master/vignettes/collision_removal.Rmd"><code>vignettes/collision_removal.Rmd</code></a></small>
      <div class="hidden name"><code>collision_removal.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<div id="how-to-install-isanalytics" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-install-isanalytics" class="anchor"></a>How to install ISAnalytics</h2>
<p>To install the package run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, quietly = <span class="fl">TRUE</span>)) {
      <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
  }
<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"ISAnalytics"</span>)
</pre></div>
<p>To install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="kw"><a href="https://devtools.r-lib.org">devtools</a></span>)) {
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
}
<span class="kw">devtools</span>::<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"calabrialab/ISAnalytics"</span>, build_vignettes = <span class="fl">TRUE</span>)
</pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com//calabrialab/isanalytics">ISAnalytics</a></span>)
<span class="co">#&gt; Loading required package: magrittr</span>
</pre></div>
</div>
<div id="setting-options" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-options" class="anchor"></a>Setting options</h2>
<p><code>ISAnalytics</code> has a verbose option that allows some functions to print additional information to the console while they’re executing. To disable this feature do:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># DISABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.verbose"</span> = <span class="fl">FALSE</span>)

<span class="co"># ENABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.verbose"</span> = <span class="fl">TRUE</span>)
</pre></div>
<p>Some functions also produce report in a user-friendly HTML format, to set this feature:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># DISABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.widgets"</span> = <span class="fl">FALSE</span>)

<span class="co"># ENABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.widgets"</span> = <span class="fl">TRUE</span>)
</pre></div>
</div>
<div id="what-is-a-collision-and-why-should-you-care" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-a-collision-and-why-should-you-care" class="anchor"></a>What is a collision and why should you care?</h2>
<p>We’re not going into too much detail here, but we’re going to explain in a very simple way what is a “collision” and how the function in this package deal with them.</p>
<p>We say that an integration (aka a unique combination of chromosome, integration locus and strand) is a <em>collision</em> if this combination is shared between different independent samples: an independent sample is a unique combination of <code>ProjectID</code> and <code>SubjectID</code> (where subjects usually represent patients). The reason behind this is that it’s highly improbable to observe the very same integration in two different subjects and this phenomenon might be an indicator of some kind of contamination in the sequencing phase or in PCR phase, for this reason we might want to exclude such contamination from our analysis.<br><code>ISAnalytics</code> provides a function that processes the imported data for the removal or reassignment of these “problematic” integrations, <code>remove_collisions</code>.</p>
<p>The processing is done on the sequence count matrix (after import) and matrices of other quantification types are re-aligned accordingly.</p>
</div>
<div id="the-logic-behind-the-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-logic-behind-the-function" class="anchor"></a>The logic behind the function</h2>
<p>The <code>remove_collisions</code> function follows several logical steps to decide whether an integration is a collision and if it is it decides whether to re-assign it or remove it entirely based on different criterias.</p>
<div id="identifying-the-collisions" class="section level3">
<h3 class="hasAnchor">
<a href="#identifying-the-collisions" class="anchor"></a>Identifying the collisions</h3>
<p>As we said before, a collision is a triplet made of <code>chr</code>, <code>integration locus</code> and <code>strand</code>, which is shared between different independent samples, aka a pair made of <code>ProjectID</code> and <code>SubjectID</code>. The function uses the information stored in the association file to assess which independent samples are present and counts the number of independent samples for each integration: those who have a count &gt; 1 are considered collisions.</p>
</div>
<div id="re-assign-vs-remove" class="section level3">
<h3 class="hasAnchor">
<a href="#re-assign-vs-remove" class="anchor"></a>Re-assign vs remove</h3>
<p>Once the collisions are identified, the function follows 3 steps where it tries to re-assign the combination to a single independent sample. The criterias are:</p>
<ol style="list-style-type: decimal">
<li>Compare dates: if it’s possible to have an absolute ordering on dates, the integration is re-assigned to the sample that has the earliest date. If two samples share the same date it’s impossible to decide, so the next criteria is tested</li>
<li>Compare replicate number: if a sample has the same integration in more than one replicate, it’s more probable the integration is not an artifact. If it’s possible to have an absolute ordering, the collision is re-assigned to the sample whose grouping is largest</li>
<li>Compare the sequence count value: if the previous criteria wasn’t sufficient to make a decision, for each group of independent samples it’s evaluated the sum of the sequence count value - for each group there is a cumulative value of the sequence count and this is compared to the value of other groups. If there is a single group which has a ratio n times bigger than other groups, this one is chosen for re-assignment. The factor n is passed as a parameter in the function ( <code>reads_ratio</code>), the default value is 10.</li>
</ol>
<p>If none of the criterias were sufficient to make a decision, the integration is simply removed from the matrix.</p>
</div>
</div>
</div>
<div id="typical-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#typical-workflow" class="anchor"></a>Typical workflow</h1>
<p>To know more about import functions take a look at the vignette “How to use import functions”.</p>
<div id="import-the-association-file" class="section level2">
<h2 class="hasAnchor">
<a href="#import-the-association-file" class="anchor"></a>Import the association file</h2>
<p>Import your association file:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(ISAnalytics.widgets = <span class="fl">FALSE</span>)
<span class="kw">path_AF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ex_association_file.tsv"</span>,
    package = <span class="st">"ISAnalytics"</span>
)

<span class="kw">root_correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"fs.zip"</span>,
    package = <span class="st">"ISAnalytics"</span>
)
<span class="kw">root_correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unzip_file_system.html">unzip_file_system</a></span>(<span class="kw">root_correct</span>, <span class="st">"fs"</span>)
<span class="kw">association_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_association_file.html">import_association_file</a></span>(<span class="kw">path_AF</span>, <span class="kw">root_correct</span>)
</pre></div>
<p>Important notes on the association file:</p>
<ul>
<li>You have to be sure your association file is properly filled out. The function requires you to specify a date column (by default “SequencingDate”), you have to ensure this column doesn’t contain NA values or incorrect values.</li>
<li>You have to ensure that your association file contains ALL the information regarding CompleteAmplificationIDs present in the matrices you’re analyzing - an error is thrown otherwise</li>
<li>If you have the verbose option set to true and your association file holds additional information on other samples which for any reason are not present in the matrix you’re analyzing, you’ll be notified with a console message.</li>
</ul>
</div>
<div id="import-the-matrices-for-your-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#import-the-matrices-for-your-analysis" class="anchor"></a>Import the matrices for your analysis</h2>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co"># This imports both sequence count and fragment estimate matrices</span>
<span class="kw">matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices_auto.html">import_parallel_Vispa2Matrices_auto</a></span>(
    association_file = <span class="kw">association_file</span>, root = <span class="kw">NULL</span>,
    quantification_type = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"fragmentEstimate"</span>, <span class="st">"seqCount"</span>),
    matrix_type = <span class="st">"annotated"</span>, workers = <span class="fl">2</span>, patterns = <span class="kw">NULL</span>,
    matching_opt = <span class="st">"ANY"</span>
)
</pre></div>
<p>As stated in the introduction, it is fundamental that the sequence count matrix is present for the collision removal process to take place.</p>
</div>
<div id="process-the-collisions" class="section level2">
<h2 class="hasAnchor">
<a href="#process-the-collisions" class="anchor"></a>Process the collisions</h2>
<p>You can process the collisions in 3 different ways.</p>
<div id="pass-the-entire-named-list-to-the-function" class="section level3">
<h3 class="hasAnchor">
<a href="#pass-the-entire-named-list-to-the-function" class="anchor"></a>Pass the entire named list to the function</h3>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co"># Pass the whole named list</span>
<span class="kw">matrices_processed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_collisions.html">remove_collisions</a></span>(x = <span class="kw">matrices</span>, 
                                        association_file = <span class="kw">association_file</span>, 
                                        date_col = <span class="st">"SequencingDate"</span>, 
                                        reads_ratio = <span class="fl">10</span>)
<span class="co">#&gt; Found additional data relative to some projects that are not included in the imported matrices. Here is a summary</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   ProjectID   PoolID       CompleteAmplificationID                              </span>
<span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;                                                </span>
<span class="co">#&gt; 1 PROJECT1101 ABY-LR-PL4-… PROJECT1101_ABY-LR-PL4-POOL54_LTR51LC51_VA2020-mix14…</span>
<span class="co">#&gt; Identifying collisions...</span>
<span class="co">#&gt; Processing collisions...</span>
<span class="co">#&gt; Realigning matrices...</span>
<span class="co">#&gt; Finished!</span>
</pre></div>
<p>If you have the “widgets” option active, a report file is produced at the end that shows the before and after for each subject (and some other details). This report is an HTML widget, so you can save it or export it for future reference if you need it.<br>
In this case, collision removal is done on the sequence count matrix and other matrices are re-aligned automatically.</p>
</div>
<div id="give-only-the-sequence-count-matrix-as-input" class="section level3">
<h3 class="hasAnchor">
<a href="#give-only-the-sequence-count-matrix-as-input" class="anchor"></a>Give only the sequence count matrix as input</h3>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co"># Pass the sequence count matrix only</span>
<span class="kw">matrices_processed_single</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_collisions.html">remove_collisions</a></span>(x = <span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>, 
                                               association_file = 
                                                 <span class="kw">association_file</span>, 
                                               date_col = <span class="st">"SequencingDate"</span>, 
                                               reads_ratio = <span class="fl">10</span>)
<span class="co">#&gt; Found additional data relative to some projects that are not included in the imported matrices. Here is a summary</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   ProjectID   PoolID       CompleteAmplificationID                              </span>
<span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;                                                </span>
<span class="co">#&gt; 1 PROJECT1101 ABY-LR-PL4-… PROJECT1101_ABY-LR-PL4-POOL54_LTR51LC51_VA2020-mix14…</span>
<span class="co">#&gt; Identifying collisions...</span>
<span class="co">#&gt; Processing collisions...</span>
<span class="co">#&gt; Finished! You provided a single sequence count as matrix, to re-align other related matrices see ?realign_after_collisions</span>
</pre></div>
<p>If you have the “verbose” option active, a console message will remind you to align other matrices if you have them at a later time.</p>
</div>
<div id="support-for-multi-quantification-matrices" class="section level3">
<h3 class="hasAnchor">
<a href="#support-for-multi-quantification-matrices" class="anchor"></a>Support for multi-quantification matrices</h3>
<p>If you’d like to avoid the re-alignment phase, you can call collision removal on a multi-quantification matrix obtained via the function <code>comparison_matrix</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co"># Obtain multi-quantification matrix</span>
<span class="kw">multi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comparison_matrix.html">comparison_matrix</a></span>(<span class="kw">matrices</span>)
<span class="kw">multi</span>
<span class="co">#&gt; # A tibble: 5,904 x 8</span>
<span class="co">#&gt;    chr   integration_loc… strand GeneName GeneStrand CompleteAmplifi…</span>
<span class="co">#&gt;    &lt;chr&gt;            &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;           </span>
<span class="co">#&gt;  1 1            100903226 +      CDC14A   +          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  2 1            101724095 -      S1PR1    +          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  3 1            108208779 +      VAV3     -          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  4 1            108412817 -      VAV3     -          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  5 1            108412817 -      VAV3     -          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  6 1            108511440 +      VAV3-AS1 +          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  7 1            109689050 +      KIAA1324 +          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  8 1            111411778 +      CD53     +          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt;  9 1            111411778 +      CD53     +          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt; 10 1            111466980 -      LRIF1    -          CLOEXP_POOL6_LT…</span>
<span class="co">#&gt; # … with 5,894 more rows, and 2 more variables: fragmentEstimate &lt;dbl&gt;,</span>
<span class="co">#&gt; #   seqCount &lt;int&gt;</span>
<span class="kw">matrices_processed_multi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_collisions.html">remove_collisions</a></span>(x = <span class="kw">multi</span>,
                                              association_file = 
                                                 <span class="kw">association_file</span>, 
                                               date_col = <span class="st">"SequencingDate"</span>, 
                                               reads_ratio = <span class="fl">10</span>, 
                                              seq_count_col = <span class="st">"seqCount"</span>)
<span class="co">#&gt; Found additional data relative to some projects that are not included in the imported matrices. Here is a summary</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   ProjectID   PoolID       CompleteAmplificationID                              </span>
<span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;                                                </span>
<span class="co">#&gt; 1 PROJECT1101 ABY-LR-PL4-… PROJECT1101_ABY-LR-PL4-POOL54_LTR51LC51_VA2020-mix14…</span>
<span class="co">#&gt; Identifying collisions...</span>
<span class="co">#&gt; Processing collisions...</span>
<span class="co">#&gt; Finished!</span>
</pre></div>
<p>As you can see, <code>comparison_matrix</code> produces a single integration matrix from the named list of single quantification matrices. This is the recommended approach if you don’t have specific needs as it negates the necessity of realigning matrices altogether.</p>
</div>
</div>
<div id="re-align-other-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#re-align-other-matrices" class="anchor"></a>Re-align other matrices</h2>
<p>If you have opted for the second way, to realign other matrices you have to call the function <code>realign_after_collisions</code>, passing as input the processed sequence count matrix and the named list of other matrices to realign.<br><strong>NOTE: the names in the list must be quantification types.</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">seq_count_proc</span> <span class="op">&lt;-</span> <span class="kw">matrices_processed_single</span>
<span class="kw">other_matrices</span> <span class="op">&lt;-</span> <span class="kw">matrices</span>[<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">matrices</span>) <span class="op">%in%</span> <span class="st">"seqCount"</span>] 
<span class="co"># Select only matrices that are not relative to sequence count</span>
<span class="kw">other_realigned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/realign_after_collisions.html">realign_after_collisions</a></span>(<span class="kw">seq_count_proc</span>, <span class="kw">other_matrices</span>)
</pre></div>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>The <em><a href="https://bioconductor.org/packages/3.12/ISAnalytics">ISAnalytics</a></em> package <a id="cite-calabrialab_2020"></a>(<a href="http://www.bioconductor.org/packages/ISAnalytics">calabrialab, 2020</a>) was made possible thanks to:</p>
<ul>
<li>R <a id="cite-R_Core_Team_2020"></a>(<a href="https://www.R-project.org/">R Core Team, 2020</a>)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> <a id="cite-Oles_2020"></a>(<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> <a id="cite-Boettiger_2019"></a>(<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> <a id="cite-Xie_2020"></a>(<a href="https://yihui.org/knitr/">Xie, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> <a id="cite-Allaire_2020"></a>(<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=sessioninfo">sessioninfo</a></em> <a id="cite-Csardi_2018"></a>(<a href="https://github.com/r-lib/sessioninfo#readme">Csárdi, core, Wickham, Chang, et al., 2018</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=testthat">testthat</a></em> <a id="cite-Wickham_2011"></a>(<a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">Wickham, 2011</a>)</li>
</ul>
<p>This package was developed using <em><a href="https://github.com/lcolladotor/biocthis">biocthis</a></em>.</p>
<p><code>R</code> session information.</p>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  setting  value                                      
#&gt;  version  R version 4.0.2 Patched (2020-09-10 r79182)
#&gt;  os       Ubuntu 20.04.1 LTS                         
#&gt;  system   x86_64, linux-gnu                          
#&gt;  ui       X11                                        
#&gt;  language (EN)                                       
#&gt;  collate  en_US.UTF-8                                
#&gt;  ctype    en_US.UTF-8                                
#&gt;  tz       UTC                                        
#&gt;  date     2020-09-14                                 
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  package       * version    date       lib source                        
#&gt;  assertthat      0.2.1      2019-03-21 [2] RSPM (R 4.0.0)                
#&gt;  backports       1.1.9      2020-08-24 [2] RSPM (R 4.0.2)                
#&gt;  bibtex          0.4.2.2    2020-01-02 [1] RSPM (R 4.0.0)                
#&gt;  BiocManager     1.30.10    2019-11-16 [2] CRAN (R 4.0.2)                
#&gt;  BiocParallel    1.23.2     2020-07-06 [1] Bioconductor                  
#&gt;  BiocStyle     * 2.17.0     2020-04-27 [1] Bioconductor                  
#&gt;  bookdown        0.20       2020-06-23 [1] RSPM (R 4.0.2)                
#&gt;  cli             2.0.2      2020-02-28 [2] RSPM (R 4.0.0)                
#&gt;  cpp11           0.2.1      2020-08-11 [1] RSPM (R 4.0.2)                
#&gt;  crayon          1.3.4      2017-09-16 [2] RSPM (R 4.0.0)                
#&gt;  desc            1.2.0      2018-05-01 [2] RSPM (R 4.0.0)                
#&gt;  digest          0.6.25     2020-02-23 [2] RSPM (R 4.0.0)                
#&gt;  dplyr           1.0.2      2020-08-18 [1] RSPM (R 4.0.2)                
#&gt;  ellipsis        0.3.1      2020-05-15 [2] RSPM (R 4.0.0)                
#&gt;  evaluate        0.14       2019-05-28 [2] RSPM (R 4.0.0)                
#&gt;  fansi           0.4.1      2020-01-08 [2] RSPM (R 4.0.0)                
#&gt;  forcats         0.5.0      2020-03-01 [1] RSPM (R 4.0.0)                
#&gt;  fs              1.5.0      2020-07-31 [2] RSPM (R 4.0.2)                
#&gt;  generics        0.0.2      2018-11-29 [1] RSPM (R 4.0.0)                
#&gt;  glue            1.4.2      2020-08-27 [2] RSPM (R 4.0.2)                
#&gt;  hms             0.5.3      2020-01-08 [1] RSPM (R 4.0.0)                
#&gt;  htmltools       0.5.0      2020-06-16 [2] RSPM (R 4.0.1)                
#&gt;  htmlwidgets     1.5.1      2019-10-08 [2] RSPM (R 4.0.0)                
#&gt;  httr            1.4.2      2020-07-20 [2] RSPM (R 4.0.2)                
#&gt;  ISAnalytics   * 0.99.10    2020-09-14 [1] Bioconductor                  
#&gt;  jsonlite        1.7.1      2020-09-07 [2] RSPM (R 4.0.2)                
#&gt;  knitcitations * 1.0.10     2019-09-15 [1] RSPM (R 4.0.0)                
#&gt;  knitr           1.29       2020-06-23 [2] RSPM (R 4.0.2)                
#&gt;  lifecycle       0.2.0      2020-03-06 [2] RSPM (R 4.0.0)                
#&gt;  lubridate       1.7.9      2020-06-08 [1] RSPM (R 4.0.2)                
#&gt;  magrittr      * 1.5        2014-11-22 [2] RSPM (R 4.0.0)                
#&gt;  memoise         1.1.0      2017-04-21 [2] RSPM (R 4.0.0)                
#&gt;  pillar          1.4.6      2020-07-10 [2] RSPM (R 4.0.2)                
#&gt;  pkgconfig       2.0.3      2019-09-22 [2] RSPM (R 4.0.0)                
#&gt;  pkgdown         1.6.1.9000 2020-09-14 [1] Github (r-lib/pkgdown@cc9c7b1)
#&gt;  plyr            1.8.6      2020-03-03 [1] RSPM (R 4.0.2)                
#&gt;  purrr           0.3.4      2020-04-17 [2] RSPM (R 4.0.0)                
#&gt;  R6              2.4.1      2019-11-12 [2] RSPM (R 4.0.0)                
#&gt;  ragg            0.3.1      2020-07-03 [1] RSPM (R 4.0.2)                
#&gt;  Rcpp            1.0.5      2020-07-06 [2] RSPM (R 4.0.2)                
#&gt;  reactable       0.2.1      2020-09-01 [1] RSPM (R 4.0.2)                
#&gt;  readr           1.3.1      2018-12-21 [1] RSPM (R 4.0.2)                
#&gt;  RefManageR      1.2.12     2019-04-03 [1] RSPM (R 4.0.0)                
#&gt;  rlang           0.4.7      2020-07-09 [2] RSPM (R 4.0.2)                
#&gt;  rmarkdown       2.3        2020-06-18 [1] RSPM (R 4.0.1)                
#&gt;  rprojroot       1.3-2      2018-01-03 [2] RSPM (R 4.0.0)                
#&gt;  sessioninfo   * 1.1.1      2018-11-05 [2] RSPM (R 4.0.0)                
#&gt;  stringi         1.5.3      2020-09-09 [2] RSPM (R 4.0.2)                
#&gt;  stringr         1.4.0      2019-02-10 [2] RSPM (R 4.0.0)                
#&gt;  systemfonts     0.3.1      2020-09-08 [1] RSPM (R 4.0.2)                
#&gt;  tibble          3.0.3      2020-07-10 [2] RSPM (R 4.0.2)                
#&gt;  tidyr           1.1.2      2020-08-27 [1] RSPM (R 4.0.2)                
#&gt;  tidyselect      1.1.0      2020-05-11 [1] RSPM (R 4.0.0)                
#&gt;  utf8            1.1.4      2018-05-24 [2] RSPM (R 4.0.0)                
#&gt;  vctrs           0.3.4      2020-08-29 [2] RSPM (R 4.0.2)                
#&gt;  withr           2.2.0      2020-04-20 [2] RSPM (R 4.0.0)                
#&gt;  xfun            0.17       2020-09-09 [2] RSPM (R 4.0.2)                
#&gt;  xml2            1.3.2      2020-04-23 [2] RSPM (R 4.0.0)                
#&gt;  yaml            2.2.1      2020-02-01 [2] RSPM (R 4.0.0)                
#&gt;  zip             2.1.1      2020-08-27 [1] RSPM (R 4.0.2)                
#&gt; 
#&gt; [1] /usr/local/lib/R/host-site-library
#&gt; [2] /usr/local/lib/R/site-library
#&gt; [3] /usr/local/lib/R/library</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> (<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (<a href="https://yihui.org/knitr/">Xie, 2020</a>) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> (<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> (<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>).</p>
<p>[1] J. Allaire, Y. Xie, J. McPherson, J. Luraschi, et al. <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.3. 2020. &lt;URL: <a href="https://github.com/rstudio/rmarkdown" class="uri">https://github.com/rstudio/rmarkdown</a>&gt;.</p>
<p>[2] C. Boettiger. <em>knitcitations: Citations for ‘Knitr’ Markdown Files</em>. R package version 1.0.10. 2019. &lt;URL: <a href="https://github.com/cboettig/knitcitations" class="uri">https://github.com/cboettig/knitcitations</a>&gt;.</p>
<p>[3] calabrialab. <em>Analyze gene therapy vector insertion sites data identified from genomics next generation sequencing reads for clonal tracking studies</em>. <a href="https://github.com/calabrialab/ISAnalytics" class="uri">https://github.com/calabrialab/ISAnalytics</a> - R package version 0.99.10. 2020. DOI: 10.18129/B9.bioc.ISAnalytics. &lt;URL: <a href="http://www.bioconductor.org/packages/ISAnalytics" class="uri">http://www.bioconductor.org/packages/ISAnalytics</a>&gt;.</p>
<p>[4] G. Csárdi, R. core, H. Wickham, W. Chang, et al. <em>sessioninfo: R Session Information</em>. R package version 1.1.1. 2018. &lt;URL: <a href="https://github.com/r-lib/sessioninfo#readme" class="uri">https://github.com/r-lib/sessioninfo#readme</a>&gt;.</p>
<p>[5] A. Oleś, M. Morgan, and W. Huber. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.17.0. 2020. &lt;URL: <a href="https://github.com/Bioconductor/BiocStyle" class="uri">https://github.com/Bioconductor/BiocStyle</a>&gt;.</p>
<p>[6] R Core Team. <em>R: A Language and Environment for Statistical Computing</em>. R Foundation for Statistical Computing. Vienna, Austria, 2020. &lt;URL: <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>&gt;.</p>
<p>[7] H. Wickham. “testthat: Get Started with Testing”. In: <em>The R Journal</em> 3 (2011), pp. 5-10. &lt;URL: <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf" class="uri">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>&gt;.</p>
<p>[8] Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.29. 2020. &lt;URL: <a href="https://yihui.org/knitr/" class="uri">https://yihui.org/knitr/</a>&gt;.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
