<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>How to use import functions • ISAnalytics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/slate/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="How to use import functions">
<meta property="og:description" content="ISAnalytics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISAnalytics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/aggregate_function_usage.html">Working with aggregate functions</a>
    </li>
    <li>
      <a href="../articles/collision_removal.html">Collision removal functionality</a>
    </li>
    <li>
      <a href="../articles/how_to_import_functions.html">How to use import functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="how_to_import_functions_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>How to use import functions</h1>
                        <h4 class="author">Giulia Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget, Via Olgettina 60, 20132 Milano - Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 class="date">21 September 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/master/vignettes/how_to_import_functions.Rmd"><code>vignettes/how_to_import_functions.Rmd</code></a></small>
      <div class="hidden name"><code>how_to_import_functions.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction-to-isanalytics-import-functions-family" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-isanalytics-import-functions-family" class="anchor"></a>Introduction to <code>ISAnalytics</code> import functions family</h1>
<p>In this vignette we’re going to explain more in detail how functions of the import family should be used, the most common workflows to follow and more.</p>
<div id="how-to-install-isanalytics" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-install-isanalytics" class="anchor"></a>How to install ISAnalytics</h2>
<p>To install the package run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co">if</span>(<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, quietly = <span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"ISAnalytics"</span>)
</pre></div>
<p>To install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="kw"><a href="https://devtools.r-lib.org">devtools</a></span>)) {
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
}
<span class="kw">devtools</span>::<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"calabrialab/ISAnalytics"</span>, build_vignettes = <span class="fl">TRUE</span>)
</pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com//calabrialab/isanalytics">ISAnalytics</a></span>)
<span class="co">#&gt; Loading required package: magrittr</span>
</pre></div>
</div>
<div id="setting-options" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-options" class="anchor"></a>Setting options</h2>
<p><code>ISAnalytics</code> has a verbose option that allows some functions to print additional information to the console while they’re executing. To disable this feature do:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># DISABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.verbose"</span> = <span class="fl">FALSE</span>)

<span class="co"># ENABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.verbose"</span> = <span class="fl">TRUE</span>)
</pre></div>
<p>Some functions also produce report in a user-friendly HTML format, to set this feature:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># DISABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.widgets"</span> = <span class="fl">FALSE</span>)

<span class="co"># ENABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="st">"ISAnalytics.widgets"</span> = <span class="fl">TRUE</span>)
</pre></div>
</div>
<div id="designed-to-work-with-vispa2-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#designed-to-work-with-vispa2-pipeline" class="anchor"></a>Designed to work with Vispa2 pipeline</h2>
<p>The vast majority of the functions included in this package is designed to work in combination with Vispa2 pipeline. If you don’t know what it is, we strongly recommend you to take a look at these links:</p>
<ul>
<li>Article: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5702242/">VISPA2: Article</a><br>
</li>
<li>BitBucket Wiki: <a href="https://bitbucket.org/andreacalabria/vispa2/wiki/Home">VISPA2 Wiki</a>
</li>
</ul>
</div>
<div id="file-system-structure-generated" class="section level2">
<h2 class="hasAnchor">
<a href="#file-system-structure-generated" class="anchor"></a>File system structure generated</h2>
<p>Vispa2 produces a standard file system structure starting from a folder you specify as your workbench or root. The structure always follows this schema:</p>
<ul>
<li>root/
<ul>
<li>Optional intermediate folders
<ul>
<li>Projects (PROJECTID)
<ul>
<li>bam</li>
<li>bcmuxall</li>
<li>bed</li>
<li>iss</li>
<li>quality</li>
<li>quantification
<ul>
<li>Pools (concatenatePoolIDSeqRun)</li>
</ul>
</li>
<li>report</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>We’ve included 2 examples of this structure in our package, one correct and the other one including errors or potential problems. They are both in .zip format, so you might want to unzip them if you plan to experiment with them.<br>
An example on how to access them:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">root_correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"fs.zip"</span>, package = <span class="st">"ISAnalytics"</span>)
<span class="kw">root_correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unzip_file_system.html">unzip_file_system</a></span>(<span class="kw">root_correct</span>, <span class="st">"fs"</span>)
<span class="kw">fs</span>::<span class="fu"><a href="http://fs.r-lib.org/reference/dir_tree.html">dir_tree</a></span>(<span class="kw">root_correct</span>)
<span class="co">#&gt; /tmp/RtmpzcGFHO/fs</span>
<span class="co">#&gt; ├── PROJECT1100</span>
<span class="co">#&gt; │   ├── bam</span>
<span class="co">#&gt; │   ├── bcmuxall</span>
<span class="co">#&gt; │   ├── bed</span>
<span class="co">#&gt; │   ├── iss</span>
<span class="co">#&gt; │   │   ├── stats.sequence_PROJECT1100.ABX-LR-PL5-POOL14-1.tsv</span>
<span class="co">#&gt; │   │   └── stats.sequence_PROJECT1100.ABX-LR-PL6-POOL15-1.tsv</span>
<span class="co">#&gt; │   ├── quality</span>
<span class="co">#&gt; │   ├── quantification</span>
<span class="co">#&gt; │   │   ├── ABX-LR-PL5-POOL14-1</span>
<span class="co">#&gt; │   │   │   ├── PROJECT1100_ABX-LR-PL5-POOL14-1_fragmentEstimate_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │   │   ├── PROJECT1100_ABX-LR-PL5-POOL14-1_fragmentEstimate_matrix.tsv.xz</span>
<span class="co">#&gt; │   │   │   ├── PROJECT1100_ABX-LR-PL5-POOL14-1_seqCount_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │   │   └── PROJECT1100_ABX-LR-PL5-POOL14-1_seqCount_matrix.tsv.xz</span>
<span class="co">#&gt; │   │   └── ABX-LR-PL6-POOL15-1</span>
<span class="co">#&gt; │   │       ├── PROJECT1100_ABX-LR-PL6-POOL15-1_fragmentEstimate_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │       ├── PROJECT1100_ABX-LR-PL6-POOL15-1_fragmentEstimate_matrix.tsv.xz</span>
<span class="co">#&gt; │   │       ├── PROJECT1100_ABX-LR-PL6-POOL15-1_seqCount_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │       └── PROJECT1100_ABX-LR-PL6-POOL15-1_seqCount_matrix.tsv.xz</span>
<span class="co">#&gt; │   └── report</span>
<span class="co">#&gt; ├── PROJECT1101</span>
<span class="co">#&gt; │   ├── bam</span>
<span class="co">#&gt; │   ├── bcmuxall</span>
<span class="co">#&gt; │   ├── bed</span>
<span class="co">#&gt; │   ├── iss</span>
<span class="co">#&gt; │   │   └── stats.sequence_PROJECT1101.ABY-LR-PL4-POOL54-2.tsv</span>
<span class="co">#&gt; │   ├── quality</span>
<span class="co">#&gt; │   ├── quantification</span>
<span class="co">#&gt; │   │   └── ABY-LR-PL4-POOL54-2</span>
<span class="co">#&gt; │   │       ├── PROJECT1101_ABY-LR-PL4-POOL54-2_fragmentEstimate_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │       ├── PROJECT1101_ABY-LR-PL4-POOL54-2_fragmentEstimate_matrix.tsv.xz</span>
<span class="co">#&gt; │   │       ├── PROJECT1101_ABY-LR-PL4-POOL54-2_seqCount_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │       └── PROJECT1101_ABY-LR-PL4-POOL54-2_seqCount_matrix.tsv.xz</span>
<span class="co">#&gt; │   └── report</span>
<span class="co">#&gt; └── VA</span>
<span class="co">#&gt;     └── CLOEXP</span>
<span class="co">#&gt;         ├── bam</span>
<span class="co">#&gt;         ├── bcmuxall</span>
<span class="co">#&gt;         ├── bed</span>
<span class="co">#&gt;         ├── iss</span>
<span class="co">#&gt;         │   └── stats.sequence_VA.POOL6-1.tsv</span>
<span class="co">#&gt;         ├── quality</span>
<span class="co">#&gt;         ├── quantification</span>
<span class="co">#&gt;         │   └── POOL6-1</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_fragmentEstimate_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_fragmentEstimate_matrix.tsv.xz</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_seqCount_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt;         │       └── VA-CLOEXP-POOL6-1_seqCount_matrix.tsv.xz</span>
<span class="co">#&gt;         └── report</span>
</pre></div>
</div>
</div>
<div id="importing-a-single-integration-matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-a-single-integration-matrix" class="anchor"></a>Importing a single integration matrix</h1>
<p>If you want to import a single integration matrix you can do so by using the <code>import_single_Vispa2Matrix</code> function.<br>
This function reads the .tsv file and converts it into a tidy tibble.<br>
If you’re not familiar with the “tidy” concept, we recommend to take a look at this link to get the basics:</p>
<ul>
<li><a href="https://bookdown.org/rdpeng/RProgDA/the-importance-of-tidy-data.html">The importance of tidy data</a></li>
</ul>
<p>This package is in fact based on the <code>tidyverse</code> and tries to follow its philosophy and guidelines as close as possible.<br>
Vispa2 pipeline and the associated Create Matrix tool produce matrices with a standard structure which we’ll refer to as “messy”, because different experimental data is divided in different columns and there are a lot of NA values.</p>
<table class="table">
<caption>A simple example of messy matrix.</caption>
<thead><tr class="header">
<th align="left">chr</th>
<th align="left">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">exp_1</th>
<th align="left">exp_2</th>
<th align="left">exp_3</th>
<th align="left">exp_4</th>
<th align="left">exp_5</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">20</td>
<td align="left">39085641</td>
<td align="left">+</td>
<td align="left">MAFB</td>
<td align="left">-</td>
<td align="left">NA</td>
<td align="left">94925.98</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">61143.20</td>
</tr>
<tr class="even">
<td align="left">7</td>
<td align="left">155446768</td>
<td align="left">+</td>
<td align="left">RBM33</td>
<td align="left">+</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">68981.02</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">19</td>
<td align="left">12278742</td>
<td align="left">-</td>
<td align="left">ZNF136</td>
<td align="left">+</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">60410.18</td>
<td align="left">NA</td>
<td align="left">68463.19</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">1865825</td>
<td align="left">-</td>
<td align="left">GMDS</td>
<td align="left">-</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">46666.04</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">7</td>
<td align="left">21508806</td>
<td align="left">+</td>
<td align="left">SP4</td>
<td align="left">+</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">17</td>
<td align="left">5187666</td>
<td align="left">+</td>
<td align="left">RABEP1</td>
<td align="left">+</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">example_matrix_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ex_annotated_ISMatrix.tsv.xz"</span>, 
                                   package = <span class="st">"ISAnalytics"</span>)
<span class="kw">imported_im</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_single_Vispa2Matrix.html">import_single_Vispa2Matrix</a></span>(<span class="kw">example_matrix_path</span>)
</pre></div>
<table class="table">
<caption>Example of tidy integration matrix</caption>
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="left">GeneName</th>
<th align="left">GeneStrand</th>
<th align="left">CompleteAmplificationID</th>
<th align="right">Value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">8744319</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">exp_2</td>
<td align="right">85050.01</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">8744319</td>
<td align="left">-</td>
<td align="left">RERE</td>
<td align="left">-</td>
<td align="left">exp_3</td>
<td align="right">36910.66</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="right">89274341</td>
<td align="left">+</td>
<td align="left">PKN2</td>
<td align="left">+</td>
<td align="left">exp_2</td>
<td align="right">12566.61</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">89274341</td>
<td align="left">+</td>
<td align="left">PKN2</td>
<td align="left">+</td>
<td align="left">exp_3</td>
<td align="right">61319.67</td>
</tr>
<tr class="odd">
<td align="left">2</td>
<td align="right">33962502</td>
<td align="left">+</td>
<td align="left">LINC01317</td>
<td align="left">+</td>
<td align="left">exp_2</td>
<td align="right">20246.97</td>
</tr>
<tr class="even">
<td align="left">2</td>
<td align="right">33962502</td>
<td align="left">+</td>
<td align="left">LINC01317</td>
<td align="left">+</td>
<td align="left">exp_3</td>
<td align="right">98047.22</td>
</tr>
</tbody>
</table>
<p>We will refer to the structure generated by <code>import_single_Vispa2Matrix</code> as “integration matrix” for convenience.<br>
To be considered an integration matrix the tibble must contain the mandatory variables, which are “chr” (chromosome), “integration_locus” and “strand”. It might also contain annotation variables if the matrix was annotated during the Vispa2 pipeline run.<br>
You can access these names by using two functions:</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co"># Displays the mandatory vars, can be called also for manipulation purposes </span>
<span class="co"># on tibble instead of calling individual variables</span>
<span class="fu"><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars</a></span>()
<span class="co">#&gt; [1] "chr"               "integration_locus" "strand"</span>

<span class="co"># Displays the annotation variables</span>
<span class="fu"><a href="../reference/annotation_IS_vars.html">annotation_IS_vars</a></span>()
<span class="co">#&gt; [1] "GeneName"   "GeneStrand"</span>
</pre></div>
<p>You can of course operate on the integration matrices as you would on any other tibble, but some functions will check the presence of specific columns because they’re needed in that context.</p>
</div>
<div id="importing-the-association-file" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-the-association-file" class="anchor"></a>Importing the association file</h1>
<p>While you can import single matrices for brief analysis, what you would like to do most of the times is import multiple matrices at once, based on certain parameters. To do that you must first import the association file, which is the file that holds all associated metadata and information about every project, pool and single experiment.<br>
The function that imports this file does not simply read it into your R environment, but performs an alignment check with your file system, so you have to specify the path to the root folder where your Vispa2 runs produce output (see the previous section). To import the association file do:</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(ISAnalytics.widgets = <span class="fl">FALSE</span>)
<span class="kw">path_as_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ex_association_file.tsv"</span>,
                            package = <span class="st">"ISAnalytics"</span>)
<span class="kw">association_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_association_file.html">import_association_file</a></span>(<span class="kw">path_as_file</span>, <span class="kw">root_correct</span>)
</pre></div>
<p>If you have the “widgets” option active, this will produce a visual HTML report of the results of the alignment check, either in Rstudio or in your browser.<br>
If projects or pools are missing you will be notified: until you fix those problems, those elements will be ignored until you re-import the association file.</p>
</div>
<div id="importing-multiple-matrices-in-parallel" class="section level1">
<h1 class="hasAnchor">
<a href="#importing-multiple-matrices-in-parallel" class="anchor"></a>Importing multiple matrices in parallel</h1>
<p>There are 2 different functions for importing multiple matrices in parallel:</p>
<ul>
<li><code>import_parallel_Vispa2Matrices_interactive</code></li>
<li><code>import_parallel_Vispa2Matrices_auto</code></li>
</ul>
<p>The interactive version will ask you to input your choices directly into the console, the automatic version will not, but has some limitations.<br>
Both functions rely on the association file and some basic parameters, most notably:</p>
<ul>
<li>
<code>quantification_type</code>: this is a string or a vector of characters indicating which quantification types you want the function to look for. The possible values are fragmentEstimate, seqCount, barcodeCount, cellCount, ShsCount</li>
<li>
<code>matrix_type</code>: tells the function if it should consider annotated or not annotated matrices. The only possible options are “annotated” and “not_annotated”</li>
<li>
<code>workers</code>: indicates the number of parallel workers to instantiate when importing. Keep in mind that the higher is the number, the faster the process is, but also higher is the RAM peak, so you should be aware of this especially if you’re dealing with really big matrices. Set this parameter according to your needs and according to your hardware specifications.</li>
</ul>
<p>Both the versions will produce several HTML reports after all key steps, such as:</p>
<ul>
<li>Report on which files were found, reporting eventual anomalies (missing files or duplicates)</li>
<li>Report on which files were chosen for import after interactive selection or after automatic filtering</li>
<li>Report on which files were actually imported, signaling potential errors during the import phase</li>
</ul>
<p>The functions both return a named list which names are the quantification types specified in input and the elements are integration matrices, obtained by binding rows of all imported matrices for the same quantification type.</p>
<div id="interactive-version" class="section level2">
<h2 class="hasAnchor">
<a href="#interactive-version" class="anchor"></a>Interactive version</h2>
<p>As stated before, with the interactive version you have more control and you can directly choose:</p>
<ul>
<li>Which projects to import</li>
<li>Which pools to import</li>
<li>If duplicates files are found which ones are to be kept</li>
</ul>
<p>If you haven’t imported the association file yet, you can directly pass the path to the association file and the path to the root folder into the function: in this way the association file will automatically be imported.</p>
<p>Example:</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw">matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices_interactive.html">import_parallel_Vispa2Matrices_interactive</a></span>(
  association_file = <span class="kw">path_as_file</span>,
  root = <span class="kw">root_correct</span>,
  quantification_type = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"fragmentEstimate"</span>, <span class="st">"seqCount"</span>),
  matrix_type = <span class="st">"annotated"</span>,
  workers = <span class="fl">2</span>)
</pre></div>
<p>If you’ve already imported the association file you can instead call the function like this:</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices_interactive.html">import_parallel_Vispa2Matrices_interactive</a></span>(
  association_file = <span class="kw">association_file</span>,
  root = <span class="kw">NULL</span>,
  quantification_type = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"fragmentEstimate"</span>, <span class="st">"seqCount"</span>),
  matrix_type = <span class="st">"annotated"</span>,
  workers = <span class="fl">2</span>)
</pre></div>
<p>You can simply access the data frames by doing:</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">matrices</span><span class="op">$</span><span class="kw">fragmentEstimate</span>
<span class="kw">matrices</span><span class="op">$</span><span class="kw">seqCount</span>
</pre></div>
</div>
<div id="automatic-version" class="section level2">
<h2 class="hasAnchor">
<a href="#automatic-version" class="anchor"></a>Automatic version</h2>
<p>If you choose to opt for the automatic version you should keep in mind that the function automatically considers everything included in the association file, so if you want to import only a subset of projects and/or pools you should filter the association file according to your criteria before calling the function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">magrittr</span>)
<span class="kw">refined_af</span> <span class="op">&lt;-</span> <span class="kw">association_file</span> <span class="op">%&gt;%</span> <span class="kw">dplyr</span>::<span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw">.data</span><span class="op">$</span><span class="kw">ProjectID</span> <span class="op">==</span> <span class="st">"CLOEXP"</span>)
</pre></div>
<p><strong>NOTE: do not drop any columns when filtering the association file or the file will be flagged as malformed</strong></p>
<p>In automatic version there is no way of discriminating duplicates, so there is the possibility to specify additional patterns to look for in file names to mitigate this problem. However, if after matching of the additional patterns duplicates are still found they’re simply discarded.</p>
<p>There are 2 additional parameters to set:</p>
<ul>
<li>
<code>patterns</code>: a string or a character vector containing regular expressions to be matched on file names. If you’re not familiar with regular expressions, I suggest you to start from here <a href="https://github.com/rstudio/cheatsheets/blob/master/strings.pdf">stringr cheatsheet</a>
</li>
<li>
<code>matching_opt</code>: a single string that tells the function how to match the patterns. The possible values for this parameter are ANY, ALL, OPTIONAL:
<ul>
<li>ANY : looks for files that match at least one of the patterns specified in <code>patterns</code>
</li>
<li>ALL: looks for files that match all the patterns specified in <code>patterns</code>
</li>
<li>OPTIONAL: looks for files that preferentially match all the patterns, if none are found looks for files that match any of the patterns and finally if none are found simply looks for files present that match the quantification type.</li>
</ul>
</li>
</ul>
<p>You can call the function with <code>patterns</code> set to NULL if you don’t wish to match anything:</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">matrices_auto</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices_auto.html">import_parallel_Vispa2Matrices_auto</a></span>(
  association_file = <span class="kw">refined_af</span>,
  root = <span class="kw">NULL</span>,
  quantification_type = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"fragmentEstimate"</span>,<span class="st">"seqCount"</span>),
  matrix_type = <span class="st">"annotated"</span>,
  workers = <span class="fl">2</span>,
  patterns = <span class="kw">NULL</span>,
  matching_opt = <span class="st">"ANY"</span> <span class="co"># Same if you choose "ALL" or "OPTIONAL"</span>
)
<span class="kw">matrices_auto</span>
<span class="co">#&gt; $fragmentEstimate</span>
<span class="co">#&gt; # A tibble: 1,476 x 7</span>
<span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand CompleteAmplificati… Value</span>
<span class="co">#&gt;    &lt;chr&gt;             &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                &lt;dbl&gt;</span>
<span class="co">#&gt;  1 1             100903226 +      CDC14A   +          CLOEXP_POOL6_LTR30L…  1.00</span>
<span class="co">#&gt;  2 1             101724095 -      S1PR1    +          CLOEXP_POOL6_LTR10L…  2.00</span>
<span class="co">#&gt;  3 1             108208779 +      VAV3     -          CLOEXP_POOL6_LTR32L…  1.00</span>
<span class="co">#&gt;  4 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR10L…  1.00</span>
<span class="co">#&gt;  5 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR12L…  1.00</span>
<span class="co">#&gt;  6 1             108511440 +      VAV3-AS1 +          CLOEXP_POOL6_LTR49L…  1.00</span>
<span class="co">#&gt;  7 1             109689050 +      KIAA1324 +          CLOEXP_POOL6_LTR29L…  1.00</span>
<span class="co">#&gt;  8 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR10L…  1.00</span>
<span class="co">#&gt;  9 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR32L…  1.00</span>
<span class="co">#&gt; 10 1             111466980 -      LRIF1    -          CLOEXP_POOL6_LTR10L…  1.00</span>
<span class="co">#&gt; # … with 1,466 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $seqCount</span>
<span class="co">#&gt; # A tibble: 1,476 x 7</span>
<span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand CompleteAmplificati… Value</span>
<span class="co">#&gt;    &lt;chr&gt;             &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                &lt;int&gt;</span>
<span class="co">#&gt;  1 1             100903226 +      CDC14A   +          CLOEXP_POOL6_LTR30L…     1</span>
<span class="co">#&gt;  2 1             101724095 -      S1PR1    +          CLOEXP_POOL6_LTR10L…     2</span>
<span class="co">#&gt;  3 1             108208779 +      VAV3     -          CLOEXP_POOL6_LTR32L…     1</span>
<span class="co">#&gt;  4 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR10L…     1</span>
<span class="co">#&gt;  5 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR12L…     1</span>
<span class="co">#&gt;  6 1             108511440 +      VAV3-AS1 +          CLOEXP_POOL6_LTR49L…     1</span>
<span class="co">#&gt;  7 1             109689050 +      KIAA1324 +          CLOEXP_POOL6_LTR29L…     1</span>
<span class="co">#&gt;  8 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR10L…     1</span>
<span class="co">#&gt;  9 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR32L…     1</span>
<span class="co">#&gt; 10 1             111466980 -      LRIF1    -          CLOEXP_POOL6_LTR10L…     1</span>
<span class="co">#&gt; # … with 1,466 more rows</span>
</pre></div>
<p>Let’s do an example with a file system where there are issues, such as duplicates:</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">root_err</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"fserr.zip"</span>, package = <span class="st">"ISAnalytics"</span>)
<span class="kw">root_err</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unzip_file_system.html">unzip_file_system</a></span>(<span class="kw">root_err</span>, <span class="st">"fserr"</span>)
<span class="kw">fs</span>::<span class="fu"><a href="http://fs.r-lib.org/reference/dir_tree.html">dir_tree</a></span>(<span class="kw">root_err</span>)
<span class="co">#&gt; /tmp/RtmpzcGFHO/fserr</span>
<span class="co">#&gt; ├── PROJECT1100</span>
<span class="co">#&gt; │   ├── bam</span>
<span class="co">#&gt; │   ├── bcmuxall</span>
<span class="co">#&gt; │   ├── bed</span>
<span class="co">#&gt; │   ├── iss</span>
<span class="co">#&gt; │   │   └── stats.sequence_PROJECT1100.ABX-LR-PL5-POOL14-1.tsv</span>
<span class="co">#&gt; │   ├── quality</span>
<span class="co">#&gt; │   ├── quantification</span>
<span class="co">#&gt; │   │   ├── ABX-LR-PL5-POOL14-1</span>
<span class="co">#&gt; │   │   │   ├── PROJECT1100_ABX-LR-PL5-POOL14-1_fragmentEstimate_matrix-NoMate.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │   │   ├── PROJECT1100_ABX-LR-PL5-POOL14-1_fragmentEstimate_matrix-NoMate2.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │   │   ├── PROJECT1100_ABX-LR-PL5-POOL14-1_fragmentEstimate_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │   │   └── PROJECT1100_ABX-LR-PL5-POOL14-1_seqCount_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │   └── ABX-LR-PL6-POOL15-1</span>
<span class="co">#&gt; │   │       ├── PROJECT1100_ABX-LR-PL6-POOL15-1_fragmentEstimate_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   │       └── PROJECT1100_ABX-LR-PL6-POOL15-1_seqCount_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt; │   └── report</span>
<span class="co">#&gt; └── VA</span>
<span class="co">#&gt;     └── CLOEXP</span>
<span class="co">#&gt;         ├── bam</span>
<span class="co">#&gt;         ├── bcmuxall</span>
<span class="co">#&gt;         ├── bed</span>
<span class="co">#&gt;         ├── iss</span>
<span class="co">#&gt;         │   └── stats.sequence_VA.POOL6-1.tsv</span>
<span class="co">#&gt;         ├── quality</span>
<span class="co">#&gt;         ├── quantification</span>
<span class="co">#&gt;         │   └── POOL6-1</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_fragmentEstimate_matrix-NoMate.no0.annotated.tsv.xz</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_fragmentEstimate_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_fragmentEstimate_matrix.tsv.xz</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_seqCount_matrix-NoMate.no0.annotated.tsv.xz</span>
<span class="co">#&gt;         │       ├── VA-CLOEXP-POOL6-1_seqCount_matrix.no0.annotated.tsv.xz</span>
<span class="co">#&gt;         │       └── VA-CLOEXP-POOL6-1_seqCount_matrix.tsv.xz</span>
<span class="co">#&gt;         └── report</span>
<span class="kw">association_file_fserr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_association_file.html">import_association_file</a></span>(<span class="kw">path_as_file</span>, <span class="kw">root_err</span>)
<span class="co">#&gt; Warning: One or more projects were not found in the filesystem starting from /</span>
<span class="co">#&gt; tmp/RtmpzcGFHO/fserr, please check yourassociation file for errors and/or your</span>
<span class="co">#&gt; file system.Until you re-import the association file these missing files will be</span>
<span class="co">#&gt; ignored.</span>
<span class="kw">refined_af_err</span> <span class="op">&lt;-</span> <span class="kw">association_file_fserr</span> <span class="op">%&gt;%</span> 
  <span class="kw">dplyr</span>::<span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw">.data</span><span class="op">$</span><span class="kw">ProjectID</span> <span class="op">==</span> <span class="st">"CLOEXP"</span>)
<span class="kw">matrices_auto2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices_auto.html">import_parallel_Vispa2Matrices_auto</a></span>(
  association_file = <span class="kw">refined_af_err</span>,
  root = <span class="kw">NULL</span>,
  quantification_type = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"fragmentEstimate"</span>,<span class="st">"seqCount"</span>),
  matrix_type = <span class="st">"annotated"</span>,
  workers = <span class="fl">2</span>,
  patterns = <span class="st">"NoMate"</span>,
  matching_opt = <span class="st">"ANY"</span> <span class="co"># Same if you choose "ALL" or "OPTIONAL"</span>
)
<span class="kw">matrices_auto</span>
<span class="co">#&gt; $fragmentEstimate</span>
<span class="co">#&gt; # A tibble: 1,476 x 7</span>
<span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand CompleteAmplificati… Value</span>
<span class="co">#&gt;    &lt;chr&gt;             &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                &lt;dbl&gt;</span>
<span class="co">#&gt;  1 1             100903226 +      CDC14A   +          CLOEXP_POOL6_LTR30L…  1.00</span>
<span class="co">#&gt;  2 1             101724095 -      S1PR1    +          CLOEXP_POOL6_LTR10L…  2.00</span>
<span class="co">#&gt;  3 1             108208779 +      VAV3     -          CLOEXP_POOL6_LTR32L…  1.00</span>
<span class="co">#&gt;  4 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR10L…  1.00</span>
<span class="co">#&gt;  5 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR12L…  1.00</span>
<span class="co">#&gt;  6 1             108511440 +      VAV3-AS1 +          CLOEXP_POOL6_LTR49L…  1.00</span>
<span class="co">#&gt;  7 1             109689050 +      KIAA1324 +          CLOEXP_POOL6_LTR29L…  1.00</span>
<span class="co">#&gt;  8 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR10L…  1.00</span>
<span class="co">#&gt;  9 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR32L…  1.00</span>
<span class="co">#&gt; 10 1             111466980 -      LRIF1    -          CLOEXP_POOL6_LTR10L…  1.00</span>
<span class="co">#&gt; # … with 1,466 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $seqCount</span>
<span class="co">#&gt; # A tibble: 1,476 x 7</span>
<span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand CompleteAmplificati… Value</span>
<span class="co">#&gt;    &lt;chr&gt;             &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;                &lt;int&gt;</span>
<span class="co">#&gt;  1 1             100903226 +      CDC14A   +          CLOEXP_POOL6_LTR30L…     1</span>
<span class="co">#&gt;  2 1             101724095 -      S1PR1    +          CLOEXP_POOL6_LTR10L…     2</span>
<span class="co">#&gt;  3 1             108208779 +      VAV3     -          CLOEXP_POOL6_LTR32L…     1</span>
<span class="co">#&gt;  4 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR10L…     1</span>
<span class="co">#&gt;  5 1             108412817 -      VAV3     -          CLOEXP_POOL6_LTR12L…     1</span>
<span class="co">#&gt;  6 1             108511440 +      VAV3-AS1 +          CLOEXP_POOL6_LTR49L…     1</span>
<span class="co">#&gt;  7 1             109689050 +      KIAA1324 +          CLOEXP_POOL6_LTR29L…     1</span>
<span class="co">#&gt;  8 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR10L…     1</span>
<span class="co">#&gt;  9 1             111411778 +      CD53     +          CLOEXP_POOL6_LTR32L…     1</span>
<span class="co">#&gt; 10 1             111466980 -      LRIF1    -          CLOEXP_POOL6_LTR10L…     1</span>
<span class="co">#&gt; # … with 1,466 more rows</span>
</pre></div>
<p>As you can see, in the file system with issues we have more than one file for quantification type, duplicates have “NoMate” suffix in their file name. By specifying this pattern in the function, we’re only going to import those files.</p>
<p>As for the interactive version, you can call the function with path to the association file and root if you want to simply import everything without filtering.</p>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>The <em><a href="https://bioconductor.org/packages/3.12/ISAnalytics">ISAnalytics</a></em> package <a id="cite-calabrialab_2020"></a>(<a href="http://www.bioconductor.org/packages/ISAnalytics">calabrialab, 2020</a>) was made possible thanks to:</p>
<ul>
<li>R <a id="cite-R_Core_Team_2020"></a>(<a href="https://www.R-project.org/">R Core Team, 2020</a>)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> <a id="cite-Oles_2020"></a>(<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> <a id="cite-Boettiger_2019"></a>(<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> <a id="cite-Xie_2020"></a>(<a href="https://yihui.org/knitr/">Xie, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> <a id="cite-Allaire_2020"></a>(<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=sessioninfo">sessioninfo</a></em> <a id="cite-Csardi_2018"></a>(<a href="https://github.com/r-lib/sessioninfo#readme">Csárdi, core, Wickham, Chang, et al., 2018</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=testthat">testthat</a></em> <a id="cite-Wickham_2011"></a>(<a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">Wickham, 2011</a>)</li>
</ul>
<p>This package was developed using <em><a href="https://github.com/lcolladotor/biocthis">biocthis</a></em>.</p>
<p><code>R</code> session information.</p>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  setting  value                                      
#&gt;  version  R version 4.0.2 Patched (2020-09-17 r79226)
#&gt;  os       Ubuntu 20.04.1 LTS                         
#&gt;  system   x86_64, linux-gnu                          
#&gt;  ui       X11                                        
#&gt;  language (EN)                                       
#&gt;  collate  en_US.UTF-8                                
#&gt;  ctype    en_US.UTF-8                                
#&gt;  tz       UTC                                        
#&gt;  date     2020-09-21                                 
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  package       * version    date       lib source                        
#&gt;  assertthat      0.2.1      2019-03-21 [2] RSPM (R 4.0.0)                
#&gt;  backports       1.1.10     2020-09-15 [2] RSPM (R 4.0.2)                
#&gt;  bibtex          0.4.2.2    2020-01-02 [1] RSPM (R 4.0.0)                
#&gt;  BiocManager     1.30.10    2019-11-16 [2] CRAN (R 4.0.2)                
#&gt;  BiocParallel    1.23.2     2020-07-06 [1] Bioconductor                  
#&gt;  BiocStyle     * 2.17.0     2020-04-27 [1] Bioconductor                  
#&gt;  bookdown        0.20       2020-06-23 [1] RSPM (R 4.0.2)                
#&gt;  cli             2.0.2      2020-02-28 [2] RSPM (R 4.0.0)                
#&gt;  cpp11           0.2.1      2020-08-11 [1] RSPM (R 4.0.2)                
#&gt;  crayon          1.3.4      2017-09-16 [2] RSPM (R 4.0.0)                
#&gt;  desc            1.2.0      2018-05-01 [2] RSPM (R 4.0.0)                
#&gt;  digest          0.6.25     2020-02-23 [2] RSPM (R 4.0.0)                
#&gt;  dplyr           1.0.2      2020-08-18 [1] RSPM (R 4.0.2)                
#&gt;  ellipsis        0.3.1      2020-05-15 [2] RSPM (R 4.0.0)                
#&gt;  evaluate        0.14       2019-05-28 [2] RSPM (R 4.0.0)                
#&gt;  fansi           0.4.1      2020-01-08 [2] RSPM (R 4.0.0)                
#&gt;  forcats         0.5.0      2020-03-01 [1] RSPM (R 4.0.0)                
#&gt;  fs              1.5.0      2020-07-31 [2] RSPM (R 4.0.2)                
#&gt;  generics        0.0.2      2018-11-29 [1] RSPM (R 4.0.0)                
#&gt;  glue            1.4.2      2020-08-27 [2] RSPM (R 4.0.2)                
#&gt;  highr           0.8        2019-03-20 [2] RSPM (R 4.0.0)                
#&gt;  hms             0.5.3      2020-01-08 [1] RSPM (R 4.0.0)                
#&gt;  htmltools       0.5.0      2020-06-16 [2] RSPM (R 4.0.1)                
#&gt;  htmlwidgets     1.5.1      2019-10-08 [2] RSPM (R 4.0.0)                
#&gt;  httr            1.4.2      2020-07-20 [2] RSPM (R 4.0.2)                
#&gt;  ISAnalytics   * 0.99.11    2020-09-21 [1] Bioconductor                  
#&gt;  jsonlite        1.7.1      2020-09-07 [2] RSPM (R 4.0.2)                
#&gt;  knitcitations * 1.0.10     2019-09-15 [1] RSPM (R 4.0.0)                
#&gt;  knitr           1.29       2020-06-23 [2] RSPM (R 4.0.2)                
#&gt;  lifecycle       0.2.0      2020-03-06 [2] RSPM (R 4.0.0)                
#&gt;  lubridate       1.7.9      2020-06-08 [1] RSPM (R 4.0.2)                
#&gt;  magrittr      * 1.5        2014-11-22 [2] RSPM (R 4.0.0)                
#&gt;  memoise         1.1.0      2017-04-21 [2] RSPM (R 4.0.0)                
#&gt;  pillar          1.4.6      2020-07-10 [2] RSPM (R 4.0.2)                
#&gt;  pkgconfig       2.0.3      2019-09-22 [2] RSPM (R 4.0.0)                
#&gt;  pkgdown         1.6.1.9000 2020-09-21 [1] Github (r-lib/pkgdown@e1e46f2)
#&gt;  plyr            1.8.6      2020-03-03 [1] RSPM (R 4.0.2)                
#&gt;  purrr           0.3.4      2020-04-17 [2] RSPM (R 4.0.0)                
#&gt;  R6              2.4.1      2019-11-12 [2] RSPM (R 4.0.0)                
#&gt;  ragg            0.3.1      2020-07-03 [1] RSPM (R 4.0.2)                
#&gt;  Rcpp            1.0.5      2020-07-06 [2] RSPM (R 4.0.2)                
#&gt;  reactable       0.2.2      2020-09-13 [1] RSPM (R 4.0.2)                
#&gt;  readr           1.3.1      2018-12-21 [1] RSPM (R 4.0.2)                
#&gt;  RefManageR      1.2.12     2019-04-03 [1] RSPM (R 4.0.0)                
#&gt;  rlang           0.4.7      2020-07-09 [2] RSPM (R 4.0.2)                
#&gt;  rmarkdown       2.3        2020-06-18 [1] RSPM (R 4.0.1)                
#&gt;  rprojroot       1.3-2      2018-01-03 [2] RSPM (R 4.0.0)                
#&gt;  sessioninfo   * 1.1.1      2018-11-05 [2] RSPM (R 4.0.0)                
#&gt;  stringi         1.5.3      2020-09-09 [2] RSPM (R 4.0.2)                
#&gt;  stringr         1.4.0      2019-02-10 [2] RSPM (R 4.0.0)                
#&gt;  systemfonts     0.3.1      2020-09-08 [1] RSPM (R 4.0.2)                
#&gt;  tibble          3.0.3      2020-07-10 [2] RSPM (R 4.0.2)                
#&gt;  tidyr           1.1.2      2020-08-27 [1] RSPM (R 4.0.2)                
#&gt;  tidyselect      1.1.0      2020-05-11 [1] RSPM (R 4.0.0)                
#&gt;  utf8            1.1.4      2018-05-24 [2] RSPM (R 4.0.0)                
#&gt;  vctrs           0.3.4      2020-08-29 [2] RSPM (R 4.0.2)                
#&gt;  withr           2.2.0      2020-04-20 [2] RSPM (R 4.0.0)                
#&gt;  xfun            0.17       2020-09-09 [2] RSPM (R 4.0.2)                
#&gt;  xml2            1.3.2      2020-04-23 [2] RSPM (R 4.0.0)                
#&gt;  yaml            2.2.1      2020-02-01 [2] RSPM (R 4.0.0)                
#&gt;  zip             2.1.1      2020-08-27 [1] RSPM (R 4.0.2)                
#&gt; 
#&gt; [1] /usr/local/lib/R/host-site-library
#&gt; [2] /usr/local/lib/R/site-library
#&gt; [3] /usr/local/lib/R/library</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> (<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (<a href="https://yihui.org/knitr/">Xie, 2020</a>) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> (<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> (<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>).</p>
<p>[1] J. Allaire, Y. Xie, J. McPherson, J. Luraschi, et al. <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.3. 2020. &lt;URL: <a href="https://github.com/rstudio/rmarkdown" class="uri">https://github.com/rstudio/rmarkdown</a>&gt;.</p>
<p>[2] C. Boettiger. <em>knitcitations: Citations for ‘Knitr’ Markdown Files</em>. R package version 1.0.10. 2019. &lt;URL: <a href="https://github.com/cboettig/knitcitations" class="uri">https://github.com/cboettig/knitcitations</a>&gt;.</p>
<p>[3] calabrialab. <em>Analyze gene therapy vector insertion sites data identified from genomics next generation sequencing reads for clonal tracking studies</em>. <a href="https://github.com/calabrialab/ISAnalytics" class="uri">https://github.com/calabrialab/ISAnalytics</a> - R package version 0.99.11. 2020. DOI: 10.18129/B9.bioc.ISAnalytics. &lt;URL: <a href="http://www.bioconductor.org/packages/ISAnalytics" class="uri">http://www.bioconductor.org/packages/ISAnalytics</a>&gt;.</p>
<p>[4] G. Csárdi, R. core, H. Wickham, W. Chang, et al. <em>sessioninfo: R Session Information</em>. R package version 1.1.1. 2018. &lt;URL: <a href="https://github.com/r-lib/sessioninfo#readme" class="uri">https://github.com/r-lib/sessioninfo#readme</a>&gt;.</p>
<p>[5] A. Oleś, M. Morgan, and W. Huber. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.17.0. 2020. &lt;URL: <a href="https://github.com/Bioconductor/BiocStyle" class="uri">https://github.com/Bioconductor/BiocStyle</a>&gt;.</p>
<p>[6] R Core Team. <em>R: A Language and Environment for Statistical Computing</em>. R Foundation for Statistical Computing. Vienna, Austria, 2020. &lt;URL: <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>&gt;.</p>
<p>[7] H. Wickham. “testthat: Get Started with Testing”. In: <em>The R Journal</em> 3 (2011), pp. 5-10. &lt;URL: <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf" class="uri">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>&gt;.</p>
<p>[8] Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.29. 2020. &lt;URL: <a href="https://yihui.org/knitr/" class="uri">https://yihui.org/knitr/</a>&gt;.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
