<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ISAnalytics">
<title>Getting started with ISAnalytics • ISAnalytics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with ISAnalytics">
<meta property="og:description" content="ISAnalytics">
<meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ISAnalytics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/ISAnalytics.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/workflow_start.html">Setting up the workflow and first steps</a>
    <a class="dropdown-item" href="../articles/sharing_analyses.html">Sharing analyses with ISAnalytics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="https://calabrialab.github.io/ISAnalytics/dev/">
    <span class="fas fa fas fa-code-branch"></span>
     
    RELEASE
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting started with ISAnalytics</h1>
                        <h4 data-toc-skip class="author">Giulia
Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget,
Via Olgettina 60, 20132 Milano -
Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 data-toc-skip class="date">1 December 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/HEAD/vignettes/ISAnalytics.Rmd" class="external-link"><code>vignettes/ISAnalytics.Rmd</code></a></small>
      <div class="d-none name"><code>ISAnalytics.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>ISAnalytics is an R package developed to analyze gene therapy vector
insertion sites data identified from genomics next generation sequencing
reads for clonal tracking studies.</p>
</div>
<div class="section level2">
<h2 id="installation-and-options">Installation and options<a class="anchor" aria-label="anchor" href="#installation-and-options"></a>
</h2>
<p><code>ISAnalytics</code> can be installed quickly in different
ways:</p>
<ul>
<li>You can install it via <a href="http://bioconductor.org" class="external-link">Bioconductor</a>
</li>
<li>You can install it via GitHub using the package
<code>devtools</code>
</li>
</ul>
<p>There are always 2 versions of the package active:</p>
<ul>
<li>
<code>RELEASE</code> is the latest stable version</li>
<li>
<code>DEVEL</code> is the development version, it is the most
up-to-date version where all new features are introduced</li>
</ul>
<div class="section level3">
<h3 id="installation-from-bioconductor">Installation from bioconductor<a class="anchor" aria-label="anchor" href="#installation-from-bioconductor"></a>
</h3>
<p>RELEASE version:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span></span></code></pre></div>
<p>DEVEL version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The following initializes usage of Bioc devel</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="installation-from-github">Installation from GitHub<a class="anchor" aria-label="anchor" href="#installation-from-github"></a>
</h3>
<p>RELEASE:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,</span>
<span>                         ref <span class="op">=</span> <span class="st">"RELEASE_3_16"</span>,</span>
<span>                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>DEVEL:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,</span>
<span>                         ref <span class="op">=</span> <span class="st">"master"</span>,</span>
<span>                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="setting-options">Setting options<a class="anchor" aria-label="anchor" href="#setting-options"></a>
</h3>
<p><code>ISAnalytics</code> has a verbose option that allows some
functions to print additional information to the console while they’re
executing. To disable this feature do:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DISABLE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ENABLE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Some functions also produce report in a user-friendly HTML format, to
set this feature:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DISABLE HTML REPORTS</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.reports"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ENABLE HTML REPORTS</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.reports"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="setting-up-the-workflow">Setting up the workflow<a class="anchor" aria-label="anchor" href="#setting-up-the-workflow"></a>
</h2>
<p>In the newer version of ISAnalytics, we introduced a “dynamic
variables system”, to allow more flexibility in terms of input formats.
Before starting with the analysis workflow, you can specify how your
inputs are structured so that the package can process them. For more
information on how to do this take a look at
<code><a href="../articles/workflow_start.html">vignette("workflow_start", package = "ISAnalytics")</a></code>.</p>
</div>
<div class="section level2">
<h2 id="the-first-steps">The first steps<a class="anchor" aria-label="anchor" href="#the-first-steps"></a>
</h2>
<p>The first steps of the analysis workflow involve the import and
parsing of data and metadata files from disk.</p>
<ul>
<li>Import metadata with <code><a href="../reference/import_association_file.html">import_association_file()</a></code> and/or
<code><a href="../reference/import_Vispa2_stats.html">import_Vispa2_stats()</a></code>
</li>
<li>Import data with <code><a href="../reference/import_single_Vispa2Matrix.html">import_single_Vispa2Matrix()</a></code> or
<code><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices()</a></code>
</li>
</ul>
<p>Refer to the vignette
<code><a href="../articles/workflow_start.html">vignette("workflow_start", package = "ISAnalytics")</a></code> for
more details.</p>
</div>
<div class="section level2">
<h2 id="data-cleaning-and-pre-processing">Data cleaning and pre-processing<a class="anchor" aria-label="anchor" href="#data-cleaning-and-pre-processing"></a>
</h2>
<p>ISAnalytics offers several different functions for cleaning and
pre-processing your data.</p>
<ul>
<li>Recalibration: identifies integration events that are near to each
other and condenses them into a single event whenever appropriate -
<code><a href="../reference/compute_near_integrations.html">compute_near_integrations()</a></code>
</li>
<li>Outliers identification and removal: identifies samples that are
considered outliers according to user-defined logic and filters them out
- <code><a href="../reference/outlier_filter.html">outlier_filter()</a></code>
</li>
<li>Collision removal: identifies collision events between independent
samples - <code><a href="../reference/remove_collisions.html">remove_collisions()</a></code>, see also the dedicated
vignette
<code><a href="../articles/workflow_start.html">vignette("workflow_start", package = "ISAnalytics")</a></code>
</li>
<li>Filter based on cell lineage purity: identifies and removes
contamination between different cell types -
<code><a href="../reference/purity_filter.html">purity_filter()</a></code>
</li>
<li>Data and metadata aggregation: allows the union of biological
samples from single pcr replicates or other arbitrary aggregations -
<code><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key()</a></code>,
<code><a href="../reference/aggregate_metadata.html">aggregate_metadata()</a></code>, see also the dedicated vignette
<code><a href="../articles/workflow_start.html">vignette("workflow_start", package = "ISAnalytics")</a></code>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="answering-biological-questions">Answering biological questions<a class="anchor" aria-label="anchor" href="#answering-biological-questions"></a>
</h2>
<p>You can answer very different biological questions by using the
provided functions with appropriate inputs.</p>
<ul>
<li>Descriptive statistics: <code><a href="../reference/sample_statistics.html">sample_statistics()</a></code>
</li>
<li>IS relative abundance: <code><a href="../reference/compute_abundance.html">compute_abundance()</a></code>,
<code><a href="../reference/integration_alluvial_plot.html">integration_alluvial_plot()</a></code>
</li>
<li>Top abundant IS: <code><a href="../reference/top_integrations.html">top_integrations()</a></code>
</li>
<li>Top targeted genes: <code><a href="../reference/top_targeted_genes.html">top_targeted_genes()</a></code>
</li>
<li>Grubbs test for common insertion sites (CIS):
<code><a href="../reference/CIS_grubbs.html">CIS_grubbs()</a></code>, <code><a href="../reference/CIS_volcano_plot.html">CIS_volcano_plot()</a></code>
</li>
<li>Fisher’s exact test for gene frequency and IS distribution on target
genome: <code><a href="../reference/gene_frequency_fisher.html">gene_frequency_fisher()</a></code>,
<code><a href="../reference/fisher_scatterplot.html">fisher_scatterplot()</a></code>,
<code><a href="../reference/circos_genomic_density.html">circos_genomic_density()</a></code>
</li>
<li>Clonal sharing analyses: <code><a href="../reference/is_sharing.html">is_sharing()</a></code>,
<code><a href="../reference/iss_source.html">iss_source()</a></code>, <code><a href="../reference/sharing_heatmap.html">sharing_heatmap()</a></code>,
<code><a href="../reference/sharing_venn.html">sharing_venn()</a></code>
</li>
<li>Estimate HSPCs population size:
<code><a href="../reference/HSC_population_size_estimate.html">HSC_population_size_estimate()</a></code>,
<code><a href="../reference/HSC_population_plot.html">HSC_population_plot()</a></code>
</li>
</ul>
<p>For more, please refer to the full function reference.</p>
</div>
<div class="section level2">
<h2 id="working-with-other-kinds-of-data">Working with other kinds of data<a class="anchor" aria-label="anchor" href="#working-with-other-kinds-of-data"></a>
</h2>
<p>ISAnalytics is designed to be flexible concerning input formats, thus
it is suited to process various kinds of data provided the correct
dynamic configuration is set.</p>
<p>We demonstrate this with an example that uses barcodes data. The
matrix is publicly available <a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE144340&amp;format=file&amp;file=GSE144340%5FMatrix%5F542%2Etsv%2Egz" class="external-link">here</a>
(Ferrari Samuele Jacob Aurelien, 2020), metadata was provided to us by
the authors and it is available in the package additional files.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://calabrialab.github.io/ISAnalytics">ISAnalytics</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: magrittr</span></span>
<span></span>
<span><span class="co"># Set appropriate data and metadata specs ----</span></span>
<span><span class="va">metadata_specs</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span><span class="va">names</span>, <span class="op">~</span><span class="va">types</span>, <span class="op">~</span><span class="va">transform</span>, <span class="op">~</span><span class="va">flag</span>, <span class="op">~</span><span class="va">tag</span>,</span>
<span>    <span class="st">"ProjectID"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"project_id"</span>,</span>
<span>    <span class="st">"SubjectID"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"subject"</span>,</span>
<span>    <span class="st">"Tissue"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"tissue"</span>,</span>
<span>    <span class="st">"TimePoint"</span>, <span class="st">"int"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"tp_days"</span>,</span>
<span>    <span class="st">"CellMarker"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"cell_marker"</span>,</span>
<span>    <span class="st">"ID"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"pcr_repl_id"</span>,</span>
<span>    <span class="st">"SourceFileName"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"optional"</span>, <span class="cn">NA_character_</span>,</span>
<span>    <span class="st">"Link"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"optional"</span>, <span class="cn">NA_character_</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/set_mandatory_IS_vars.html">set_af_columns_def</a></span><span class="op">(</span><span class="va">metadata_specs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Warning: important tags missing</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Some tags are required for proper execution of some functions. If these tags are not provided, execution of dependent functions might fail. Review your inputs carefully.</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Missing tags: pool_id, fusion_id, tag_seq, vector_id, tag_id, pcr_replicate, vispa_concatenate, proj_folder</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> To see where these are involved type `inspect_tags(c('pool_id','fusion_id','tag_seq','vector_id','tag_id','pcr_replicate','vispa_concatenate','proj_folder'))`</span></span>
<span><span class="co">#&gt; Association file columns specs successfully changed</span></span>
<span></span>
<span><span class="va">mandatory_specs</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span><span class="va">names</span>, <span class="op">~</span><span class="va">types</span>, <span class="op">~</span><span class="va">transform</span>, <span class="op">~</span><span class="va">flag</span>, <span class="op">~</span><span class="va">tag</span>,</span>
<span>    <span class="st">"BarcodeSeq"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="cn">NA_character_</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/set_mandatory_IS_vars.html">set_mandatory_IS_vars</a></span><span class="op">(</span><span class="va">mandatory_specs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Warning: important tags missing</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Some tags are required for proper execution of some functions. If these tags are not provided, execution of dependent functions might fail. Review your inputs carefully.</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Missing tags: chromosome, locus, is_strand</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> To see where these are involved type `inspect_tags(c('chromosome','locus','is_strand'))`</span></span>
<span><span class="co">#&gt; Mandatory IS vars successfully changed</span></span>
<span></span>
<span><span class="co"># Files ----</span></span>
<span><span class="va">data_folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span>
<span><span class="va">meta_file</span> <span class="op">&lt;-</span> <span class="st">"barcodes_example_af.tsv.xz"</span></span>
<span><span class="va">matrix_file</span> <span class="op">&lt;-</span> <span class="st">"GSE144340_Matrix_542.tsv.xz"</span></span>
<span></span>
<span><span class="co"># Data import ----</span></span>
<span><span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_association_file.html">import_association_file</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">data_folder</span>, <span class="va">meta_file</span><span class="op">)</span>,</span>
<span>    report_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="va">af</span></span>
<span><span class="co">#&gt;        ProjectID SubjectID Tissue TimePoint CellMarker      ID</span></span>
<span><span class="co">#&gt;  1: PMID32601433        A0     BM        21      Whole   BM_A0</span></span>
<span><span class="co">#&gt;  2: PMID32601433        A0     PB        21      Whole PB21_A0</span></span>
<span><span class="co">#&gt;  3: PMID32601433        A1     BM        21      Whole   BM_A1</span></span>
<span><span class="co">#&gt;  4: PMID32601433        A1     PB        21      Whole PB21_A1</span></span>
<span><span class="co">#&gt;  5: PMID32601433        A2     PB        21      Whole PB21_A2</span></span>
<span><span class="co">#&gt;  6: PMID32601433        A3     PB        21      Whole PB21_A3</span></span>
<span><span class="co">#&gt;  7: PMID32601433        A4     BM        21      Whole   BM_A4</span></span>
<span><span class="co">#&gt;  8: PMID32601433        A4     PB        21      Whole PB21_A4</span></span>
<span><span class="co">#&gt;  9: PMID32601433        C0     PB        21      Whole PB21_C0</span></span>
<span><span class="co">#&gt; 10: PMID32601433        C1     BM        21      Whole   BM_C1</span></span>
<span><span class="co">#&gt; 11: PMID32601433        C1     PB        21      Whole PB21_C1</span></span>
<span><span class="co">#&gt; 12: PMID32601433        C2     BM        21      Whole   BM_C2</span></span>
<span><span class="co">#&gt; 13: PMID32601433        C2     PB        21      Whole PB21_C2</span></span>
<span><span class="co">#&gt; 14: PMID32601433        C3     BM        21      Whole   BM_C3</span></span>
<span><span class="co">#&gt; 15: PMID32601433        C3     PB        21      Whole PB21_C3</span></span>
<span><span class="co">#&gt;                  SourceFileName</span></span>
<span><span class="co">#&gt;  1: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  2: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  3: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  4: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  5: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  6: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  7: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  8: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  9: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 10: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 11: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 12: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 13: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 14: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 15: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;                                                                                                      Link</span></span>
<span><span class="co">#&gt;  1: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  2: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  3: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  4: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  5: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  6: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  7: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  8: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  9: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 10: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 11: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 12: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 13: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 14: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 15: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;     TimepointMonths TimepointYears</span></span>
<span><span class="co">#&gt;  1:              01             01</span></span>
<span><span class="co">#&gt;  2:              01             01</span></span>
<span><span class="co">#&gt;  3:              01             01</span></span>
<span><span class="co">#&gt;  4:              01             01</span></span>
<span><span class="co">#&gt;  5:              01             01</span></span>
<span><span class="co">#&gt;  6:              01             01</span></span>
<span><span class="co">#&gt;  7:              01             01</span></span>
<span><span class="co">#&gt;  8:              01             01</span></span>
<span><span class="co">#&gt;  9:              01             01</span></span>
<span><span class="co">#&gt; 10:              01             01</span></span>
<span><span class="co">#&gt; 11:              01             01</span></span>
<span><span class="co">#&gt; 12:              01             01</span></span>
<span><span class="co">#&gt; 13:              01             01</span></span>
<span><span class="co">#&gt; 14:              01             01</span></span>
<span><span class="co">#&gt; 15:              01             01</span></span>
<span></span>
<span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_single_Vispa2Matrix.html">import_single_Vispa2Matrix</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">data_folder</span>, <span class="va">matrix_file</span><span class="op">)</span>,</span>
<span>    sample_names_to <span class="op">=</span> <span class="st">"ID"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: compression format not supported by fread</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> File will be read using readr</span></span>
<span><span class="co">#&gt; Reading file...</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Mode: classic</span></span>
<span><span class="co">#&gt; Reshaping...</span></span>
<span><span class="co">#&gt; *** File info *** </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">•</span> --- Annotated: FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">•</span> --- Dimensions: 31757 x 16</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">•</span> --- Read mode: classic</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">•</span> --- Sample count: 15</span></span>
<span><span class="va">matrix</span></span>
<span><span class="co">#&gt;                    BarcodeSeq      ID Value</span></span>
<span><span class="co">#&gt;     1: AAAAAAAATTTTTAAACGTACC   BM_A0     1</span></span>
<span><span class="co">#&gt;     2: AAAAAACATATCTATAGTTACC   BM_A0     1</span></span>
<span><span class="co">#&gt;     3: AAAAAATATATAAATAGATACC   BM_A0     1</span></span>
<span><span class="co">#&gt;     4: AAAAACAACAAGGAAATTCAAT   BM_A0     1</span></span>
<span><span class="co">#&gt;     5: AAAAACAACGAGGATAGTGAAT   BM_A0     1</span></span>
<span><span class="co">#&gt;    ---                                     </span></span>
<span><span class="co">#&gt; 33772: TTTTGAGACCTTCACACCTACT PB21_C3     1</span></span>
<span><span class="co">#&gt; 33773: TTTTGCCACCTTCATACCCAAC PB21_C3     1</span></span>
<span><span class="co">#&gt; 33774: TTTTTAAACCGTTAGACCCGCA PB21_C3     1</span></span>
<span><span class="co">#&gt; 33775: TTTTTTCACGACAATAGCCAAT PB21_C3     1</span></span>
<span><span class="co">#&gt; 33776: TTTTTTCACTTGCACATCCGGC PB21_C3     1</span></span>
<span></span>
<span><span class="co"># Descriptive stats ----</span></span>
<span><span class="va">desc_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sample_statistics.html">sample_statistics</a></span><span class="op">(</span><span class="va">matrix</span>, <span class="va">af</span>,</span>
<span>    sample_key <span class="op">=</span> <span class="fu"><a href="../reference/pcr_id_column.html">pcr_id_column</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    value_columns <span class="op">=</span> <span class="st">"Value"</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>distinct_barcodes <span class="op">=</span> <span class="st">"nIS"</span><span class="op">)</span></span>
<span><span class="va">desc_stats</span></span>
<span><span class="co">#&gt;        ProjectID SubjectID Tissue TimePoint CellMarker      ID</span></span>
<span><span class="co">#&gt;  1: PMID32601433        A0     BM        21      Whole   BM_A0</span></span>
<span><span class="co">#&gt;  2: PMID32601433        A0     PB        21      Whole PB21_A0</span></span>
<span><span class="co">#&gt;  3: PMID32601433        A1     BM        21      Whole   BM_A1</span></span>
<span><span class="co">#&gt;  4: PMID32601433        A1     PB        21      Whole PB21_A1</span></span>
<span><span class="co">#&gt;  5: PMID32601433        A2     PB        21      Whole PB21_A2</span></span>
<span><span class="co">#&gt;  6: PMID32601433        A3     PB        21      Whole PB21_A3</span></span>
<span><span class="co">#&gt;  7: PMID32601433        A4     BM        21      Whole   BM_A4</span></span>
<span><span class="co">#&gt;  8: PMID32601433        A4     PB        21      Whole PB21_A4</span></span>
<span><span class="co">#&gt;  9: PMID32601433        C0     PB        21      Whole PB21_C0</span></span>
<span><span class="co">#&gt; 10: PMID32601433        C1     BM        21      Whole   BM_C1</span></span>
<span><span class="co">#&gt; 11: PMID32601433        C1     PB        21      Whole PB21_C1</span></span>
<span><span class="co">#&gt; 12: PMID32601433        C2     BM        21      Whole   BM_C2</span></span>
<span><span class="co">#&gt; 13: PMID32601433        C2     PB        21      Whole PB21_C2</span></span>
<span><span class="co">#&gt; 14: PMID32601433        C3     BM        21      Whole   BM_C3</span></span>
<span><span class="co">#&gt; 15: PMID32601433        C3     PB        21      Whole PB21_C3</span></span>
<span><span class="co">#&gt;                  SourceFileName</span></span>
<span><span class="co">#&gt;  1: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  2: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  3: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  4: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  5: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  6: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  7: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  8: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;  9: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 10: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 11: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 12: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 13: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 14: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt; 15: GSE144340_Matrix_542.tsv.gz</span></span>
<span><span class="co">#&gt;                                                                                                      Link</span></span>
<span><span class="co">#&gt;  1: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  2: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  3: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  4: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  5: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  6: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  7: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  8: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;  9: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 10: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 11: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 12: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 13: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 14: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt; 15: https://ftp.ncbi.nlm.nih.gov/geo/series/GSE144nnn/GSE144340/suppl/GSE144340%5FMatrix%5F542%2Etsv%2Egz</span></span>
<span><span class="co">#&gt;     TimepointMonths TimepointYears Value_shannon Value_simpson Value_invsimpson</span></span>
<span><span class="co">#&gt;  1:              01             01      2.952237     0.9113968        11.286277</span></span>
<span><span class="co">#&gt;  2:              01             01      3.459660     0.9488818        19.562506</span></span>
<span><span class="co">#&gt;  3:              01             01      3.006200     0.8870954         8.857038</span></span>
<span><span class="co">#&gt;  4:              01             01      3.774526     0.9453922        18.312391</span></span>
<span><span class="co">#&gt;  5:              01             01      3.181671     0.9222549        12.862544</span></span>
<span><span class="co">#&gt;  6:              01             01      3.389893     0.9401115        16.697684</span></span>
<span><span class="co">#&gt;  7:              01             01      2.820483     0.8857002         8.748925</span></span>
<span><span class="co">#&gt;  8:              01             01      3.345492     0.9344783        15.262125</span></span>
<span><span class="co">#&gt;  9:              01             01      3.843335     0.9588449        24.298324</span></span>
<span><span class="co">#&gt; 10:              01             01      3.201240     0.8979416         9.798313</span></span>
<span><span class="co">#&gt; 11:              01             01      3.194270     0.8805337         8.370564</span></span>
<span><span class="co">#&gt; 12:              01             01      2.615324     0.8670521         7.521742</span></span>
<span><span class="co">#&gt; 13:              01             01      3.515454     0.9382745        16.200749</span></span>
<span><span class="co">#&gt; 14:              01             01      2.557280     0.8453968         6.468172</span></span>
<span><span class="co">#&gt; 15:              01             01      3.929186     0.9397792        16.605561</span></span>
<span><span class="co">#&gt;     Value_sum Value_count Value_describe_vars Value_describe_n</span></span>
<span><span class="co">#&gt;  1:    244879        2284                   1             2284</span></span>
<span><span class="co">#&gt;  2:     81588        1080                   1             1080</span></span>
<span><span class="co">#&gt;  3:    274792        2477                   1             2477</span></span>
<span><span class="co">#&gt;  4:    104195        2269                   1             2269</span></span>
<span><span class="co">#&gt;  5:    124676        1465                   1             1465</span></span>
<span><span class="co">#&gt;  6:    180497        1786                   1             1786</span></span>
<span><span class="co">#&gt;  7:    296246        2255                   1             2255</span></span>
<span><span class="co">#&gt;  8:    177010        1538                   1             1538</span></span>
<span><span class="co">#&gt;  9:     59966        2644                   1             2644</span></span>
<span><span class="co">#&gt; 10:    303345        2993                   1             2993</span></span>
<span><span class="co">#&gt; 11:     95971        2636                   1             2636</span></span>
<span><span class="co">#&gt; 12:    343490        2223                   1             2223</span></span>
<span><span class="co">#&gt; 13:    149100        2386                   1             2386</span></span>
<span><span class="co">#&gt; 14:    277048        1817                   1             1817</span></span>
<span><span class="co">#&gt; 15:     64118        3923                   1             3923</span></span>
<span><span class="co">#&gt;     Value_describe_mean Value_describe_sd Value_describe_median</span></span>
<span><span class="co">#&gt;  1:           107.21497         1521.7649                     1</span></span>
<span><span class="co">#&gt;  2:            75.54444          556.4600                     1</span></span>
<span><span class="co">#&gt;  3:           110.93742         1852.2801                     1</span></span>
<span><span class="co">#&gt;  4:            45.92111          509.2056                     1</span></span>
<span><span class="co">#&gt;  5:            85.10307          904.5528                     1</span></span>
<span><span class="co">#&gt;  6:           101.06215         1040.5982                     1</span></span>
<span><span class="co">#&gt;  7:           131.37295         2105.4943                     1</span></span>
<span><span class="co">#&gt;  8:           115.09103         1149.9733                     1</span></span>
<span><span class="co">#&gt;  9:            22.68003          235.5394                     1</span></span>
<span><span class="co">#&gt; 10:           101.35149         1768.7577                     1</span></span>
<span><span class="co">#&gt; 11:            36.40781          645.1812                     1</span></span>
<span><span class="co">#&gt; 12:           154.51642         2652.4479                     1</span></span>
<span><span class="co">#&gt; 13:            62.48952          755.9385                     1</span></span>
<span><span class="co">#&gt; 14:           152.47551         2551.7136                     1</span></span>
<span><span class="co">#&gt; 15:            16.34412          250.7138                     1</span></span>
<span><span class="co">#&gt;     Value_describe_trimmed Value_describe_mad Value_describe_min</span></span>
<span><span class="co">#&gt;  1:               1.160284                  0                  1</span></span>
<span><span class="co">#&gt;  2:               1.082176                  0                  1</span></span>
<span><span class="co">#&gt;  3:               1.234493                  0                  1</span></span>
<span><span class="co">#&gt;  4:               1.068795                  0                  1</span></span>
<span><span class="co">#&gt;  5:               1.083546                  0                  1</span></span>
<span><span class="co">#&gt;  6:               1.080420                  0                  1</span></span>
<span><span class="co">#&gt;  7:               1.248199                  0                  1</span></span>
<span><span class="co">#&gt;  8:               1.065747                  0                  1</span></span>
<span><span class="co">#&gt;  9:               1.086484                  0                  1</span></span>
<span><span class="co">#&gt; 10:               1.208768                  0                  1</span></span>
<span><span class="co">#&gt; 11:               1.136019                  0                  1</span></span>
<span><span class="co">#&gt; 12:               1.295110                  0                  1</span></span>
<span><span class="co">#&gt; 13:               1.115707                  0                  1</span></span>
<span><span class="co">#&gt; 14:               1.252234                  0                  1</span></span>
<span><span class="co">#&gt; 15:               1.083147                  0                  1</span></span>
<span><span class="co">#&gt;     Value_describe_max Value_describe_range Value_describe_skew</span></span>
<span><span class="co">#&gt;  1:              48411                48410            21.84543</span></span>
<span><span class="co">#&gt;  2:              10411                10410            11.49921</span></span>
<span><span class="co">#&gt;  3:              68661                68660            29.32449</span></span>
<span><span class="co">#&gt;  4:              18252                18251            24.92426</span></span>
<span><span class="co">#&gt;  5:              19865                19864            17.48608</span></span>
<span><span class="co">#&gt;  6:              26055                26054            16.62791</span></span>
<span><span class="co">#&gt;  7:              70946                70945            25.02978</span></span>
<span><span class="co">#&gt;  8:              32536                32535            19.00613</span></span>
<span><span class="co">#&gt;  9:               5379                 5378            15.44781</span></span>
<span><span class="co">#&gt; 10:              80197                80196            35.45585</span></span>
<span><span class="co">#&gt; 11:              28170                28169            35.63449</span></span>
<span><span class="co">#&gt; 12:              81381                81380            23.99526</span></span>
<span><span class="co">#&gt; 13:              27428                27427            24.99102</span></span>
<span><span class="co">#&gt; 14:              83741                83740            26.01411</span></span>
<span><span class="co">#&gt; 15:              10463                10462            29.78536</span></span>
<span><span class="co">#&gt;     Value_describe_kurtosis Value_describe_se distinct_barcodes</span></span>
<span><span class="co">#&gt;  1:                564.8589         31.841939              2284</span></span>
<span><span class="co">#&gt;  2:                159.6777         16.932540              1080</span></span>
<span><span class="co">#&gt;  3:                969.0394         37.217196              2477</span></span>
<span><span class="co">#&gt;  4:                788.0556         10.689956              2269</span></span>
<span><span class="co">#&gt;  5:                342.6372         23.632797              1465</span></span>
<span><span class="co">#&gt;  6:                330.3889         24.623077              1786</span></span>
<span><span class="co">#&gt;  7:                716.3839         44.338480              2255</span></span>
<span><span class="co">#&gt;  8:                457.8748         29.323081              1538</span></span>
<span><span class="co">#&gt;  9:                276.7000          4.580710              2644</span></span>
<span><span class="co">#&gt; 10:               1478.2344         32.330691              2993</span></span>
<span><span class="co">#&gt; 11:               1447.9758         12.566345              2636</span></span>
<span><span class="co">#&gt; 12:                625.3701         56.257072              2223</span></span>
<span><span class="co">#&gt; 13:                793.6929         15.475733              2386</span></span>
<span><span class="co">#&gt; 14:                757.2388         59.862447              1817</span></span>
<span><span class="co">#&gt; 15:               1045.4960          4.002848              3923</span></span>
<span></span>
<span><span class="co"># Aggregation and new stats ----</span></span>
<span><span class="va">agg_key</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SubjectID"</span><span class="op">)</span></span>
<span><span class="va">agg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span><span class="va">matrix</span>, <span class="va">af</span>,</span>
<span>    key <span class="op">=</span> <span class="va">agg_key</span>,</span>
<span>    group <span class="op">=</span> <span class="st">"BarcodeSeq"</span>,</span>
<span>    join_af_by <span class="op">=</span> <span class="fu"><a href="../reference/pcr_id_column.html">pcr_id_column</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">agg</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 33,267 × 3</span></span></span>
<span><span class="co">#&gt;    BarcodeSeq             SubjectID Value_sum</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AAAAAAAACACGGAGAACGACG C3                2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AAAAAAAACGCGAACAACTACG C3                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AAAAAAAACTCAAAAAAGAAAT C3                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AAAAAAAATTTACACAAAGAAA A4                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AAAAAAAATTTTTAAACGTACC A0                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AAAAAACATATCTATAGTTACC A0                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AAAAAAGACGACGATAGGCACG C1                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AAAAAAGACGTTTATAGGTGTA A2                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AAAAAAGACTGCGACAAAAGGG A4                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AAAAAAGACTTTGATAACCACG C3                1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 33,257 more rows</span></span></span>
<span></span>
<span><span class="va">agg_meta_functions</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span><span class="va">Column</span>, <span class="op">~</span><span class="va">Function</span>, <span class="op">~</span><span class="va">Args</span>, <span class="op">~</span><span class="va">Output_colname</span>,</span>
<span>    <span class="st">"TimePoint"</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">.x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="cn">NA</span>, <span class="st">"{.col}_avg"</span>,</span>
<span>    <span class="st">"CellMarker"</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span>, <span class="cn">NA</span>, <span class="st">"distinct_cell_marker_count"</span>,</span>
<span>    <span class="st">"ID"</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span>, <span class="cn">NA</span>, <span class="st">"distinct_id_count"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">agg_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_metadata.html">aggregate_metadata</a></span><span class="op">(</span></span>
<span>    <span class="va">af</span>,</span>
<span>    aggregating_functions <span class="op">=</span> <span class="va">agg_meta_functions</span>,</span>
<span>    grouping_keys <span class="op">=</span> <span class="va">agg_key</span></span>
<span><span class="op">)</span></span>
<span><span class="va">agg_meta</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 4</span></span></span>
<span><span class="co">#&gt;   SubjectID TimePoint_avg distinct_cell_marker_count distinct_id_count</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> A0                   21                          1                 2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> A1                   21                          1                 2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> A2                   21                          1                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> A3                   21                          1                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> A4                   21                          1                 2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> C0                   21                          1                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> C1                   21                          1                 2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> C2                   21                          1                 2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> C3                   21                          1                 2</span></span>
<span></span>
<span><span class="va">agg_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sample_statistics.html">sample_statistics</a></span><span class="op">(</span><span class="va">agg</span>, <span class="va">agg_meta</span>,</span>
<span>    sample_key <span class="op">=</span> <span class="va">agg_key</span>,</span>
<span>    value_columns <span class="op">=</span> <span class="st">"Value_sum"</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>distinct_barcodes <span class="op">=</span> <span class="st">"nIS"</span><span class="op">)</span></span>
<span><span class="va">agg_stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 23</span></span></span>
<span><span class="co">#&gt;   SubjectID TimePoint_…¹ disti…² disti…³ Value…⁴ Value…⁵ Value…⁶ Value…⁷ Value…⁸</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> A0                  21       1       2    3.24   0.929   14.1   <span style="text-decoration: underline;">326</span>467    <span style="text-decoration: underline;">3</span>304</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> A1                  21       1       2    3.47   0.927   13.7   <span style="text-decoration: underline;">378</span>987    <span style="text-decoration: underline;">4</span>631</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> A2                  21       1       1    3.18   0.922   12.9   <span style="text-decoration: underline;">124</span>676    <span style="text-decoration: underline;">1</span>465</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> A3                  21       1       1    3.39   0.940   16.7   <span style="text-decoration: underline;">180</span>497    <span style="text-decoration: underline;">1</span>786</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> A4                  21       1       2    3.29   0.930   14.3   <span style="text-decoration: underline;">473</span>256    <span style="text-decoration: underline;">3</span>718</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> C0                  21       1       1    3.84   0.959   24.3    <span style="text-decoration: underline;">59</span>966    <span style="text-decoration: underline;">2</span>644</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> C1                  21       1       2    3.39   0.920   12.6   <span style="text-decoration: underline;">399</span>316    <span style="text-decoration: underline;">5</span>538</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> C2                  21       1       2    3.05   0.903   10.3   <span style="text-decoration: underline;">492</span>590    <span style="text-decoration: underline;">4</span>526</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> C3                  21       1       2    3.00   0.886    8.77  <span style="text-decoration: underline;">341</span>166    <span style="text-decoration: underline;">5</span>655</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 14 more variables: Value_sum_describe_vars &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Value_sum_describe_n &lt;dbl&gt;, Value_sum_describe_mean &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Value_sum_describe_sd &lt;dbl&gt;, Value_sum_describe_median &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Value_sum_describe_trimmed &lt;dbl&gt;, Value_sum_describe_mad &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Value_sum_describe_min &lt;dbl&gt;, Value_sum_describe_max &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Value_sum_describe_range &lt;dbl&gt;, Value_sum_describe_skew &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Value_sum_describe_kurtosis &lt;dbl&gt;, Value_sum_describe_se &lt;dbl&gt;, …</span></span></span>
<span></span>
<span><span class="co"># Abundance ----</span></span>
<span><span class="va">abundance</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_abundance.html">compute_abundance</a></span><span class="op">(</span><span class="va">agg</span>, columns <span class="op">=</span> <span class="st">"Value_sum"</span>, key <span class="op">=</span> <span class="va">agg_key</span><span class="op">)</span></span>
<span><span class="va">abundance</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 33,267 × 5</span></span></span>
<span><span class="co">#&gt;    BarcodeSeq             SubjectID Value_sum Value_sum_RelAbundance Value_sum…¹</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AAAAAAAACACGGAGAACGACG C3                2             0.000<span style="text-decoration: underline;">005</span>86    0.000<span style="text-decoration: underline;">586</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AAAAAAAACGCGAACAACTACG C3                1             0.000<span style="text-decoration: underline;">002</span>93    0.000<span style="text-decoration: underline;">293</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AAAAAAAACTCAAAAAAGAAAT C3                1             0.000<span style="text-decoration: underline;">002</span>93    0.000<span style="text-decoration: underline;">293</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AAAAAAAATTTACACAAAGAAA A4                1             0.000<span style="text-decoration: underline;">002</span>11    0.000<span style="text-decoration: underline;">211</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AAAAAAAATTTTTAAACGTACC A0                1             0.000<span style="text-decoration: underline;">003</span>06    0.000<span style="text-decoration: underline;">306</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AAAAAACATATCTATAGTTACC A0                1             0.000<span style="text-decoration: underline;">003</span>06    0.000<span style="text-decoration: underline;">306</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AAAAAAGACGACGATAGGCACG C1                1             0.000<span style="text-decoration: underline;">002</span>50    0.000<span style="text-decoration: underline;">250</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AAAAAAGACGTTTATAGGTGTA A2                1             0.000<span style="text-decoration: underline;">008</span>02    0.000<span style="text-decoration: underline;">802</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AAAAAAGACTGCGACAAAAGGG A4                1             0.000<span style="text-decoration: underline;">002</span>11    0.000<span style="text-decoration: underline;">211</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AAAAAAGACTTTGATAACCACG C3                1             0.000<span style="text-decoration: underline;">002</span>93    0.000<span style="text-decoration: underline;">293</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 33,257 more rows, and abbreviated variable name</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ¹​Value_sum_PercAbundance</span></span></span>
<span></span>
<span><span class="fu"><a href="../reference/reset_mandatory_IS_vars.html">reset_dyn_vars_config</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Mandatory IS vars reset to default</span></span>
<span><span class="co">#&gt; Annotation IS vars reset to default</span></span>
<span><span class="co">#&gt; Association file columns specs reset to default</span></span>
<span><span class="co">#&gt; ISS stats specs reset to default</span></span>
<span><span class="co">#&gt; Matrix suffixes specs reset to default</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="using-the-shiny-interface">Using the Shiny interface<a class="anchor" aria-label="anchor" href="#using-the-shiny-interface"></a>
</h2>
<p>The package provides a simple Shiny interface for data exploration
and plotting. To start the interface use:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NGSdataExplorer.html">NGSdataExplorer</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The application main page will show a loading screen for a file. It
is possible to load files also from the R environment, for example,
before opening the app, we can load the included association file:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"association_file"</span><span class="op">)</span></span></code></pre></div>
<p>Once in the application we can choose <code>"association_file"</code>
from the R environment loading option screen and click on “Import data”.
Once data is imported, we can click on the “Explore” tab in the upper
navbar: here we will see 2 tabs, one allows interactive exploration of
data in tabular form, in the other tab we can plot data. It is possible
to customize several different parameters for the plot and finally save
it to file with the dedicated button at the end of the page.</p>
<p>The Shiny interface is still currently under active development and
new features will be added in the near future.</p>
</div>
<div class="section level2">
<h2 id="ensuring-reproducibility-of-results">Ensuring reproducibility of results<a class="anchor" aria-label="anchor" href="#ensuring-reproducibility-of-results"></a>
</h2>
<p>Several implemented functions produce static HTML reports that can be
saved on disk, or tabular files. Reports contain the relevant
information on how the function was called, inputs and outputs
statistics, and session info for reproducibility.</p>
</div>
<div class="section level2">
<h2 id="browse-documentation-online-and-keep-updated">Browse documentation online and keep updated<a class="anchor" aria-label="anchor" href="#browse-documentation-online-and-keep-updated"></a>
</h2>
<p>ISAnalytics has it’s dedicated package website where you can browse
the documentation and vignettes easily, in addition to keeping up to
date with all relevant updates. Visit the website at <a href="https://calabrialab.github.io/ISAnalytics/" class="uri">https://calabrialab.github.io/ISAnalytics/</a></p>
</div>
<div class="section level2">
<h2 id="problems">Problems?<a class="anchor" aria-label="anchor" href="#problems"></a>
</h2>
<p>If you have any issues the documentation can’t solve, get in touch by
opening an issue on GitHub or contacting the maintainers</p>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p>
<a id="bib-ferrarisamueleefficient"></a><a href="#cite-ferrarisamueleefficient">[1]</a><cite>
B. S. Ferrari Samuele Jacob Aurelien. “Efficient gene editing of human
long-term hematopoietic stem cells validated by clonal tracking”. In:
<em>Nat Biotechnol 38, 1298–1308</em> (Nov. 2020). DOI:
<a href="https://doi.org/https://doi.org/10.1038/s41587-020-0551-y" class="external-link">https://doi.org/10.1038/s41587-020-0551-y</a>.</cite>
</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
