<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using ISAnalytics without RStudio support • ISAnalytics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/slate/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using ISAnalytics without RStudio support">
<meta property="og:description" content="ISAnalytics">
<meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISAnalytics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/aggregate_function_usage.html">Working with aggregate functions</a>
    </li>
    <li>
      <a href="../articles/collision_removal.html">Collision removal functionality</a>
    </li>
    <li>
      <a href="../articles/how_to_import_functions.html">How to use import functions</a>
    </li>
    <li>
      <a href="../articles/no_rstudio_usage.html">Using ISAnalytics without RStudio support</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="no_rstudio_usage_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using ISAnalytics without RStudio support</h1>
                        <h4 class="author">Giulia Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget, Via Olgettina 60, 20132 Milano - Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 class="date">17 February 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/master/vignettes/no_rstudio_usage.Rmd"><code>vignettes/no_rstudio_usage.Rmd</code></a></small>
      <div class="hidden name"><code>no_rstudio_usage.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>In this vignette we offer support for users that wish to use our package outside of RStudio and optionally in remote servers.</p>
<div id="how-to-install-isanalytics" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-install-isanalytics" class="anchor"></a>How to install ISAnalytics</h2>
<p>To install the package run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## For release version</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span>

<span class="co">## For devel version</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="co"># The following initializes usage of Bioc devel</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ISAnalytics"</span><span class="op">)</span></code></pre></div>
<p>To install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># For release version</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">devtools</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"RELEASE_3_12"</span>,
                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,
                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co">## Safer option for vignette building issue</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"RELEASE_3_12"</span><span class="op">)</span>

<span class="co"># For devel version</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">devtools</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"master"</span>,
                         dependencies <span class="op">=</span> <span class="cn">TRUE</span>,
                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co">## Safer option for vignette building issue</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"calabrialab/ISAnalytics"</span>,
                         ref <span class="op">=</span> <span class="st">"master"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://calabrialab.github.io/ISAnalytics">ISAnalytics</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span></code></pre></div>
</div>
<div id="setting-options" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-options" class="anchor"></a>Setting options</h2>
<p><code>ISAnalytics</code> has a verbose option that allows some functions to print additional information to the console while they’re executing. To disable this feature do:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DISABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># ENABLE</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.verbose"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Some functions also produce report in a user-friendly HTML format, to set this feature:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DISABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.widgets"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># ENABLE HTML REPORTS</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span><span class="st">"ISAnalytics.widgets"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="recommendations" class="section level1">
<h1 class="hasAnchor">
<a href="#recommendations" class="anchor"></a>Recommendations</h1>
<p>We highly encourage the use of widgets because some of the functions included in this package produce extensive and detailed reports that are made easy to explore via HTML format: we are talking, usually, about potentially very large data frames which are highly impractical to print on console.</p>
<p>Since we are aware that not all users can or want to use RStudio, here we present a few proposed solutions to maintain widget functionality.</p>
<div id="running-the-package-on-a-client-machine" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-package-on-a-client-machine" class="anchor"></a>Running the package on a client machine</h2>
<p>If code is run on the user machine but from R gui or console, user should not encounter problems, since widgets will be visualized in your preferred browser. If that’s not the case try:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>browser <span class="op">=</span> <span class="st">"firefox"</span><span class="op">)</span></code></pre></div>
<p>of course “firefox” can be replaced with whatever is needed.</p>
</div>
<div id="running-code-on-a-remote-machine-via-vpn" class="section level2">
<h2 class="hasAnchor">
<a href="#running-code-on-a-remote-machine-via-vpn" class="anchor"></a>Running code on a remote machine via VPN</h2>
<p>Users can visualize widgets even if they’re running code on a remote machine via VPN with the use of X11.</p>
<p>Provided the server machine has a browser installed (for example firefox), you can enable X11 via ssh by doing:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">ssh</span> YOUR_USERNAME@server_address -X -Y</span></code></pre></div>
<p>On the client side, users must have an X11 client such as XQuartz to allow X11 forwarding.</p>
<p>If users are working from a Windows machine the procedure is a bit more complicated, since the operating system is not Unix-like.</p>
<ol style="list-style-type: decimal">
<li>Download Xming here: <a href="http://www.straightrunning.com/XmingNotes/" class="uri">http://www.straightrunning.com/XmingNotes/</a> and puTTY here: <a href="https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html" class="uri">https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html</a>
</li>
<li>Once installed, run Xming and puTTY</li>
<li>In puTTY configurations, on the left side, find “SSH” -&gt; “X11”</li>
<li>Check the option “Enable X11 forwarding”</li>
<li>Back in the “Session” panel enter the IP address and as connection type select “SSH”</li>
<li>Finally open a connection and try opening the browser from console</li>
</ol>
<p>Once connected, open R and set the browser option, then load ISAnalytics:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>browser <span class="op">=</span> <span class="st">"firefox"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://calabrialab.github.io/ISAnalytics">ISAnalytics</a></span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>The <em><a href="https://bioconductor.org/packages/3.12/ISAnalytics">ISAnalytics</a></em> package <a id="cite-calabrialab_2021"></a>(<a href="http://www.bioconductor.org/packages/ISAnalytics">calabrialab, 2021</a>) was made possible thanks to:</p>
<ul>
<li>R <a id="cite-R_Core_Team_2021"></a>(<a href="https://www.R-project.org/">R Core Team, 2021</a>)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> <a id="cite-Oles_2020"></a>(<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> <a id="cite-Boettiger_2021"></a>(<a href="https://CRAN.R-project.org/package=knitcitations">Boettiger, 2021</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> <a id="cite-Xie_2021"></a>(<a href="https://yihui.org/knitr/">Xie, 2021</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> <a id="cite-Allaire_2020"></a>(<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=sessioninfo">sessioninfo</a></em> <a id="cite-Csardi_2018"></a>(<a href="https://CRAN.R-project.org/package=sessioninfo">Csárdi, core, Wickham, Chang, et al., 2018</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=testthat">testthat</a></em> <a id="cite-Wickham_2011"></a>(<a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">Wickham, 2011</a>)</li>
</ul>
<p>This package was developed using <em><a href="https://github.com/lcolladotor/biocthis">biocthis</a></em>.</p>
<p><code>R</code> session information.</p>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  setting  value                       
#&gt;  version  R version 4.0.4 (2021-02-15)
#&gt;  os       macOS Catalina 10.15.7      
#&gt;  system   x86_64, darwin17.0          
#&gt;  ui       X11                         
#&gt;  language (EN)                        
#&gt;  collate  en_US.UTF-8                 
#&gt;  ctype    en_US.UTF-8                 
#&gt;  tz       UTC                         
#&gt;  date     2021-02-17                  
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  package       * version date       lib source        
#&gt;  assertthat      0.2.1   2019-03-21 [1] CRAN (R 4.0.2)
#&gt;  BiocManager     1.30.10 2019-11-16 [1] CRAN (R 4.0.2)
#&gt;  BiocParallel    1.24.1  2020-11-06 [1] Bioconductor  
#&gt;  BiocStyle     * 2.18.1  2020-11-24 [1] Bioconductor  
#&gt;  bookdown        0.21    2020-10-13 [1] CRAN (R 4.0.2)
#&gt;  cachem          1.0.3   2021-02-04 [1] CRAN (R 4.0.2)
#&gt;  cli             2.3.0   2021-01-31 [1] CRAN (R 4.0.2)
#&gt;  colorspace      2.0-0   2020-11-11 [1] CRAN (R 4.0.2)
#&gt;  crayon          1.4.1   2021-02-08 [1] CRAN (R 4.0.2)
#&gt;  desc            1.2.0   2018-05-01 [1] CRAN (R 4.0.2)
#&gt;  digest          0.6.27  2020-10-24 [1] CRAN (R 4.0.2)
#&gt;  dplyr           1.0.4   2021-02-02 [1] CRAN (R 4.0.2)
#&gt;  ellipsis        0.3.1   2020-05-15 [1] CRAN (R 4.0.2)
#&gt;  evaluate        0.14    2019-05-28 [1] CRAN (R 4.0.1)
#&gt;  fastmap         1.1.0   2021-01-25 [1] CRAN (R 4.0.2)
#&gt;  forcats         0.5.1   2021-01-27 [1] CRAN (R 4.0.2)
#&gt;  fs              1.5.0   2020-07-31 [1] CRAN (R 4.0.2)
#&gt;  generics        0.1.0   2020-10-31 [1] CRAN (R 4.0.2)
#&gt;  ggplot2         3.3.3   2020-12-30 [1] CRAN (R 4.0.2)
#&gt;  ggrepel         0.9.1   2021-01-15 [1] CRAN (R 4.0.2)
#&gt;  glue            1.4.2   2020-08-27 [1] CRAN (R 4.0.2)
#&gt;  gtable          0.3.0   2019-03-25 [1] CRAN (R 4.0.2)
#&gt;  hms             1.0.0   2021-01-13 [1] CRAN (R 4.0.2)
#&gt;  htmltools       0.5.1.1 2021-01-22 [1] CRAN (R 4.0.2)
#&gt;  htmlwidgets     1.5.3   2020-12-10 [1] CRAN (R 4.0.2)
#&gt;  httr            1.4.2   2020-07-20 [1] CRAN (R 4.0.2)
#&gt;  ISAnalytics   * 1.0.9   2021-02-17 [1] Bioconductor  
#&gt;  jsonlite        1.7.2   2020-12-09 [1] CRAN (R 4.0.2)
#&gt;  knitcitations * 1.0.12  2021-01-10 [1] CRAN (R 4.0.2)
#&gt;  knitr           1.31    2021-01-27 [1] CRAN (R 4.0.2)
#&gt;  lattice         0.20-41 2020-04-02 [2] CRAN (R 4.0.4)
#&gt;  lifecycle       1.0.0   2021-02-15 [1] CRAN (R 4.0.3)
#&gt;  lubridate       1.7.9.2 2020-11-13 [1] CRAN (R 4.0.2)
#&gt;  magrittr      * 2.0.1   2020-11-17 [1] CRAN (R 4.0.2)
#&gt;  memoise         2.0.0   2021-01-26 [1] CRAN (R 4.0.2)
#&gt;  mnormt          2.0.2   2020-09-01 [1] CRAN (R 4.0.2)
#&gt;  munsell         0.5.0   2018-06-12 [1] CRAN (R 4.0.2)
#&gt;  nlme            3.1-152 2021-02-04 [2] CRAN (R 4.0.4)
#&gt;  pillar          1.4.7   2020-11-20 [1] CRAN (R 4.0.2)
#&gt;  pkgconfig       2.0.3   2019-09-22 [1] CRAN (R 4.0.2)
#&gt;  pkgdown         1.6.1   2020-09-12 [1] CRAN (R 4.0.2)
#&gt;  plyr            1.8.6   2020-03-03 [1] CRAN (R 4.0.2)
#&gt;  psych           2.0.12  2020-12-16 [1] CRAN (R 4.0.2)
#&gt;  purrr           0.3.4   2020-04-17 [1] CRAN (R 4.0.2)
#&gt;  R6              2.5.0   2020-10-28 [1] CRAN (R 4.0.2)
#&gt;  ragg            0.4.1   2021-01-11 [1] CRAN (R 4.0.2)
#&gt;  Rcpp            1.0.6   2021-01-15 [1] CRAN (R 4.0.2)
#&gt;  reactable       0.2.3   2020-10-04 [1] CRAN (R 4.0.2)
#&gt;  readr           1.4.0   2020-10-05 [1] CRAN (R 4.0.2)
#&gt;  RefManageR      1.3.0   2020-11-13 [1] CRAN (R 4.0.2)
#&gt;  rlang           0.4.10  2020-12-30 [1] CRAN (R 4.0.2)
#&gt;  rmarkdown       2.6     2020-12-14 [1] CRAN (R 4.0.2)
#&gt;  rprojroot       2.0.2   2020-11-15 [1] CRAN (R 4.0.2)
#&gt;  scales          1.1.1   2020-05-11 [1] CRAN (R 4.0.2)
#&gt;  sessioninfo   * 1.1.1   2018-11-05 [1] CRAN (R 4.0.2)
#&gt;  stringi         1.5.3   2020-09-09 [1] CRAN (R 4.0.2)
#&gt;  stringr         1.4.0   2019-02-10 [1] CRAN (R 4.0.2)
#&gt;  systemfonts     1.0.1   2021-02-09 [1] CRAN (R 4.0.2)
#&gt;  textshaping     0.3.0   2021-02-10 [1] CRAN (R 4.0.2)
#&gt;  tibble          3.0.6   2021-01-29 [1] CRAN (R 4.0.2)
#&gt;  tidyr           1.1.2   2020-08-27 [1] CRAN (R 4.0.2)
#&gt;  tidyselect      1.1.0   2020-05-11 [1] CRAN (R 4.0.2)
#&gt;  tmvnsim         1.0-2   2016-12-15 [1] CRAN (R 4.0.2)
#&gt;  upsetjs         1.9.0   2021-02-15 [1] CRAN (R 4.0.3)
#&gt;  vctrs           0.3.6   2020-12-17 [1] CRAN (R 4.0.2)
#&gt;  withr           2.4.1   2021-01-26 [1] CRAN (R 4.0.2)
#&gt;  xfun            0.21    2021-02-10 [1] CRAN (R 4.0.2)
#&gt;  xml2            1.3.2   2020-04-23 [1] CRAN (R 4.0.2)
#&gt;  yaml            2.2.1   2020-02-01 [1] CRAN (R 4.0.2)
#&gt;  zip             2.1.1   2020-08-27 [1] CRAN (R 4.0.2)
#&gt; 
#&gt; [1] /Users/runner/work/_temp/Library
#&gt; [2] /Library/Frameworks/R.framework/Versions/4.0/Resources/library</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> (<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (<a href="https://yihui.org/knitr/">Xie, 2021</a>) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> (<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> (<a href="https://CRAN.R-project.org/package=knitcitations">Boettiger, 2021</a>).</p>
<p>[1] J. Allaire, Y. Xie, J. McPherson, J. Luraschi, et al. <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.6. 2020. &lt;URL: <a href="https://github.com/rstudio/rmarkdown" class="uri">https://github.com/rstudio/rmarkdown</a>&gt;.</p>
<p>[2] C. Boettiger. <em>knitcitations: Citations for ‘Knitr’ Markdown Files</em>. R package version 1.0.12. 2021. &lt;URL: <a href="https://CRAN.R-project.org/package=knitcitations" class="uri">https://CRAN.R-project.org/package=knitcitations</a>&gt;.</p>
<p>[3] calabrialab. <em>Analyze gene therapy vector insertion sites data identified from genomics next generation sequencing reads for clonal tracking studies</em>. <a href="https://github.com/calabrialab/ISAnalytics" class="uri">https://github.com/calabrialab/ISAnalytics</a> - R package version 1.0.9. 2021. DOI: 10.18129/B9.bioc.ISAnalytics. &lt;URL: <a href="http://www.bioconductor.org/packages/ISAnalytics" class="uri">http://www.bioconductor.org/packages/ISAnalytics</a>&gt;.</p>
<p>[4] G. Csárdi, R. core, H. Wickham, W. Chang, et al. <em>sessioninfo: R Session Information</em>. R package version 1.1.1. 2018. &lt;URL: <a href="https://CRAN.R-project.org/package=sessioninfo" class="uri">https://CRAN.R-project.org/package=sessioninfo</a>&gt;.</p>
<p>[5] A. Oleś, M. Morgan, and W. Huber. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.18.1. 2020. &lt;URL: <a href="https://github.com/Bioconductor/BiocStyle" class="uri">https://github.com/Bioconductor/BiocStyle</a>&gt;.</p>
<p>[6] R Core Team. <em>R: A Language and Environment for Statistical Computing</em>. R Foundation for Statistical Computing. Vienna, Austria, 2021. &lt;URL: <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>&gt;.</p>
<p>[7] H. Wickham. “testthat: Get Started with Testing”. In: <em>The R Journal</em> 3 (2011), pp. 5-10. &lt;URL: <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf" class="uri">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>&gt;.</p>
<p>[8] Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.31. 2021. &lt;URL: <a href="https://yihui.org/knitr/" class="uri">https://yihui.org/knitr/</a>&gt;.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Andrea Calabria, Giulio Spinozzi, Giulia Pais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
