<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ISAnalytics">
<title>Setting up the workflow and first steps • ISAnalytics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Setting up the workflow and first steps">
<meta property="og:description" content="ISAnalytics">
<meta property="og:image" content="https://calabrialab.github.io/ISAnalytics/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ISAnalytics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.10.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/ISAnalytics.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/workflow_start.html">Setting up the workflow and first steps</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="https://calabrialab.github.io/ISAnalytics/">
    <span class="fas fa fas fa-code-branch"></span>
     
    RELEASE
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/calabrialab/ISAnalytics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="workflow_start_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="workflow_start_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="workflow_start_files/datatables-binding-0.27/datatables.js"></script><link href="workflow_start_files/dt-core-1.12.1/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="workflow_start_files/dt-core-1.12.1/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="workflow_start_files/dt-core-1.12.1/js/jquery.dataTables.min.js"></script><link href="workflow_start_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="workflow_start_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Setting up the workflow and first steps</h1>
                        <h4 data-toc-skip class="author">Giulia
Pais</h4>
            <address class="author_afil">
      <p>San Raffaele Telethon Institute for Gene Therapy - SR-Tiget,
Via Olgettina 60, 20132 Milano -
Italia</p>
<br><a class="author_email" href="mailto:#"></a><a href="mailto:giuliapais1@gmail.com" class="email">giuliapais1@gmail.com</a>, <a href="mailto:calabria.andrea@hsr.it" class="email">calabria.andrea@hsr.it</a>
      </address>
                  
            <h4 data-toc-skip class="date">9 May 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/calabrialab/ISAnalytics/blob/HEAD/vignettes/workflow_start.Rmd" class="external-link"><code>vignettes/workflow_start.Rmd</code></a></small>
      <div class="d-none name"><code>workflow_start.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>ISAnalytics is an R package developed to analyze gene therapy vector
insertion sites data identified from genomics next generation sequencing
reads for clonal tracking studies.</p>
<p>In this vignette we will explain how to properly setup the workflow
and the first steps of data import and data cleaning.</p>
</div>
<div class="section level2">
<h2 id="setting-up-your-workflow-with-dynamic-vars">Setting up your workflow with dynamic vars<a class="anchor" aria-label="anchor" href="#setting-up-your-workflow-with-dynamic-vars"></a>
</h2>
<p>This section demonstrates how to properly setup your workflow with
<code>ISAnalytics</code> using the “dynamic vars” system.</p>
<p>From <code>ISAnalytics 1.5.4</code> onwards, a new system here
referred to as “dynamic vars” has been implemented to improve the
flexibility of the package, by allowing multiple input formats based on
user needs rather than enforcing hard-coded names and structures. In
this way, users that do not follow the standard name conventions used by
the package have to put minimal effort into making their inputs
compliant to the package requirements.</p>
<p>There are 5 main categories of inputs you can customize:</p>
<ul>
<li>The <strong>“mandatory IS vars”</strong>: this set of variables is
used to uniquely identify integration events across several functions
implemented in the package</li>
<li>The <strong>“annotation IS vars”</strong>: this set of variables
holds the names of the columns that contain genomic annotations</li>
<li>The <strong>“association file columns”</strong>: this set contains
information on how metadata is structured</li>
<li>The <strong>“VISPA2 stats specs”</strong>: this set contains
information on the format of pool statistics files produced
automatically by VISPA2</li>
<li>The <strong>“matrix files suffixes”</strong>: this set contains all
default file names for each quantification type and it is used by
automated import functions</li>
</ul>
<div class="section level3">
<h3 id="general-approach">General approach<a class="anchor" aria-label="anchor" href="#general-approach"></a>
</h3>
<p>The general approach is based on the specification of predefined tags
and their associated information in the form of simple data frames with
a standard structure, namely:</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="12%">
<col width="32%">
<col width="12%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th>names</th>
<th>types</th>
<th>transform</th>
<th>flag</th>
<th>tag</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>&lt;name of the column&gt;</code></td>
<td><code>&lt;type&gt;</code></td>
<td><code>&lt;a lambda or NULL&gt;</code></td>
<td><code>&lt;flag&gt;</code></td>
<td><code>&lt;tag&gt;</code></td>
</tr></tbody>
</table>
<p>where</p>
<ul>
<li>
<code>names</code> contains the name of the column as a
character</li>
<li>
<code>types</code> contains the type of the column. Type should be
expressed as a string and should be in one of the allowed types
<ul>
<li>
<code>char</code> for character (strings)</li>
<li>
<code>int</code> for integers</li>
<li>
<code>logi</code> for logical values (TRUE / FALSE)</li>
<li>
<code>numeric</code> for numeric values</li>
<li>
<code>factor</code> for factors</li>
<li>
<code>date</code> for generic date format - note that functions that
need to read and parse files will try to guess the format and parsing
may fail</li>
<li>One of the accepted date/datetime formats by <em><a href="https://CRAN.R-project.org/package=lubridate" class="external-link">lubridate</a></em>,
you can use <code><a href="../reference/date_formats.html">ISAnalytics::date_formats()</a></code> to view the
accepted formats</li>
</ul>
</li>
<li>
<code>transform</code>: a purrr-style lambda that is applied
immediately after importing. This is useful to operate simple
transformations like removing unwanted characters or rounding to a
certain precision. Please note that these lambdas need to be functions
that accept a vector as input and only operate a <u>transformation</u>,
aka they output a vector of the same length as the input. For more
complicated applications that may require the value of other columns,
appropriate functions should be manually applied post-import.</li>
<li>
<code>flag</code>: as of now, it should be set either to
<code>required</code> or <code>optional</code> - some functions
internally check for only required tags presence and if those are
missing from inputs they fail, signaling failure to the user</li>
<li>
<code>tag</code>: a specific tag expressed as a string - see Section
@ref(tags)</li>
</ul>
<div class="figure">
<img src="../reference/figures/dyn_vars_general.png" alt=""><p class="caption">Dynamic variables general approach</p>
</div>
</div>
<div class="section level3">
<h3 id="tags">Tags<a class="anchor" aria-label="anchor" href="#tags"></a>
</h3>
<p>As already mentioned, certain functions included in the package
require the presence of specific tags (and associated column names) in
the input to work properly. You can always check what a tag means and
where it is used by using the function <code><a href="../reference/inspect_tags.html">inspect_tags()</a></code> and
providing in input the tags you want to check as a character vector.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/inspect_tags.html">inspect_tags</a></span><span class="op">(</span><span class="st">"chromosome"</span><span class="op">)</span></span>
<span><span class="co">#&gt; * TAG: chromosome</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Description: Number of the chromosome</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Functions that use it: top_targeted_genes, CIS_grubbs, compute_near_integrations</span></span></code></pre></div>
<p>You should make sure the tag matches the right information in your
inputs by looking at the description of the tag. It is also possible to
add entries that are not associated with any tag. Here is an overview of
the critical tags for each category.</p>
<div class="section level4">
<h4 id="mandatory-is-vars">Mandatory IS vars<a class="anchor" aria-label="anchor" href="#mandatory-is-vars"></a>
</h4>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-cd1ee0ac159edae3cc4d" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-cd1ee0ac159edae3cc4d">{"x":{"filter":"none","vertical":false,"data":[["chromosome","locus","is_strand"],["top_targeted_genes, CIS_grubbs, compute_near_integrations","top_targeted_genes, CIS_grubbs, compute_near_integrations","CIS_grubbs, compute_near_integrations"],["Number of the chromosome","The locus at which the integration occurs","The DNA strand in which the integration occurs"]],"container":"<table class=\"stripe\">\n  <thead>\n    <tr>\n      <th>tag<\/th>\n      <th>needed_in<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","columnDefs":[],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p>The presence of all mandatory IS vars is also checked and used in
other functions - for example, when importing matrices it is ensured
that all mandatory variables are present in the input, as declared in
the look up table. Some functions may require information that needs to
be specified as input, always check the documentation if you have
doubts.</p>
</div>
<div class="section level4">
<h4 id="annotation-is-vars">Annotation IS vars<a class="anchor" aria-label="anchor" href="#annotation-is-vars"></a>
</h4>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-3569a995b161560021c2" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-3569a995b161560021c2">{"x":{"filter":"none","vertical":false,"data":[["gene_symbol","gene_strand"],["top_targeted_genes, CIS_grubbs, compute_near_integrations, CIS_volcano_plot, gene_frequency_fisher","top_targeted_genes, CIS_grubbs"],["The symbol of the gene","The strand of the gene"]],"container":"<table class=\"stripe\">\n  <thead>\n    <tr>\n      <th>tag<\/th>\n      <th>needed_in<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","columnDefs":[],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p>Although genomic annotations are not necessarily required to work
with ISAnalytics, some operations do require annotation - if you’re
working with matrices that are not annotated you can either annotate
them with a tool of your choice or skip the steps that require
annotation.</p>
</div>
<div class="section level4">
<h4 id="association-file-columns">Association file columns<a class="anchor" aria-label="anchor" href="#association-file-columns"></a>
</h4>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-491eaeec1f264c2a7de6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-491eaeec1f264c2a7de6">{"x":{"filter":"none","vertical":false,"data":[["project_id","pool_id","fusion_id","tag_seq","subject","vector_id","tissue","tp_days","pcr_method","cell_marker","tag_id","ngs_tech","dna_num","pcr_replicate","vcn","vispa_concatenate","pcr_repl_id","proj_folder","genome"],["generate_default_folder_structure, import_Vispa2_stats, remove_collisions, generate_Vispa2_launch_AF, import_association_file, import_parallel_Vispa2Matrices","generate_Vispa2_launch_AF, remove_collisions, import_association_file","generate_Vispa2_launch_AF","generate_default_folder_structure, import_association_file, import_Vispa2_stats","import_association_file, HSC_population_size_estimate","generate_Vispa2_launch_AF","generate_Vispa2_launch_AF, import_association_file, HSC_population_size_estimate","generate_Vispa2_launch_AF, import_association_file","","generate_Vispa2_launch_AF, import_association_file, HSC_population_size_estimate","generate_Vispa2_launch_AF","","","import_association_file, remove_collisions","","import_association_file, generate_Vispa2_launch_AF, generate_default_folder_structure, import_Vispa2_stats, import_parallel_Vispa2Matrices","pcr_id_column, generate_Vispa2_launch_AF, import_association_file, import_Vispa2_stats","import_association_file",""],["Unique identifier of a project","Unique identifier of a sequencing pool","Identification code/number of the barcoded (SLiM-)PCR product included in the sequencing library","The barcode tag sequence","Unique identifier of a study subject (usually a patient)","Unique identifier of the vector used","The biological tissue the sample belongs to","The time point expressed in days","The PCR method used","Cell marker associated with isolated cells carrying the IS","Unique identifier of the barcode tag, as specified in VISPA2 requirements","Technology used for next generation sequencing","Identification code/number of the DNA extraction from a specific biological sample","Number of the PCR replicate","Vector copy number","Unique identifier of a pool as specified in VISPA2","Unique identifier of the pcr replicate, used as key to join data and metadata","Path on disk containing the standard VISPA2 folder structure of the project","The reference genome (e.g. 'hg19')"]],"container":"<table class=\"stripe\">\n  <thead>\n    <tr>\n      <th>tag<\/th>\n      <th>needed_in<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","columnDefs":[],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p>Some tags in this table are not associated with any function yet, but
they exist for potential new features that will be added in the
future.</p>
</div>
<div class="section level4">
<h4 id="vispa2-stats-specs">VISPA2 stats specs<a class="anchor" aria-label="anchor" href="#vispa2-stats-specs"></a>
</h4>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-f778d2053b8a58d176d3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-f778d2053b8a58d176d3">{"x":{"filter":"none","vertical":false,"data":[["vispa_concatenate","tag_seq","barcode_mux"],["import_association_file, generate_Vispa2_launch_AF, generate_default_folder_structure, import_Vispa2_stats, import_parallel_Vispa2Matrices","generate_default_folder_structure, import_association_file, import_Vispa2_stats",""],["Unique identifier of a pool as specified in VISPA2","The barcode tag sequence","The barcode demultiplexed reads"]],"container":"<table class=\"stripe\">\n  <thead>\n    <tr>\n      <th>tag<\/th>\n      <th>needed_in<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","columnDefs":[],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p>NOTE: VISPA2 stats files usually follow standard naming conventions.
If the pipeline launch was configured with default parameters, do not
change this lookup table.</p>
</div>
</div>
<div class="section level3">
<h3 id="customizing-dynamic-vars">Customizing dynamic vars<a class="anchor" aria-label="anchor" href="#customizing-dynamic-vars"></a>
</h3>
<p>For each category of dynamic vars there are 3 functions:</p>
<ul>
<li>A getter - returns the current lookup table</li>
<li>A setter - allows the user to change the current lookup table</li>
<li>A resetter - reverts all changes to defaults</li>
</ul>
<p>Setters will take in input the new variables, validate and eventually
change the lookup table. If validation fails an error will be thrown
instead, inviting the user to review the inputs. Moreover, if some of
the critical tags for the category are missing, a warning appears, with
a list of the missing ones.</p>
<p>Let’s take a look at some examples.</p>
<p>On package loading, all lookup tables are set to default values. For
example, for mandatory IS vars we have:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 5</span></span></span>
<span><span class="co">#&gt;   names             types transform flag     tag       </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr               char  <span style="color: #949494;">&lt;NULL&gt;</span>    required chromosome</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> integration_locus int   <span style="color: #949494;">&lt;NULL&gt;</span>    required locus     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> strand            char  <span style="color: #949494;">&lt;NULL&gt;</span>    required is_strand</span></span></code></pre></div>
<p>Let’s suppose our matrices follow a different standard, and
integration events are characterized by 5 fields, like so (the example
contains random data):</p>
<table class="table">
<thead><tr class="header">
<th>chrom</th>
<th>position</th>
<th>strand</th>
<th>gap</th>
<th>junction</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>“chr1”</td>
<td>342543</td>
<td>“+”</td>
<td>100</td>
<td>50</td>
</tr>
<tr class="even">
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
</tbody>
</table>
<p>To make this work with ISAnalytics functions, we need to compile the
lookup table like this:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_mand_vars</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span><span class="va">names</span>, <span class="op">~</span><span class="va">types</span>, <span class="op">~</span><span class="va">transform</span>, <span class="op">~</span><span class="va">flag</span>, <span class="op">~</span><span class="va">tag</span>,</span>
<span>    <span class="st">"chrom"</span>, <span class="st">"char"</span>, <span class="op">~</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="va">.x</span>, <span class="st">"chr"</span>, <span class="st">""</span><span class="op">)</span>, <span class="st">"required"</span>,</span>
<span>    <span class="st">"chromosome"</span>,</span>
<span>    <span class="st">"position"</span>, <span class="st">"int"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"locus"</span>,</span>
<span>    <span class="st">"strand"</span>, <span class="st">"char"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="st">"is_strand"</span>,</span>
<span>    <span class="st">"gap"</span>, <span class="st">"int"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="cn">NA_character_</span>,</span>
<span>    <span class="st">"junction"</span>, <span class="st">"int"</span>, <span class="cn">NULL</span>, <span class="st">"required"</span>, <span class="cn">NA_character_</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Notice that we have specified a transformation for the “chromosome”
tag: in this case we would like to have only the number of the
chromosome without the prefix “chr” - this lambda will get executed
immediately after import.</p>
<p>To set the new variables simply do:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/set_mandatory_IS_vars.html">set_mandatory_IS_vars</a></span><span class="op">(</span><span class="va">new_mand_vars</span><span class="op">)</span></span>
<span><span class="co">#&gt; Mandatory IS vars successfully changed</span></span>
<span><span class="fu"><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 5</span></span></span>
<span><span class="co">#&gt;   names    types transform flag     tag       </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chrom    char  <span style="color: #949494;">&lt;formula&gt;</span> required chromosome</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> position int   <span style="color: #949494;">&lt;NULL&gt;</span>    required locus     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> strand   char  <span style="color: #949494;">&lt;NULL&gt;</span>    required is_strand </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> gap      int   <span style="color: #949494;">&lt;NULL&gt;</span>    required <span style="color: #BB0000;">NA</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> junction int   <span style="color: #949494;">&lt;NULL&gt;</span>    required <span style="color: #BB0000;">NA</span></span></span></code></pre></div>
<p>If you don’t specify a critical tag, a warning message is
displayed:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_mand_vars</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">$</span><span class="va">tag</span> <span class="op">&lt;-</span> <span class="cn">NA_character_</span></span>
<span><span class="fu"><a href="../reference/set_mandatory_IS_vars.html">set_mandatory_IS_vars</a></span><span class="op">(</span><span class="va">new_mand_vars</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Warning: important tags missing</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Some tags are required for proper execution of some functions. If these tags are not provided, execution of dependent functions might fail. Review your inputs carefully.</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Missing tags: chromosome</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> To see where these are involved type `inspect_tags(c('chromosome'))`</span></span>
<span><span class="co">#&gt; Mandatory IS vars successfully changed</span></span>
<span><span class="fu"><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 5</span></span></span>
<span><span class="co">#&gt;   names    types transform flag     tag      </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chrom    char  <span style="color: #949494;">&lt;formula&gt;</span> required <span style="color: #BB0000;">NA</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> position int   <span style="color: #949494;">&lt;NULL&gt;</span>    required locus    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> strand   char  <span style="color: #949494;">&lt;NULL&gt;</span>    required is_strand</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> gap      int   <span style="color: #949494;">&lt;NULL&gt;</span>    required <span style="color: #BB0000;">NA</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> junction int   <span style="color: #949494;">&lt;NULL&gt;</span>    required <span style="color: #BB0000;">NA</span></span></span></code></pre></div>
<p>If you change your mind and want to go back to defaults:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reset_mandatory_IS_vars.html">reset_mandatory_IS_vars</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Mandatory IS vars reset to default</span></span>
<span><span class="fu"><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 5</span></span></span>
<span><span class="co">#&gt;   names             types transform flag     tag       </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr               char  <span style="color: #949494;">&lt;NULL&gt;</span>    required chromosome</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> integration_locus int   <span style="color: #949494;">&lt;NULL&gt;</span>    required locus     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> strand            char  <span style="color: #949494;">&lt;NULL&gt;</span>    required is_strand</span></span></code></pre></div>
<p>The principle is the same for annotation IS vars, association file
columns and VISPA2 stats specs. Here is a summary of the functions for
each:</p>
<ul>
<li>mandatory IS vars: <code><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars()</a></code>,
<code><a href="../reference/set_mandatory_IS_vars.html">set_mandatory_IS_vars()</a></code>,
<code><a href="../reference/reset_mandatory_IS_vars.html">reset_mandatory_IS_vars()</a></code>
</li>
<li>annotation IS vars: <code><a href="../reference/mandatory_IS_vars.html">annotation_IS_vars()</a></code>,
<code><a href="../reference/set_mandatory_IS_vars.html">set_annotation_IS_vars()</a></code>,
<code><a href="../reference/reset_mandatory_IS_vars.html">reset_annotation_IS_vars()</a></code>
</li>
<li>association file columns: <code><a href="../reference/mandatory_IS_vars.html">association_file_columns()</a></code>,
<code><a href="../reference/set_mandatory_IS_vars.html">set_af_columns_def()</a></code>,
<code><a href="../reference/reset_mandatory_IS_vars.html">reset_af_columns_def()</a></code>
</li>
<li>VISPA2 stats specs: <code><a href="../reference/mandatory_IS_vars.html">iss_stats_specs()</a></code>,
<code><a href="../reference/set_mandatory_IS_vars.html">set_iss_stats_specs()</a></code>,
<code>reset_iss_stats_specs</code>
</li>
</ul>
<p>Matrix files suffixes work slightly different:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mandatory_IS_vars.html">matrix_file_suffixes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 3</span></span></span>
<span><span class="co">#&gt;    quantification   matrix_type   file_suffix                                 </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> seqCount         annotated     seqCount_matrix.no0.annotated.tsv.gz        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> seqCount         not_annotated seqCount_matrix.tsv.gz                      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> fragmentEstimate annotated     fragmentEstimate_matrix.no0.annotated.tsv.gz</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> fragmentEstimate not_annotated fragmentEstimate_matrix.tsv.gz              </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> barcodeCount     annotated     barcodeCount_matrix.no0.annotated.tsv.gz    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> barcodeCount     not_annotated barcodeCount_matrix.tsv.gz                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> cellCount        annotated     cellCount_matrix.no0.annotated.tsv.gz       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> cellCount        not_annotated cellCount_matrix.tsv.gz                     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ShsCount         annotated     ShsCount_matrix.no0.annotated.tsv.gz        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ShsCount         not_annotated ShsCount_matrix.tsv.gz</span></span></code></pre></div>
<p>To change this lookup table use the function
<code><a href="../reference/set_matrix_file_suffixes.html">set_matrix_file_suffixes()</a></code>: the function will ask to
specify a suffix for each quantification and for both annotated and not
annotated versions. These suffixes are used in the automated matrix
import function when scanning the file system.</p>
<p>To reset all lookup tables to their default configurations you can
also use the function <code><a href="../reference/reset_mandatory_IS_vars.html">reset_dyn_vars_config()</a></code>, which
reverts all changes.</p>
</div>
<div class="section level3">
<h3 id="faqs">FAQs<a class="anchor" aria-label="anchor" href="#faqs"></a>
</h3>
<div class="section level4">
<h4 id="do-i-have-to-do-this-every-time-the-package-loads">Do I have to do this every time the package loads?<a class="anchor" aria-label="anchor" href="#do-i-have-to-do-this-every-time-the-package-loads"></a>
</h4>
<p>No, if you frequently have to work with a non-standard settings
profile, you can use the functions <code><a href="../reference/export_ISA_settings.html">export_ISA_settings()</a></code>
and <code><a href="../reference/import_ISA_settings.html">import_ISA_settings()</a></code>: these functions allow the
import/export of setting profiles in *.json format.</p>
<p>Once you set your variables for the first time through the procedure
described before, simply call the export function and all will be saved
to a json file, which can then be imported for the next workflow.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="reporting-progress">Reporting progress<a class="anchor" aria-label="anchor" href="#reporting-progress"></a>
</h2>
<p>From <code>ISAnalytics 1.7.4</code>, functions that make use of
parallel workers or process long tasks report progress via the functions
offered by <a href="https://progressr.futureverse.org/" class="external-link">progressr</a>.
To enable progress bars for all functions in ISAnalytics do</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/enable_progress_bars.html">enable_progress_bars</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>before calling other functions. For customizing the appearance of the
progress bar please refer to <code>progressr</code> documentation.</p>
</div>
<div class="section level2">
<h2 id="introduction-to-isanalytics-import-functions-family">Introduction to <code>ISAnalytics</code> import functions
family<a class="anchor" aria-label="anchor" href="#introduction-to-isanalytics-import-functions-family"></a>
</h2>
<p>In this section we’re going to explain more in detail how functions
of the import family should be used, the most common workflows to follow
and more.</p>
<div class="section level3">
<h3 id="designed-to-work-with-vispa2-pipeline">Designed to work with VISPA2 pipeline<a class="anchor" aria-label="anchor" href="#designed-to-work-with-vispa2-pipeline"></a>
</h3>
<p>The vast majority of the functions included in this package is
designed to work in combination with VISPA2 pipeline (Giulio Spinozzi
Andrea Calabria, 2017). If you don’t know what it is, we strongly
recommend you to take a look at these links:</p>
<ul>
<li>Article: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5702242/" class="external-link">VISPA2:
Article</a><br>
</li>
<li>BitBucket Wiki: <a href="https://bitbucket.org/andreacalabria/vispa2/wiki/Home" class="external-link">VISPA2
Wiki</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="file-system-structure-generated">File system structure generated<a class="anchor" aria-label="anchor" href="#file-system-structure-generated"></a>
</h3>
<p>VISPA2 produces a standard file system structure starting from a
folder you specify as your workbench or root. The structure always
follows this schema:</p>
<ul>
<li>root/
<ul>
<li>Optional intermediate folders
<ul>
<li>Projects (PROJECTID)
<ul>
<li>bam</li>
<li>bcmuxall</li>
<li>bed</li>
<li>iss
<ul>
<li>Pools (concatenatePoolIDSeqRun)</li>
</ul>
</li>
<li>quality</li>
<li>quantification
<ul>
<li>Pools (concatenatePoolIDSeqRun)</li>
</ul>
</li>
<li>report</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>Most of the functions implemented expect a standard file system
structure as the one described above.</p>
</div>
<div class="section level3">
<h3 id="notation">Notation<a class="anchor" aria-label="anchor" href="#notation"></a>
</h3>
<p>We call an <em>“integration matrix”</em> a tabular structure
characterized by:</p>
<ul>
<li>k mandatory columns of genomic features that characterize a viral
insertion site in the genome, which are specified via
<code><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars()</a></code>. By default they’re set to
<code>chr</code>, <code>integration_locus</code> and
<code>strand</code>
</li>
<li>a (optional) annotation columns, provided via
<code><a href="../reference/mandatory_IS_vars.html">annotation_IS_vars()</a></code>. By default they’re set to
<code>GeneName</code> and <code>GeneStrand</code>
</li>
<li>A variable number n of sample columns containing the quantification
of the corresponding integration site</li>
</ul>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 8</span></span></span>
<span><span class="co">#&gt;   chr   integration_locus strand GeneName     GeneStrand  exp1  exp2  exp3</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1                 <span style="text-decoration: underline;">12</span>324 +      NFATC3       +           <span style="text-decoration: underline;">4</span>553  <span style="text-decoration: underline;">5</span>345    <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 6                <span style="text-decoration: underline;">657</span>532 +      LOC100507487 +             76   545     5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 7                <span style="text-decoration: underline;">657</span>532 +      EDIL3        -             <span style="color: #BB0000;">NA</span>    56    <span style="color: #BB0000;">NA</span></span></span></code></pre>
<p>The package uses a more compact form of these matrices, limiting the
amount of NA values and optimizing time and memory consumption. For more
info on this take a look at: <a href="https://r4ds.had.co.nz/tidy-data.html" class="external-link">Tidy data</a></p>
<p>While integration matrices contain the actual data, we also need
associated sample metadata to perform the vast majority of the analyses.
<code>ISAnalytics</code> expects the metadata to be contained in a so
called <em>“association file”</em>, which is a simple tabular file.</p>
<p>To generate a blank association file you can use the function
<code>generate_blank_association_file</code>. You can also view the
standard column names with <code><a href="../reference/mandatory_IS_vars.html">association_file_columns()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="metadata">Importing metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>To import metadata we use <code><a href="../reference/import_association_file.html">import_association_file()</a></code>.
This function is not only responsible for reading the file into the R
environment as a data frame, but it is capable to perform a file system
alignment operation, that is, for each project and pool contained in the
file, it scans the file system starting from the provided root to check
if the corresponding folders (contained in the appropriate column) can
be found. Remember that to work properly, this operation expects a
standard folder structure, such as the one provided by VISPA2. This
function also produces an interactive HTML report.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fs_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_default_folder_structure.html">generate_default_folder_structure</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_options.html" class="external-link">with_options</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ISAnalytics.reports <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, code <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_association_file.html">import_association_file</a></span><span class="op">(</span><span class="va">fs_path</span><span class="op">$</span><span class="va">af</span>, root <span class="op">=</span> <span class="va">fs_path</span><span class="op">$</span><span class="va">root</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; *** Association file import summary ***</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> For detailed report please set option 'ISAnalytics.reports' to TRUE</span></span>
<span><span class="co">#&gt; Parsing problems detected: FALSE</span></span>
<span><span class="co">#&gt; Date parsing problems: FALSE</span></span>
<span><span class="co">#&gt; Column problems detected: FALSE</span></span>
<span><span class="co">#&gt; NAs found in important columns: FALSE</span></span>
<span><span class="co">#&gt; File system alignment: no problems detected</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 74</span></span></span>
<span><span class="co">#&gt;   ProjectID FUSIONID  PoolID TagSequence SubjectID VectorType VectorID</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> PJ01      ET#382.46 POOL01 LTR75LC38   PT001     lenti      GLOBE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> PJ01      ET#381.40 POOL01 LTR53LC32   PT001     lenti      GLOBE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> PJ01      ET#381.9  POOL01 LTR83LC66   PT001     lenti      GLOBE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> PJ01      ET#381.71 POOL01 LTR27LC94   PT001     lenti      GLOBE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> PJ01      ET#381.2  POOL01 LTR69LC52   PT001     lenti      GLOBE   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> PJ01      ET#382.28 POOL01 LTR37LC2    PT001     lenti      GLOBE   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 67 more variables: ExperimentID &lt;chr&gt;, Tissue &lt;chr&gt;, TimePoint &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   DNAFragmentation &lt;chr&gt;, PCRMethod &lt;chr&gt;, TagIDextended &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Keywords &lt;chr&gt;, CellMarker &lt;chr&gt;, TagID &lt;chr&gt;, NGSProvider &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   NGSTechnology &lt;chr&gt;, ConverrtedFilesDir &lt;chr&gt;, ConverrtedFilesName &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   SourceFileFolder &lt;chr&gt;, SourceFileNameR1 &lt;chr&gt;, SourceFileNameR2 &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   DNAnumber &lt;chr&gt;, ReplicateNumber &lt;int&gt;, DNAextractionDate &lt;date&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   DNAngUsed &lt;dbl&gt;, LinearPCRID &lt;chr&gt;, LinearPCRDate &lt;date&gt;, …</span></span></span></code></pre>
<div class="section level4">
<h4 id="function-arguments">Function arguments<a class="anchor" aria-label="anchor" href="#function-arguments"></a>
</h4>
<p>You can change several arguments in the function call to modify the
behavior of the function.</p>
<ul>
<li>
<code>root</code>
<ul>
<li>Set it to <code>NULL</code> if you only want to import the
association file without file system alignment. Beware that some of the
automated import functionalities won’t work!</li>
<li>Set it to a non-empty string (path on disk): in this case, the
column associated with the tag <code>proj_folder</code> (by default
<code>PathToFolderProjectID</code>) in the file should contain
<strong>relative</strong> file paths, so if for example your root is set
to “/home” and your project folder in the association file is set to
“/PJ01”, the function will check that the directory exists under
“/home/PJ01”</li>
<li>Set it to an empty string: ideal if you want to store paths in the
association file as <strong>absolute</strong> file paths. In this case
if your project folder is in “/home/PJ01” you should have this path in
the <code>PathToFolderProjectID</code> column and set <code>root</code>
= “”</li>
</ul>
</li>
<li>
<code>dates_format</code>: a string that is useful for properly
parsing dates from tabular formats</li>
<li>
<code>separator</code>: the column separator used in the file.
Defaults to “\t”, other valid separators are “,” (comma), “;”
(semi-colon)</li>
<li>
<code>filter_for</code>: you can set this argument to a
<strong>named</strong> list of filters, where names are column names.
For example <code>list(ProjectID = "PJ01")</code> will return only those
rows whose attribute “ProjectID” equals “PJ01”</li>
<li>
<code>import_iss</code>: either <code>TRUE</code> or
<code>FALSE</code>. If set to <code>TRUE</code>, performs an internal
call to <code><a href="../reference/import_Vispa2_stats.html">import_Vispa2_stats()</a></code> (see next section), and
appends the imported files to metadata</li>
<li>
<code>convert_tp</code>: either <code>TRUE</code> or
<code>FALSE</code>. Converts the column containing the time point
expressed in days in months and years (with custom logic).</li>
<li>
<code>report_path</code>
<ul>
<li>Set it to <code>NULL</code> to avoid the production of a report</li>
<li>Set it to a folder (if it doesn’t exist, it gets automatically
created)</li>
</ul>
</li>
<li>
<code>...</code>: additional named arguments to pass to
<code><a href="../reference/import_Vispa2_stats.html">import_Vispa2_stats()</a></code> if you chose to import VISPA2
stats</li>
</ul>
<p>For further details view the dedicated function documentation.</p>
<p><em>NOTE</em>: the function supports files in various formats as long
as the correct separator is provided. It also accepts files in
<code>*.xlsx</code> and <code>*.xls</code> formats but we do not
recommend using these since the report won’t include a detailed summary
of potential parsing problems.</p>
<p>The interactive report includes useful information such as</p>
<ul>
<li>General issues: parsing problems, missing columns, NA values in
important columns etc. This allows you to immediately spot problems and
correct them before proceeding with the analyses</li>
<li>File system alignment issues: very useful to know if all data can be
imported or folders are missing</li>
<li>Info on VISPA2 stats (if <code>import_iss</code> was
<code>TRUE</code>)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="importing-vispa2-stats-files">Importing VISPA2 stats files<a class="anchor" aria-label="anchor" href="#importing-vispa2-stats-files"></a>
</h3>
<p>VISPA2 automatically produces summary files for each pool holding
information that can be useful for other analyses downstream, so it is
recommended to import them in the first steps of the workflow. To do
that, you can use <code>import_VISPA2_stats</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vispa_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_Vispa2_stats.html">import_Vispa2_stats</a></span><span class="op">(</span></span>
<span>    association_file <span class="op">=</span> <span class="va">af</span>,</span>
<span>    join_with_af <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    report_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 14</span></span></span>
<span><span class="co">#&gt;   POOL     TAG       RUN_NAME     PHIX_MAPPING PLASMID_MAPPED_BYPOOL BARCODE_MUX</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> POOL01-1 LTR75LC38 PJ01|POOL01…     43<span style="text-decoration: underline;">586</span>699               2<span style="text-decoration: underline;">256</span>176      <span style="text-decoration: underline;">645</span>026</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> POOL01-1 LTR53LC32 PJ01|POOL01…     43<span style="text-decoration: underline;">586</span>699               2<span style="text-decoration: underline;">256</span>176      <span style="text-decoration: underline;">652</span>208</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> POOL01-1 LTR83LC66 PJ01|POOL01…     43<span style="text-decoration: underline;">586</span>699               2<span style="text-decoration: underline;">256</span>176      <span style="text-decoration: underline;">451</span>519</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> POOL01-1 LTR27LC94 PJ01|POOL01…     43<span style="text-decoration: underline;">586</span>699               2<span style="text-decoration: underline;">256</span>176      <span style="text-decoration: underline;">426</span>500</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> POOL01-1 LTR69LC52 PJ01|POOL01…     43<span style="text-decoration: underline;">586</span>699               2<span style="text-decoration: underline;">256</span>176       <span style="text-decoration: underline;">18</span>300</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> POOL01-1 LTR37LC2  PJ01|POOL01…     43<span style="text-decoration: underline;">586</span>699               2<span style="text-decoration: underline;">256</span>176      <span style="text-decoration: underline;">729</span>327</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 8 more variables: LTR_IDENTIFIED &lt;dbl&gt;, TRIMMING_FINAL_LTRLC &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   LV_MAPPED &lt;dbl&gt;, BWA_MAPPED_OVERALL &lt;dbl&gt;, ISS_MAPPED_OVERALL &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   RAW_READS &lt;lgl&gt;, QUALITY_PASSED &lt;lgl&gt;, ISS_MAPPED_PP &lt;lgl&gt;</span></span></span></code></pre>
<p>The function requires as input the imported and file system aligned
association file and it will scan the <code>iss</code> folder for files
that match some known prefixes (defaults are already provided but you
can change them as you see fit). You can either choose to join the
imported data frames with the association file in input and obtain a
single data frame or keep it as it is, just set the parameter
<code>join_with_af</code> accordingly. At the end of the process an HTML
report is produced, signaling potential problems.</p>
<p>You can directly call this function when you import the association
file by setting the <code>import_iss</code> argument of
<code>import_association_file</code> to <code>TRUE</code>.</p>
</div>
<div class="section level3">
<h3 id="importing-a-single-integration-matrix">Importing a single integration matrix<a class="anchor" aria-label="anchor" href="#importing-a-single-integration-matrix"></a>
</h3>
<p>If you want to import a single integration matrix you can do so by
using the <code><a href="../reference/import_single_Vispa2Matrix.html">import_single_Vispa2Matrix()</a></code> function. This
function reads the file and converts it into a tidy structure: several
different formats can be read, since you can specify the column
separator.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrix_path</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span></span>
<span>    <span class="va">fs_path</span><span class="op">$</span><span class="va">root</span>,</span>
<span>    <span class="st">"PJ01"</span>,</span>
<span>    <span class="st">"quantification"</span>,</span>
<span>    <span class="st">"POOL01-1"</span>,</span>
<span>    <span class="st">"PJ01_POOL01-1_seqCount_matrix.no0.annotated.tsv.gz"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_single_Vispa2Matrix.html">import_single_Vispa2Matrix</a></span><span class="op">(</span><span class="va">matrix_path</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 802 × 7</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand GeneName     GeneStrand CompleteAmplificatio…¹</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 16             68<span style="text-decoration: underline;">164</span>148 +      NFATC3       +          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 4             129<span style="text-decoration: underline;">390</span>130 +      LOC100507487 +          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 5              84<span style="text-decoration: underline;">009</span>671 -      EDIL3        -          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 12             54<span style="text-decoration: underline;">635</span>693 -      CBX5         -          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 2             181<span style="text-decoration: underline;">930</span>711 +      UBE2E3       +          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 20             35<span style="text-decoration: underline;">920</span>986 +      MANBAL       +          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 22             26<span style="text-decoration: underline;">900</span>625 +      TFIP11       -          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 3             106<span style="text-decoration: underline;">580</span>075 +      LINC00882    -          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              16<span style="text-decoration: underline;">186</span>297 -      SPEN         +          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 17             61<span style="text-decoration: underline;">712</span>419 +      MAP3K3       +          PJ01_POOL01_LTR75LC38…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 792 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated name: ¹​CompleteAmplificationID</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: Value &lt;int&gt;</span></span></span></code></pre>
<p>For details on usage and arguments view the dedicated function
documentation.</p>
</div>
<div class="section level3">
<h3 id="automated-integration-matrices-import">Automated integration matrices import<a class="anchor" aria-label="anchor" href="#automated-integration-matrices-import"></a>
</h3>
<p>Integration matrices import can be automated when when the
association file is imported with the file system alignment option.
<code>ISAnalytics</code> provides a function,
<code><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices()</a></code>, that allows to do just
that in a fast and efficient way.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_options.html" class="external-link">with_options</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ISAnalytics.reports <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="op">{</span></span>
<span>    <span class="va">matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices</a></span><span class="op">(</span><span class="va">af</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span>,</span>
<span>        mode <span class="op">=</span> <span class="st">"AUTO"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="function-arguments-1">Function arguments<a class="anchor" aria-label="anchor" href="#function-arguments-1"></a>
</h3>
<p>Let’s see how the behavior of the function changes when we change
arguments.</p>
<div class="section level4">
<h4 id="association_file-argument">
<code>association_file</code> argument<a class="anchor" aria-label="anchor" href="#association_file-argument"></a>
</h4>
<p>You can supply a data frame object, imported via
<code><a href="../reference/import_association_file.html">import_association_file()</a></code> (see Section @ref(metadata)) or a
string (the path to the association file on disk). In the first scenario
it is necessary to perform file system alignment, since the function
scans the folders contained in the column <code>Path_quant</code>, while
in the second case you should also provide as additional
<strong>named</strong> argument (to <code>...</code>) an appropriate
<code>root</code>: the function will internally call
<code><a href="../reference/import_association_file.html">import_association_file()</a></code>, if you don’t have specific needs
we recommend doing the 2 steps separately and provide the association
file as a data frame.</p>
</div>
<div class="section level4">
<h4 id="quantification_type-argument">
<code>quantification_type</code> argument<a class="anchor" aria-label="anchor" href="#quantification_type-argument"></a>
</h4>
<p>For each pool there may be multiple available quantification types,
that is, different matrices containing the same samples and same genomic
features but a different quantification. A typical workflow contemplates
<code>seqCount</code> and <code>fragmentEstimate</code>, all the
supported quantification types can be viewed with
<code><a href="../reference/quantification_types.html">quantification_types()</a></code>.</p>
</div>
<div class="section level4">
<h4 id="matrix_type-argument">
<code>matrix_type</code> argument<a class="anchor" aria-label="anchor" href="#matrix_type-argument"></a>
</h4>
<p>As we mentioned in Section @ref(notation), annotation columns are
optional and may not be included in some matrices. This argument allows
you to specify the function to look for only a specific type of matrix,
either <code>annotated</code> or <code>not_annotated</code>.</p>
<p>File suffixes for matrices are specified via
<code><a href="../reference/mandatory_IS_vars.html">matrix_file_suffixes()</a></code>.</p>
</div>
<div class="section level4">
<h4 id="workers-argument">
<code>workers</code> argument<a class="anchor" aria-label="anchor" href="#workers-argument"></a>
</h4>
<p>Sets the number of parallel workers to set up. This highly depends on
the hardware configuration of your machine.</p>
</div>
<div class="section level4">
<h4 id="multi_quant_matrix-argument">
<code>multi_quant_matrix</code> argument<a class="anchor" aria-label="anchor" href="#multi_quant_matrix-argument"></a>
</h4>
<p>When importing more than one quantification at once, it can be very
handy to have all data in a single data frame rather than two. If set to
<code>TRUE</code> the function will internally call
<code><a href="../reference/comparison_matrix.html">comparison_matrix()</a></code> and produce a single data frames that
has a dedicated column for each quantification. For example, for the
matrices we’ve imported before:</p>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">#&gt;   chr   integration_locus strand GeneName     GeneStrand CompleteAmplificationID</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 16             68<span style="text-decoration: underline;">164</span>148 +      NFATC3       +          PJ01_POOL01_LTR75LC38_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 4             129<span style="text-decoration: underline;">390</span>130 +      LOC100507487 +          PJ01_POOL01_LTR75LC38_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 5              84<span style="text-decoration: underline;">009</span>671 -      EDIL3        -          PJ01_POOL01_LTR75LC38_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 12             54<span style="text-decoration: underline;">635</span>693 -      CBX5         -          PJ01_POOL01_LTR75LC38_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 2             181<span style="text-decoration: underline;">930</span>711 +      UBE2E3       +          PJ01_POOL01_LTR75LC38_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 20             35<span style="text-decoration: underline;">920</span>986 +      MANBAL       +          PJ01_POOL01_LTR75LC38_…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: fragmentEstimate &lt;dbl&gt;, seqCount &lt;int&gt;</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="report_path-argument">
<code>report_path</code> argument<a class="anchor" aria-label="anchor" href="#report_path-argument"></a>
</h4>
<p>As other import functions, also
<code><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices()</a></code> produces an interactive
report, use this argument to set the appropriate path were the report
should be saved.</p>
</div>
<div class="section level4">
<h4 id="mode-argument">
<code>mode</code> argument<a class="anchor" aria-label="anchor" href="#mode-argument"></a>
</h4>
<p>Since <code>ISAnalytics 1.8.3</code> this argument can only be set to
<code>AUTO</code>.</p>
<p><strong>What do you want to import?</strong><br>
In a fully automated mode, the function will try to import everything
that is contained in the input association file. This means that if you
need to import only a specific set of projects/pools, you will need to
filter the association file accordingly prior calling the function (you
can easily do that via the <code>filter_for</code> argument as explained
in Section @ref(metadata)).</p>
<p><strong>How to deal with duplicates?</strong><br>
When scanning folders for files that match a given pattern (in our case
the function looks for matrices that match the quantification type and
the matrix type), it is very possible that the same folder contains
multiple files for the same quantification. Of course this is not
recommended, we suggest to move the duplicated files in a sub directory
or remove them if they’re not necessary, but in case this happens, you
need to set two other arguments (described in the next sub sections) to
“help” the function discriminate between duplicates. Please note that if
such discrimination is not possible no files are imported.</p>
</div>
<div class="section level4">
<h4 id="patterns-argument">
<code>patterns</code> argument<a class="anchor" aria-label="anchor" href="#patterns-argument"></a>
</h4>
<p>Providing a set of patterns (interpreted as regular expressions)
helps the function to choose between duplicated files if any are found.
If you’re confident your folders don’t contain any duplicates feel free
to ignore this argument.</p>
</div>
<div class="section level4">
<h4 id="matching_opt-argument">
<code>matching_opt</code> argument<a class="anchor" aria-label="anchor" href="#matching_opt-argument"></a>
</h4>
<p>This argument is relevant only if <code>patterns</code> isn’t
<code>NULL</code>. Tells the function how to match the given patterns if
multiple are supplied: <code>ALL</code> means keep only those files
whose name matches all the given patterns, <code>ANY</code> means keep
only those files whose name matches any of the given patterns and
<code>OPTIONAL</code> expresses a preference, try to find files that
contain the patterns and if you don’t find any return whatever you
find.</p>
</div>
<div class="section level4">
<h4 id="argument">
<code>...</code> argument<a class="anchor" aria-label="anchor" href="#argument"></a>
</h4>
<p>Additional named arguments to supply to
<code><a href="../reference/comparison_matrix.html">comparison_matrix()</a></code> and
<code>import_single_Vispa2_matrix</code></p>
</div>
</div>
<div class="section level3">
<h3 id="notes">Notes<a class="anchor" aria-label="anchor" href="#notes"></a>
</h3>
<p>Earlier versions of the package featured two separated functions,
<code><a href="../reference/import_parallel_Vispa2Matrices_auto.html">import_parallel_Vispa2Matrices_auto()</a></code> and
<code><a href="../reference/import_parallel_Vispa2Matrices_interactive.html">import_parallel_Vispa2Matrices_interactive()</a></code>. Those
functions are now officially deprecated (since
<code>ISAnalytics 1.3.3</code>) and will be defunct on the next release
cycle.</p>
</div>
</div>
<div class="section level2">
<h2 id="data-cleaning-and-pre-processing">Data cleaning and pre-processing<a class="anchor" aria-label="anchor" href="#data-cleaning-and-pre-processing"></a>
</h2>
<p>This section goes more in detail on some data cleaning and
pre-processing operations you can perform with this package.</p>
<p>ISAnalytics offers several different functions for cleaning and
pre-processing your data.</p>
<ul>
<li>Recalibration: identifies integration events that are near to each
other and condenses them into a single event whenever appropriate -
<code><a href="../reference/compute_near_integrations.html">compute_near_integrations()</a></code>
</li>
<li>Outliers identification and removal: identifies samples that are
considered outliers according to user-defined logic and filters them out
- <code><a href="../reference/outlier_filter.html">outlier_filter()</a></code>
</li>
<li>Collision removal: identifies collision events between independent
samples - <code><a href="../reference/remove_collisions.html">remove_collisions()</a></code>
</li>
<li>Filter based on cell lineage purity: identifies and removes
contamination between different cell types -
<code><a href="../reference/purity_filter.html">purity_filter()</a></code>
</li>
<li>Data and metadata aggregation: allows the union of biological
samples from single pcr replicates or other arbitrary aggregations -
<code><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key()</a></code>,
<code><a href="../reference/aggregate_metadata.html">aggregate_metadata()</a></code>
</li>
</ul>
<div class="section level3">
<h3 id="removing-collisions">Removing collisions<a class="anchor" aria-label="anchor" href="#removing-collisions"></a>
</h3>
<p>In this section we illustrate the functions dedicated to collision
removal.</p>
<div class="section level4">
<h4 id="what-is-a-collision-and-why-should-you-care">What is a collision and why should you care?<a class="anchor" aria-label="anchor" href="#what-is-a-collision-and-why-should-you-care"></a>
</h4>
<p>We’re not going into too much detail here, but we’re going to explain
in a very simple way what a “collision” is and how the function in this
package deals with them.</p>
<p>We say that an integration (aka a unique combination of
<code><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars()</a></code>) is a <em>collision</em> if this
combination is shared between different independent samples: an
independent sample is a unique combination of metadata fields specified
by the user. The reason behind this is that it’s highly improbable to
observe the very same integration in two different independent samples
and this phenomenon might be an indicator of some kind of contamination
in the sequencing phase or in PCR phase, for this reason we might want
to exclude such contamination from our analysis.
<code>ISAnalytics</code> provides a function that processes the imported
data for the removal or reassignment of these “problematic”
integrations, <code><a href="../reference/remove_collisions.html">remove_collisions()</a></code>.</p>
<p>The processing is done using the sequence count value, so the
corresponding matrix is needed for this operation.</p>
</div>
<div class="section level4">
<h4 id="the-logic-behind-the-function">The logic behind the function<a class="anchor" aria-label="anchor" href="#the-logic-behind-the-function"></a>
</h4>
<p>The <code><a href="../reference/remove_collisions.html">remove_collisions()</a></code> function follows several logical
steps to decide whether an integration is a collision and if it is it
decides whether to re-assign it or remove it entirely based on different
criteria.</p>
<div class="section level5">
<h5 id="identifying-the-collisions">Identifying the collisions<a class="anchor" aria-label="anchor" href="#identifying-the-collisions"></a>
</h5>
<p>The function uses the information stored in the association file to
assess which independent samples are present and counts the number of
independent samples for each integration: those who have a count &gt; 1
are considered collisions.</p>
<table class="table">
<caption>Example of collisions: the same integration (1, 123454, +) is
found in 2 different independent samples ((SUBJ01, PJ01) &amp; (SUBJ02,
PJ01))</caption>
<colgroup>
<col width="4%">
<col width="21%">
<col width="8%">
<col width="10%">
<col width="29%">
<col width="12%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left">chr</th>
<th align="right">integration_locus</th>
<th align="left">strand</th>
<th align="right">seqCount</th>
<th align="left">CompleteAmplificationID</th>
<th align="left">SubjectID</th>
<th align="left">ProjectID</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="right">123454</td>
<td align="left">+</td>
<td align="right">653</td>
<td align="left">SAMPLE1</td>
<td align="left">SUBJ01</td>
<td align="left">PJ01</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="right">123454</td>
<td align="left">+</td>
<td align="right">456</td>
<td align="left">SAMPLE2</td>
<td align="left">SUBJ02</td>
<td align="left">PJ01</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="re-assign-vs-remove">Re-assign vs remove<a class="anchor" aria-label="anchor" href="#re-assign-vs-remove"></a>
</h5>
<p>Once the collisions are identified, the function follows 3 steps
where it tries to re-assign the combination to a single independent
sample. The criteria are:</p>
<ol style="list-style-type: decimal">
<li>Compare dates: if it’s possible to have an absolute ordering on
dates, the integration is re-assigned to the sample that has the
earliest date. If two samples share the same date it’s impossible to
decide, so the next criteria is tested</li>
<li>Compare replicate number: if a sample has the same integration in
more than one replicate, it’s more probable the integration is not an
artifact. If it’s possible to have an absolute ordering, the collision
is re-assigned to the sample whose grouping is largest</li>
<li>Compare the sequence count value: if the previous criteria wasn’t
sufficient to make a decision, for each group of independent samples
it’s evaluated the sum of the sequence count value - for each group
there is a cumulative value of the sequence count and this is compared
to the value of other groups. If there is a single group which has a
ratio n times bigger than other groups, this one is chosen for
re-assignment. The factor n is passed as a parameter in the function
(<code>reads_ratio</code>), the default value is 10.</li>
</ol>
<p>If none of the criteria were sufficient to make a decision, the
integration is simply removed from the matrix.</p>
</div>
</div>
<div class="section level4">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h4>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"integration_matrices"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"association_file"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span>
<span><span class="co">## Multi quantification matrix</span></span>
<span><span class="va">no_coll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_collisions.html">remove_collisions</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">integration_matrices</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    report_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Identifying collisions...</span></span>
<span><span class="co">#&gt; Processing collisions...</span></span>
<span><span class="co">#&gt; Finished!</span></span>
<span><span class="co">## Matrix list</span></span>
<span><span class="va">separated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/separate_quant_matrices.html">separate_quant_matrices</a></span><span class="op">(</span><span class="va">integration_matrices</span><span class="op">)</span></span>
<span><span class="va">no_coll_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_collisions.html">remove_collisions</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">separated</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    report_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Identifying collisions...</span></span>
<span><span class="co">#&gt; Processing collisions...</span></span>
<span><span class="co">#&gt; Finished!</span></span>
<span><span class="co">## Only sequence count</span></span>
<span><span class="va">no_coll_single</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_collisions.html">remove_collisions</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">separated</span><span class="op">$</span><span class="va">seqCount</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    quant_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>seqCount <span class="op">=</span> <span class="st">"Value"</span><span class="op">)</span>,</span>
<span>    report_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Identifying collisions...</span></span>
<span><span class="co">#&gt; Processing collisions...</span></span>
<span><span class="co">#&gt; Finished!</span></span></code></pre></div>
<p>Important notes on the association file:</p>
<ul>
<li>You have to be sure your association file is properly filled out.
The function requires you to specify a date column (by default
“SequencingDate”), you have to ensure this column doesn’t contain NA
values or incorrect values.</li>
</ul>
<p>The function accepts different inputs, namely:</p>
<ul>
<li>A multi-quantification matrix: this is always the recommended
approach</li>
<li>A named list of matrices where names are quantification types in
<code><a href="../reference/quantification_types.html">quantification_types()</a></code>
</li>
<li>The single sequence count matrix: this is not the recommended
approach since it requires a realignment step for other quantification
matrices if you have them.</li>
</ul>
<p>If the option <code>ISAnalytics.reports</code> is active, an
interactive report in HTML format will be produced at the specified
path.</p>
</div>
<div class="section level4">
<h4 id="re-align-other-matrices">Re-align other matrices<a class="anchor" aria-label="anchor" href="#re-align-other-matrices"></a>
</h4>
<p>If you’ve given as input the standalone sequence count matrix to
<code><a href="../reference/remove_collisions.html">remove_collisions()</a></code>, to realign other matrices you have to
call the function <code><a href="../reference/realign_after_collisions.html">realign_after_collisions()</a></code>, passing as
input the processed sequence count matrix and the named list of other
matrices to realign. <strong>NOTE: the names in the list must be
quantification types.</strong></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">other_realigned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/realign_after_collisions.html">realign_after_collisions</a></span><span class="op">(</span></span>
<span>    sc_matrix <span class="op">=</span> <span class="va">no_coll_single</span>,</span>
<span>    other_matrices <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fragmentEstimate <span class="op">=</span> <span class="va">separated</span><span class="op">$</span><span class="va">fragmentEstimate</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="performing-data-and-metadata-aggregation">Performing data and metadata aggregation<a class="anchor" aria-label="anchor" href="#performing-data-and-metadata-aggregation"></a>
</h3>
<p>In this section we’re going to explain in detail how to use functions
of the aggregate family, namely:</p>
<ol style="list-style-type: decimal">
<li><code><a href="../reference/aggregate_metadata.html">aggregate_metadata()</a></code></li>
<li><code><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key()</a></code></li>
</ol>
<div class="section level4">
<h4 id="aggregating-metadata">Aggregating metadata<a class="anchor" aria-label="anchor" href="#aggregating-metadata"></a>
</h4>
<p>We refer to information contained in the association file as
“metadata”: sometimes it’s useful to obtain collective information based
on a certain group of variables we’re interested in. The function
<code><a href="../reference/aggregate_metadata.html">aggregate_metadata()</a></code> does just that: according to the
grouping variables, meaning the names of the columns in the association
file to perform a <code>group_by</code> operation with,it creates a
summary. You can fully customize the summary by providing a “function
table” that tells the function which operation should be applied to
which column and what name to give to the output column. A default is
already supplied:</p>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 15 × 4</span></span></span>
<span><span class="co">#&gt;    Column               Function  Args  Output_colname</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> FusionPrimerPCRDate  <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_min    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> LinearPCRDate        <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_min    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> VCN                  <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_avg    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ng DNA corrected     <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_avg    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Kapa                 <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_avg    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ng DNA corrected     <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_sum    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ulForPool            <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_sum    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> BARCODE_MUX          <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_sum    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> TRIMMING_FINAL_LTRLC <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_sum    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LV_MAPPED            <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_sum    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> BWA_MAPPED_OVERALL   <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_sum    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> ISS_MAPPED_OVERALL   <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}_sum    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> PCRMethod            <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> NGSTechnology        <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> DNAnumber            <span style="color: #949494;">&lt;formula&gt;</span> <span style="color: #BB0000;">NA</span>    {.col}</span></span></code></pre>
<p>You can either provide purrr-style lambdas (as given in the example
above), or simply specify the name of the function and additional
parameters as a list in a separated column. If you choose to provide
your own table you should maintain the column names for the function to
work properly. For more details on this take a look at the function
documentation <code><a href="../reference/default_meta_agg.html">?default_meta_agg</a></code>.</p>
<div class="section level5">
<h5 id="typical-workflow">Typical workflow<a class="anchor" aria-label="anchor" href="#typical-workflow"></a>
</h5>
<ol style="list-style-type: decimal">
<li>Import the association file via
<code>import_assocition_file()</code>. If you need more information on
import function please view the vignette “How to use import functions”:
<code>vignette("how_to_import_functions", package="ISAnalytics")</code>.</li>
<li>Perform aggregation</li>
</ol>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"association_file"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span>
<span><span class="va">aggregated_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_metadata.html">aggregate_metadata</a></span><span class="op">(</span>association_file <span class="op">=</span> <span class="va">association_file</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 19</span></span></span>
<span><span class="co">#&gt;    SubjectID CellMarker Tissue TimePoint FusionPrimerPCRDate_min</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;date&gt;</span>                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> PT001     MNC        BM     0030      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> PT001     MNC        BM     0060      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> PT001     MNC        BM     0090      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> PT001     MNC        BM     0180      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PT001     MNC        BM     0360      2017-04-21             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> PT001     MNC        PB     0030      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PT001     MNC        PB     0060      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> PT001     MNC        PB     0090      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PT001     MNC        PB     0180      2016-11-03             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PT001     MNC        PB     0360      2017-04-21             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> PT002     MNC        BM     0030      2017-04-21             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> PT002     MNC        BM     0060      2017-05-05             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> PT002     MNC        BM     0090      2017-05-05             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> PT002     MNC        BM     0180      2017-05-16             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> PT002     MNC        BM     0360      2018-03-12             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> PT002     MNC        PB     0030      2017-04-21             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> PT002     MNC        PB     0060      2017-05-05             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> PT002     MNC        PB     0090      2017-05-05             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> PT002     MNC        PB     0180      2017-05-05             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> PT002     MNC        PB     0360      2018-03-12             </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 14 more variables: LinearPCRDate_min &lt;date&gt;, VCN_avg &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   `ng DNA corrected_avg` &lt;dbl&gt;, Kapa_avg &lt;dbl&gt;, `ng DNA corrected_sum` &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ulForPool_sum &lt;dbl&gt;, BARCODE_MUX_sum &lt;int&gt;, TRIMMING_FINAL_LTRLC_sum &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   LV_MAPPED_sum &lt;int&gt;, BWA_MAPPED_OVERALL_sum &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ISS_MAPPED_OVERALL_sum &lt;int&gt;, PCRMethod &lt;chr&gt;, NGSTechnology &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   DNAnumber &lt;chr&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="aggregation-of-values-by-key">Aggregation of values by key<a class="anchor" aria-label="anchor" href="#aggregation-of-values-by-key"></a>
</h4>
<p><code>ISAnalytics</code> contains useful functions to aggregate the
values contained in your imported matrices based on a key, aka a single
column or a combination of columns contained in the association file
that are related to the samples.</p>
<div class="section level5">
<h5 id="typical-workflow-1">Typical workflow<a class="anchor" aria-label="anchor" href="#typical-workflow-1"></a>
</h5>
<ol style="list-style-type: decimal">
<li>Import your association file</li>
<li>Import integration matrices via
<code><a href="../reference/import_parallel_Vispa2Matrices.html">import_parallel_Vispa2Matrices()</a></code>
</li>
<li>Perform aggregation</li>
</ol>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"integration_matrices"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"association_file"</span>, package <span class="op">=</span> <span class="st">"ISAnalytics"</span><span class="op">)</span></span>
<span><span class="va">aggreg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">integration_matrices</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    value_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,074 × 11</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand SubjectID CellMarker</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1               8<span style="text-decoration: underline;">607</span>362 -      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1               8<span style="text-decoration: underline;">850</span>362 +      RERE     -          PT002     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001     MNC       </span></span>
<span><span class="co">#&gt;    Tissue TimePoint seqCount_sum fragmentEstimate_sum</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM     0030               542                 3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> BM     0060                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> BM     0060                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> BM     0180              <span style="text-decoration: underline;">1</span>096                 5.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> BM     0360               330                34.1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> BM     0180              <span style="text-decoration: underline;">1</span>702                 4.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> BM     0360               562                 3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> BM     0060              <span style="text-decoration: underline;">1</span>605                 8.03</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PB     0060                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PB     0180                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,064 more rows</span></span></span></code></pre>
<p>The function <code>aggregate_values_by_key</code> can perform the
aggregation both on the list of matrices and a single matrix.</p>
<div class="section level6">
<h6 id="changing-parameters-to-obtain-different-results">Changing parameters to obtain different results<a class="anchor" aria-label="anchor" href="#changing-parameters-to-obtain-different-results"></a>
</h6>
<p>The function has several different parameters that have default
values that can be changed according to user preference.</p>
<ol style="list-style-type: decimal">
<li>
<strong>Changing the <code>key</code> value</strong><br>
You can change the value of the parameter key as you see fit. This
parameter should contain one or multiple columns of the association file
that you want to include in the grouping when performing the
aggregation. The default value is set to
<code>c("SubjectID", "CellMarker", "Tissue", "TimePoint")</code> (same
default key as the <code>aggregate_metadata</code> function).</li>
</ol>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agg1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">integration_matrices</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SubjectID"</span>, <span class="st">"ProjectID"</span><span class="op">)</span>,</span>
<span>    value_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 577 × 9</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand SubjectID ProjectID</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1               8<span style="text-decoration: underline;">607</span>362 -      RERE     -          PT001     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1               8<span style="text-decoration: underline;">850</span>362 +      RERE     -          PT002     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1              12<span style="text-decoration: underline;">341</span>466 -      VPS13D   +          PT002     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1              14<span style="text-decoration: underline;">034</span>054 -      PRDM2    +          PT002     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1              16<span style="text-decoration: underline;">186</span>297 -      SPEN     +          PT001     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT001     PJ01     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT002     PJ01     </span></span>
<span><span class="co">#&gt;    seqCount_sum fragmentEstimate_sum</span></span>
<span><span class="co">#&gt;           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>          543                 4.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>         <span style="text-decoration: underline;">1</span>427                40.1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>         <span style="text-decoration: underline;">1</span>702                 4.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>          562                 3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>         <span style="text-decoration: underline;">1</span>607                10.0 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>         <span style="text-decoration: underline;">1</span>843                 8.05</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>         <span style="text-decoration: underline;">1</span>938                 3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>         <span style="text-decoration: underline;">3</span>494                16.1 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>         <span style="text-decoration: underline;">2</span>947                 9.04</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>           30                 2.00</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 567 more rows</span></span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>Changing the <code>lambda</code> value</strong><br>
The <code>lambda</code> parameter indicates the function(s) to be
applied to the values for aggregation. <code>lambda</code> must be a
named list of either functions or purrr-style lambdas: if you would like
to specify additional parameters to the function the second option is
recommended. The only important note on functions is that they should
perform some kind of aggregation on numeric values: this means in
practical terms they need to accept a vector of numeric/integer values
as input and produce a SINGLE value as output. Valid options for this
purpose might be: <code>sum</code>, <code>mean</code>,
<code>median</code>, <code>min</code>, <code>max</code> and so on.</li>
</ol>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">integration_matrices</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    key <span class="op">=</span> <span class="st">"SubjectID"</span>,</span>
<span>    lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">.x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    value_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 577 × 8</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand SubjectID seqCount_mean</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001              272.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001              285.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1               8<span style="text-decoration: underline;">607</span>362 -      RERE     -          PT001              851 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1               8<span style="text-decoration: underline;">850</span>362 +      RERE     -          PT002              562 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001              321.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1              12<span style="text-decoration: underline;">341</span>466 -      VPS13D   +          PT002             <span style="text-decoration: underline;">1</span>843 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1              14<span style="text-decoration: underline;">034</span>054 -      PRDM2    +          PT002             <span style="text-decoration: underline;">1</span>938 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1              16<span style="text-decoration: underline;">186</span>297 -      SPEN     +          PT001              699.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT001              982.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT002               30 </span></span>
<span><span class="co">#&gt;    fragmentEstimate_mean</span></span>
<span><span class="co">#&gt;                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>                  2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>                  8.02</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>                  2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>                  2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>                  8.05</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>                  3.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>                  2.00</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 567 more rows</span></span></span></code></pre>
<p>Note that, when specifying purrr-style lambdas (formulas), the first
parameter needs to be set to <code>.x</code>, other parameters can be
set as usual.</p>
<p>You can also use in <code>lambda</code> functions that produce data
frames or lists. In this case all variables from the produced data frame
will be included in the final data frame. For example:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agg3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">integration_matrices</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    key <span class="op">=</span> <span class="st">"SubjectID"</span>,</span>
<span>    lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>describe <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">psych</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/psych/man/describe.html" class="external-link">describe</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    value_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 577 × 8</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand SubjectID</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1               8<span style="text-decoration: underline;">607</span>362 -      RERE     -          PT001    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1               8<span style="text-decoration: underline;">850</span>362 +      RERE     -          PT002    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1              12<span style="text-decoration: underline;">341</span>466 -      VPS13D   +          PT002    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1              14<span style="text-decoration: underline;">034</span>054 -      PRDM2    +          PT002    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1              16<span style="text-decoration: underline;">186</span>297 -      SPEN     +          PT001    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT001    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT002    </span></span>
<span><span class="co">#&gt;    seqCount_describe fragmentEstimate_describe</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>                   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>  <span style="color: #949494;">&lt;psych [1 × 13]&gt;</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 567 more rows</span></span></span></code></pre>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Changing the <code>value_cols</code> value</strong><br>
The <code>value_cols</code> parameter tells the function on which
numeric columns of x the functions should be applied. Note that every
function contained in <code>lambda</code> will be applied to every
column in <code>value_cols</code>: resulting columns will be named as
“original name_function applied”.</li>
</ol>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agg4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">integration_matrices</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    key <span class="op">=</span> <span class="st">"SubjectID"</span>,</span>
<span>    lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sum <span class="op">=</span> <span class="va">sum</span>, mean <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>,</span>
<span>    value_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 577 × 10</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand SubjectID seqCount_sum</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001              543</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001             <span style="text-decoration: underline;">1</span>427</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1               8<span style="text-decoration: underline;">607</span>362 -      RERE     -          PT001             <span style="text-decoration: underline;">1</span>702</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1               8<span style="text-decoration: underline;">850</span>362 +      RERE     -          PT002              562</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001             <span style="text-decoration: underline;">1</span>607</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1              12<span style="text-decoration: underline;">341</span>466 -      VPS13D   +          PT002             <span style="text-decoration: underline;">1</span>843</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1              14<span style="text-decoration: underline;">034</span>054 -      PRDM2    +          PT002             <span style="text-decoration: underline;">1</span>938</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1              16<span style="text-decoration: underline;">186</span>297 -      SPEN     +          PT001             <span style="text-decoration: underline;">3</span>494</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT001             <span style="text-decoration: underline;">2</span>947</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT002               30</span></span>
<span><span class="co">#&gt;    seqCount_mean fragmentEstimate_sum fragmentEstimate_mean</span></span>
<span><span class="co">#&gt;            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>          272.                 4.01                  2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>          285.                40.1                   8.02</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>          851                  4.01                  2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>          562                  3.01                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>          321.                10.0                   2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>         <span style="text-decoration: underline;">1</span>843                  8.05                  8.05</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>         <span style="text-decoration: underline;">1</span>938                  3.01                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>          699.                16.1                   3.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>          982.                 9.04                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>           30                  2.00                  2.00</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 567 more rows</span></span></span></code></pre>
<ol start="4" style="list-style-type: decimal">
<li>
<strong>Changing the <code>group</code> value</strong><br>
The <code>group</code> parameter should contain all other variables to
include in the grouping besides <code>key</code>. By default this
contains
<code>c("chr", "integration_locus","strand", "GeneName", "GeneStrand")</code>.
You can change this grouping as you see fit, if you don’t want to add
any other variable to the key, just set it to <code>NULL</code>.</li>
</ol>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agg5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">integration_matrices</span>,</span>
<span>    association_file <span class="op">=</span> <span class="va">association_file</span>,</span>
<span>    key <span class="op">=</span> <span class="st">"SubjectID"</span>,</span>
<span>    lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sum <span class="op">=</span> <span class="va">sum</span>, mean <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>,</span>
<span>    group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/mandatory_IS_vars.html">mandatory_IS_vars</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    value_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 577 × 8</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand SubjectID seqCount_sum seqCount_mean</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1               8<span style="text-decoration: underline;">464</span>757 -      PT001              543          272.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1               8<span style="text-decoration: underline;">607</span>357 +      PT001             <span style="text-decoration: underline;">1</span>427          285.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1               8<span style="text-decoration: underline;">607</span>362 -      PT001             <span style="text-decoration: underline;">1</span>702          851 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1               8<span style="text-decoration: underline;">850</span>362 +      PT002              562          562 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1              11<span style="text-decoration: underline;">339</span>120 +      PT001             <span style="text-decoration: underline;">1</span>607          321.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1              12<span style="text-decoration: underline;">341</span>466 -      PT002             <span style="text-decoration: underline;">1</span>843         <span style="text-decoration: underline;">1</span>843 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1              14<span style="text-decoration: underline;">034</span>054 -      PT002             <span style="text-decoration: underline;">1</span>938         <span style="text-decoration: underline;">1</span>938 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1              16<span style="text-decoration: underline;">186</span>297 -      PT001             <span style="text-decoration: underline;">3</span>494          699.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              16<span style="text-decoration: underline;">602</span>483 +      PT001             <span style="text-decoration: underline;">2</span>947          982.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1              16<span style="text-decoration: underline;">602</span>483 +      PT002               30           30 </span></span>
<span><span class="co">#&gt;    fragmentEstimate_sum fragmentEstimate_mean</span></span>
<span><span class="co">#&gt;                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>                 4.01                  2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>                40.1                   8.02</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>                 4.01                  2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>                 3.01                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>                10.0                   2.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>                 8.05                  8.05</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>                 3.01                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>                16.1                   3.22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>                 9.04                  3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>                 2.00                  2.00</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 567 more rows</span></span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="analysis-use-case-example-shared-integration-sites">Analysis use-case example: shared integration sites<a class="anchor" aria-label="anchor" href="#analysis-use-case-example-shared-integration-sites"></a>
</h2>
<p>An integration site is always characterized by a triple of values:
<code>(chr, integration_locus, strand)</code>, hence these attributes
are always present in integration matrices.</p>
<p>We can aggregate our data in different ways according to our needs,
obtaining therefore different groups. Each group has an associated set
of integration sites.</p>
<p>NOTE: aggregating data is not mandatory, since sharing functions in
ISAnalytics only count distinct integration sites and do not require any
quantification. The only important thing is that columns that are
included in the specified key are also included in the input
matrices.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Aggregation by standard key</span></span>
<span><span class="va">agg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_values_by_key.html">aggregate_values_by_key</a></span><span class="op">(</span><span class="va">integration_matrices</span>,</span>
<span>    <span class="va">association_file</span>,</span>
<span>    value_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqCount"</span>, <span class="st">"fragmentEstimate"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">agg</span> <span class="op">&lt;-</span> <span class="va">agg</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">TimePoint</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0030"</span>, <span class="st">"0060"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 419 × 11</span></span></span>
<span><span class="co">#&gt;    chr   integration_locus strand GeneName GeneStrand SubjectID CellMarker</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 1               8<span style="text-decoration: underline;">464</span>757 -      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 1               8<span style="text-decoration: underline;">607</span>357 +      RERE     -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 1              11<span style="text-decoration: underline;">339</span>120 +      UBIAD1   +          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 1              16<span style="text-decoration: underline;">186</span>297 -      SPEN     +          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 1              16<span style="text-decoration: underline;">186</span>297 -      SPEN     +          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 1              16<span style="text-decoration: underline;">602</span>483 +      FBXO42   -          PT001     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 1              25<span style="text-decoration: underline;">337</span>264 -      MIR4425  +          PT002     MNC       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 1              25<span style="text-decoration: underline;">337</span>264 -      MIR4425  +          PT002     MNC       </span></span>
<span><span class="co">#&gt;    Tissue TimePoint seqCount_sum fragmentEstimate_sum</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> BM     0030               542                 3.01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> BM     0060                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> BM     0060                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> BM     0060              <span style="text-decoration: underline;">1</span>605                 8.03</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PB     0060                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> BM     0030                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PB     0060                 1                 1.00</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> BM     0060              <span style="text-decoration: underline;">2</span>947                 9.04</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> BM     0030                23                 9.14</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PB     0060                36                 7.07</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 409 more rows</span></span></span></code></pre>
<p>An integration site is <em>shared</em> between two or more groups if
the same triple is observed in all the groups considered.</p>
<div class="section level3">
<h3 id="automated-sharing-counts">Automated sharing counts<a class="anchor" aria-label="anchor" href="#automated-sharing-counts"></a>
</h3>
<p>ISAnalytics provides the function <code><a href="../reference/is_sharing.html">is_sharing()</a></code> for
computing automated sharing counts. The function has several arguments
that can be tuned according to user needs.</p>
</div>
<div class="section level3">
<h3 id="scenario-1-single-input-data-frame-and-single-grouping-key">SCENARIO 1: single input data frame and single grouping key<a class="anchor" aria-label="anchor" href="#scenario-1-single-input-data-frame-and-single-grouping-key"></a>
</h3>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sharing_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span><span class="va">agg</span>,</span>
<span>    group_key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>,</span>
<span>        <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    n_comp <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    is_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    relative_is_sharing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    minimal <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    include_self_comp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    keep_genomic_coord <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sharing_1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 28 × 10</span></span></span>
<span><span class="co">#&gt;    g1           g2    shared count_g1 count_g2 count_union  on_g1 on_g2 on_union</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> PT001_MNC_B… PT00…     21       54      114         147 38.9   18.4    14.3  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> PT001_MNC_B… PT00…      7       54       28          75 13.0   25       9.33 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> PT001_MNC_B… PT00…      8       54       59         105 14.8   13.6     7.62 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> PT001_MNC_B… PT00…      0       54       98         152  0      0       0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PT001_MNC_B… PT00…      1       54       33          86  1.85   3.03    1.16 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> PT001_MNC_B… PT00…      0       54       15          69  0      0       0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PT001_MNC_B… PT00…      0       54       18          72  0      0       0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> PT001_MNC_B… PT00…      7      114       28         135  6.14  25       5.19 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PT001_MNC_B… PT00…     29      114       59         144 25.4   49.2    20.1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PT001_MNC_B… PT00…      1      114       98         211  0.877  1.02    0.474</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 18 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: is_coord &lt;list&gt;</span></span></span></code></pre></div>
<p>In this configuration we set:</p>
<ul>
<li>A single input data frame: <code>agg</code>
</li>
<li>A single grouping key by setting the argument
<code>grouping_key</code>. In this specific case, our groups will be
identified by a unique combination of <code>SubjectID</code>,
<code>CellMarker</code>, <code>Tissue</code> and
<code>TimePoint</code>
</li>
<li>
<code>n_comp</code> represents the number of comparisons to compute:
2 means we’re interested in knowing the sharing for PAIRS of distinct
groups</li>
<li>We want to keep the counts of distinct integration sites for each
group by setting <code>is_count</code> to <code>TRUE</code>
</li>
<li>
<code>relative_is_sharing</code> if set to <code>TRUE</code> adds
sharing expressed as a percentage, more precisely it adds a column
<code>on_g1</code> that is calculated as the absolute number of shared
integrations divided by the cardinality of the first group,
<code>on_g2</code> is analogous but is computed on the cardinality of
the second group and finally <code>on_union</code> is computed on the
cardinality of the union of the two groups.</li>
<li>By setting the argument <code>minimal</code> to <code>TRUE</code> we
tell the function to avoid redundant comparisons: in this way only
<strong>combinations</strong> and not permutations are included in the
output table</li>
<li>
<code>include_self_comp</code> adds rows in the table that are
labelled with the same group: these rows always have a 100% sharing with
all other groups. There are few scenarios where this is useful, but for
now we set it to <code>FALSE</code> since we don’t need it</li>
<li>
<code>keep_genomic_coord</code> allows us to keep the genomic
coordinates of the shared integration sites as a separate table</li>
</ul>
<div class="section level5">
<h5 id="changing-the-number-of-comparisons">Changing the number of comparisons<a class="anchor" aria-label="anchor" href="#changing-the-number-of-comparisons"></a>
</h5>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sharing_1_a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span><span class="va">agg</span>,</span>
<span>    group_key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>,</span>
<span>        <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    n_comp <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    is_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    relative_is_sharing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    minimal <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    include_self_comp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    keep_genomic_coord <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sharing_1_a</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 56 × 13</span></span></span>
<span><span class="co">#&gt;    g1      g2    g3    shared count_g1 count_g2 count_g3 count_union on_g1 on_g2</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> PT001_… PT00… PT00…      5       54      114       28         166  9.26  4.39</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> PT001_… PT00… PT00…      6       54      114       59         175 11.1   5.26</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> PT001_… PT00… PT00…      0       54      114       98         244  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> PT001_… PT00… PT00…      0       54      114       33         179  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PT001_… PT00… PT00…      0       54      114       15         161  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> PT001_… PT00… PT00…      0       54      114       18         165  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PT001_… PT00… PT00…      1       54       28       59         117  1.85  3.57</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> PT001_… PT00… PT00…      0       54       28       98         173  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PT001_… PT00… PT00…      0       54       28       33         107  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PT001_… PT00… PT00…      0       54       28       15          90  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 46 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: on_g3 &lt;dbl&gt;, on_union &lt;dbl&gt;, is_coord &lt;list&gt;</span></span></span></code></pre></div>
<p>Changing the <code>n_comp</code> to 3 means that we want to calculate
the sharing between 3 different groups. Note that the
<code>shared</code> column contains the counts of integrations that are
shared by <strong>ALL</strong> groups, which is equivalent to a set
intersection.</p>
<p>Beware of the fact that the more comparisons are requested the more
time the computation requires.</p>
</div>
<div class="section level5">
<h5 id="a-case-when-it-is-useful-to-set-minimal-false">A case when it is useful to set <code>minimal = FALSE</code><a class="anchor" aria-label="anchor" href="#a-case-when-it-is-useful-to-set-minimal-false"></a>
</h5>
<p>Setting <code>minimal = FALSE</code> produces all possible
permutations of the groups and the corresponding values. In combination
with <code>include_self_comp = TRUE</code>, this is useful when we want
to know the sharing between pairs of groups and plot results as a
heatmap.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sharing_1_b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span><span class="va">agg</span>,</span>
<span>    group_key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>,</span>
<span>        <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    n_comp <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    is_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    relative_is_sharing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    minimal <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    include_self_comp <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sharing_1_b</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 64 × 9</span></span></span>
<span><span class="co">#&gt;    g1            g2    shared count_g1 count_g2 count_union on_g1 on_g2 on_union</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> PT001_MNC_BM… PT00…     54       54       54          54 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> PT001_MNC_BM… PT00…    114      114      114         114 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> PT001_MNC_PB… PT00…     59       59       59          59 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> PT002_MNC_BM… PT00…     98       98       98          98 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PT002_MNC_PB… PT00…     18       18       18          18 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> PT002_MNC_PB… PT00…     15       15       15          15 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PT001_MNC_PB… PT00…     28       28       28          28 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> PT002_MNC_BM… PT00…     33       33       33          33 100   100      100  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PT001_MNC_BM… PT00…     21       54      114         147  38.9  18.4     14.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PT001_MNC_BM… PT00…     21      114       54         147  18.4  38.9     14.3</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 54 more rows</span></span></span>
<span><span class="va">heatmaps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sharing_heatmap.html">sharing_heatmap</a></span><span class="op">(</span><span class="va">sharing_1_b</span><span class="op">)</span></span></code></pre></div>
<p>The function <code><a href="../reference/sharing_heatmap.html">sharing_heatmap()</a></code> automatically plots
sharing between 2 groups. There are several arguments to this function
that allow us to obtain heatmaps for the absolute sharing values or the
relative (percentage) values.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heatmaps</span><span class="op">$</span><span class="va">absolute</span></span></code></pre></div>
<p><img src="workflow_start_files/figure-html/unnamed-chunk-45-1.png" width="700"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heatmaps</span><span class="op">$</span><span class="va">on_g1</span></span></code></pre></div>
<p><img src="workflow_start_files/figure-html/unnamed-chunk-45-2.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heatmaps</span><span class="op">$</span><span class="va">on_union</span></span></code></pre></div>
<p><img src="workflow_start_files/figure-html/unnamed-chunk-45-3.png" width="700"></p>
<p>Beware of the fact that calculating all permutations takes longer
than calculating only distinct combinations, therefore if you don’t need
your results plotted or you have more than 2 groups to compare you
should stick with <code>minimal = TRUE</code> and
<code>include_self_comp = FALSE</code>.</p>
</div>
<div class="section level4">
<h4 id="scenario-2-single-input-data-frame-and-multiple-grouping-keys">SCENARIO 2: single input data frame and multiple grouping keys<a class="anchor" aria-label="anchor" href="#scenario-2-single-input-data-frame-and-multiple-grouping-keys"></a>
</h4>
<p>In the first scenario, groups were homogeneous, that is, they were
grouped all with the same key. In this other scenario we want to start
with data contained in just one data frame but we want to compare sets
of integrations that are grouped differently. To do this we give as
input a <strong>list of keys</strong> through the argument
<code>group_keys</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sharing_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span><span class="va">agg</span>,</span>
<span>    group_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        g1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>,</span>
<span>            <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span></span>
<span>        <span class="op">)</span>,</span>
<span>        g2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span><span class="op">)</span>,</span>
<span>        g3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CellMarker"</span>, <span class="st">"Tissue"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sharing_2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 32 × 12</span></span></span>
<span><span class="co">#&gt;    g1    g2    g3    shared count_g1 count_g2 count_g3 count_union  on_g1  on_g2</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> PT00… PT00… MNC_…     54       54      186      271         310 100    29.0  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> PT00… PT00… MNC_…     14       54      186      103         211  25.9   7.53 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> PT00… PT00… MNC_…      1       54      137      271         281   1.85  0.730</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> PT00… PT00… MNC_…      0       54      137      103         252   0     0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PT00… PT00… MNC_…    114      114      186      271         310 100    61.3  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> PT00… PT00… MNC_…     35      114      186      103         211  30.7  18.8  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PT00… PT00… MNC_…      2      114      137      271         281   1.75  1.46 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> PT00… PT00… MNC_…      2      114      137      103         292   1.75  1.46 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PT00… PT00… MNC_…      9       28      186      271         310  32.1   4.84 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PT00… PT00… MNC_…     28       28      186      103         211 100    15.1  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 22 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: on_g3 &lt;dbl&gt;, on_union &lt;dbl&gt;</span></span></span></code></pre></div>
<p>There are a few things to keep in mind in this case:</p>
<ul>
<li>The arguments <code>group_key</code> (notice the absence of plural),
<code>n_comp</code> and <code>include_self_comp</code> are ignored: the
number of comparisons is automatically detected by counting the provided
keys and a self comparison doesn’t make sense since group labels are
different</li>
<li>If you provide a list of identical keys or just one key you fall
back to scenario 1</li>
</ul>
</div>
<div class="section level4">
<h4 id="scenario-3-multiple-input-data-frame-and-single-grouping-key">SCENARIO 3: multiple input data frame and single grouping key<a class="anchor" aria-label="anchor" href="#scenario-3-multiple-input-data-frame-and-single-grouping-key"></a>
</h4>
<p>Providing multiple input data frames and the same grouping key is an
effective way to reduce the number of comparisons performed. Let’s make
an example: suppose we’re interested in comparing groups labelled by a
unique combination of <code>SubjectID</code>, <code>CellMarker</code>,
<code>Tissue</code> and <code>TimePoint</code>, but this time we want
the first group to contain only integrations relative to
<code>PT001_MNC_BM_0030</code> and the second group to contain only
integrations relative to <code>PT001_MNC_BM_0060</code>.</p>
<p>We’re going to filter the original data frame in order to obtain only
relevant data in 2 separated tables and then proceed by calling the
function.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">first_sample</span> <span class="op">&lt;-</span> <span class="va">agg</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">SubjectID</span> <span class="op">==</span> <span class="st">"PT001"</span>, <span class="va">CellMarker</span> <span class="op">==</span> <span class="st">"MNC"</span>, <span class="va">Tissue</span> <span class="op">==</span> <span class="st">"BM"</span>,</span>
<span>        <span class="va">TimePoint</span> <span class="op">==</span> <span class="st">"0030"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">second_sample</span> <span class="op">&lt;-</span> <span class="va">agg</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">SubjectID</span> <span class="op">==</span> <span class="st">"PT001"</span>, <span class="va">CellMarker</span> <span class="op">==</span> <span class="st">"MNC"</span>, <span class="va">Tissue</span> <span class="op">==</span> <span class="st">"BM"</span>,</span>
<span>        <span class="va">TimePoint</span> <span class="op">==</span> <span class="st">"0060"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">sharing_3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span><span class="va">first_sample</span>, <span class="va">second_sample</span>,</span>
<span>    group_key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>,</span>
<span>        <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    is_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    relative_is_sharing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    minimal <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sharing_3</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 9</span></span></span>
<span><span class="co">#&gt;   g1             g2    shared count_g1 count_g2 count_union on_g1 on_g2 on_union</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> PT001_MNC_BM_… PT00…     21       54      114         147  38.9  18.4     14.3</span></span></code></pre></div>
<p>Once again the arguments <code>n_comp</code> and
<code>include_self_comp</code> are ignored: the number of comparisons is
equal to the number of data frames in input.</p>
<p>To handle special limit cases, the output group ids are appended with
a dash and a number (<code>-1</code>, <code>-2</code>, …) that indicates
the data frame of origin: this is useful in the case group ids are
duplicated in the inputs. To understand better let’s make an
example:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sharing_3_a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span></span>
<span>    <span class="va">first_sample</span>, <span class="va">second_sample</span>,</span>
<span>    group_key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"CellMarker"</span>, <span class="st">"Tissue"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    is_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    relative_is_sharing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    minimal <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sharing_3_a</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 9</span></span></span>
<span><span class="co">#&gt;   g1       g2       shared count_g1 count_g2 count_union on_g1 on_g2 on_union</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> MNC_BM-1 MNC_BM-2     21       54      114         147  38.9  18.4     14.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> MNC_BM-2 MNC_BM-1     21      114       54         147  18.4  38.9     14.3</span></span></code></pre></div>
<p>As you can see, the IDs of group 1 and group 2 are duplicated and
without a suffix it would be impossible to know which one came from
which data frame. In this way we know that the group “MNC_BM-1” comes
from table 1 and has 54 ISs, while “MNC_BM-2” comes from the second
input table and has 114 ISs.</p>
</div>
<div class="section level4">
<h4 id="scenario-4-multiple-input-data-frame-and-multiple-grouping-keys">SCENARIO 4: multiple input data frame and multiple grouping
keys<a class="anchor" aria-label="anchor" href="#scenario-4-multiple-input-data-frame-and-multiple-grouping-keys"></a>
</h4>
<p>Finally, the most general scenario is when we have multiple data
frames with multiple keys. In this case the number of data frames must
be equal to the number of provided keys and grouping keys are applied in
order ( data frame 1 is grouped with key 1, data frame 2 is grouped with
key 2, and so on).</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="va">agg</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">TimePoint</span> <span class="op">==</span> <span class="st">"0030"</span><span class="op">)</span></span>
<span><span class="va">df2</span> <span class="op">&lt;-</span> <span class="va">agg</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">TimePoint</span> <span class="op">==</span> <span class="st">"0060"</span><span class="op">)</span></span>
<span><span class="va">df3</span> <span class="op">&lt;-</span> <span class="va">agg</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Tissue</span> <span class="op">==</span> <span class="st">"BM"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">keys</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    g1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>, <span class="st">"Tissue"</span><span class="op">)</span>,</span>
<span>    g2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SubjectID"</span>, <span class="st">"Tissue"</span><span class="op">)</span>,</span>
<span>    g3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>, <span class="st">"Tissue"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">sharing_4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span><span class="va">df1</span>, <span class="va">df2</span>, <span class="va">df3</span>, group_keys <span class="op">=</span> <span class="va">keys</span><span class="op">)</span></span>
<span><span class="va">sharing_4</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 32 × 12</span></span></span>
<span><span class="co">#&gt;    g1      g2    g3    shared count_g1 count_g2 count_g3 count_union on_g1 on_g2</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> PT001_… PT00… PT00…     21       54      114      147         147 38.9  18.4 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> PT001_… PT00… PT00…      0       54      114      126         271  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> PT001_… PT00… PT00…      8       54       59      147         175 14.8  13.6 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> PT001_… PT00… PT00…      0       54       59      126         229  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PT001_… PT00… PT00…      1       54       33      147         179  1.85  3.03</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> PT001_… PT00… PT00…      1       54       33      126         179  1.85  3.03</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PT001_… PT00… PT00…      0       54       18      147         165  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> PT001_… PT00… PT00…      0       54       18      126         185  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PT001_… PT00… PT00…      7       28      114      147         166 25     6.14</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PT001_… PT00… PT00…      0       28      114      126         260  0     0   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 22 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: on_g3 &lt;dbl&gt;, on_union &lt;dbl&gt;</span></span></span></code></pre></div>
<p>Notice that in this example the keys for g1 and g3 are the same,
meaning the labels of the groups are actually the same, however you
should remember that the groups contain a <strong>different set of
integration sites</strong> since they come from different data
frames.</p>
</div>
</div>
<div class="section level3">
<h3 id="plotting-sharing-results">Plotting sharing results<a class="anchor" aria-label="anchor" href="#plotting-sharing-results"></a>
</h3>
<p>When we have more than 2 comparisons it is convenient to plot them as
venn or euler diagrams. ISAnalytics has a fast way to do that through
the functions <code><a href="../reference/is_sharing.html">is_sharing()</a></code> and
<code><a href="../reference/sharing_venn.html">sharing_venn()</a></code>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sharing_5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_sharing.html">is_sharing</a></span><span class="op">(</span><span class="va">agg</span>,</span>
<span>    group_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        g1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span>,</span>
<span>            <span class="st">"Tissue"</span>, <span class="st">"TimePoint"</span></span>
<span>        <span class="op">)</span>,</span>
<span>        g2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SubjectID"</span>, <span class="st">"CellMarker"</span><span class="op">)</span>,</span>
<span>        g3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CellMarker"</span>, <span class="st">"Tissue"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>, table_for_venn <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sharing_5</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 32 × 13</span></span></span>
<span><span class="co">#&gt;    g1    g2    g3    shared count_g1 count_g2 count_g3 count_union  on_g1  on_g2</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> PT00… PT00… MNC_…     54       54      186      271         310 100    29.0  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> PT00… PT00… MNC_…     14       54      186      103         211  25.9   7.53 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> PT00… PT00… MNC_…      1       54      137      271         281   1.85  0.730</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> PT00… PT00… MNC_…      0       54      137      103         252   0     0    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> PT00… PT00… MNC_…    114      114      186      271         310 100    61.3  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> PT00… PT00… MNC_…     35      114      186      103         211  30.7  18.8  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> PT00… PT00… MNC_…      2      114      137      271         281   1.75  1.46 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> PT00… PT00… MNC_…      2      114      137      103         292   1.75  1.46 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> PT00… PT00… MNC_…      9       28      186      271         310  32.1   4.84 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> PT00… PT00… MNC_…     28       28      186      103         211 100    15.1  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 22 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: on_g3 &lt;dbl&gt;, on_union &lt;dbl&gt;, truth_tbl_venn &lt;list&gt;</span></span></span></code></pre></div>
<p>The argument <code>table_for_venn = TRUE</code> will add a new column
<code>truth_tbl_venn</code> that contains corresponding truth tables for
each row.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sharing_plots1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sharing_venn.html">sharing_venn</a></span><span class="op">(</span><span class="va">sharing_5</span>, row_range <span class="op">=</span> <span class="fl">1</span>, euler <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sharing_plots2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sharing_venn.html">sharing_venn</a></span><span class="op">(</span><span class="va">sharing_5</span>, row_range <span class="op">=</span> <span class="fl">1</span>, euler <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Say that we’re interested in plotting just the first row of our
sharing data frame. Then we can call the function
<code>sharing_venn</code> and specify in the <code>row_range</code>
argument the index 1. Note that this function requires the package
<em><a href="https://CRAN.R-project.org/package=eulerr" class="external-link">eulerr</a></em>
to work. The argument <code>euler</code> indicates if the function
should produce euler or venn diagrams instead.</p>
<p>Once obtained the lists of euler/venn objects we can plot them by
simply calling the function <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code>:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sharing_plots1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_start_files/figure-html/unnamed-chunk-52-1.png" width="700"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sharing_plots2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_start_files/figure-html/unnamed-chunk-52-2.png" width="700"></p>
<p>There are several options that can be set, for this please refer to
<a href="https://jolars.github.io/eulerr/reference/plot.euler.html" class="external-link">eulerr
docs</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p><code>R</code> session information.</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2023-05-09</span></span>
<span><span class="co">#&gt;  pandoc   2.19.2 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package      </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version</span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  backports       1.4.1   <span style="color: #555555;">2021-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bibtex          0.5.1   <span style="color: #555555;">2023-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  BiocManager     1.30.20 <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  BiocParallel    1.34.1  <span style="color: #555555;">2023-05-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocStyle     * 2.28.0  <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bit             4.0.5   <span style="color: #555555;">2022-11-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bit64           4.0.5   <span style="color: #555555;">2020-08-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bookdown        0.33    <span style="color: #555555;">2023-03-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib           0.4.2   <span style="color: #555555;">2022-12-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem          1.0.8   <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cellranger      1.1.0   <span style="color: #555555;">2016-07-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli             3.6.1   <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  codetools       0.2-19  <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  colorspace      2.1-0   <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  crayon          1.5.2   <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  crosstalk       1.2.0   <span style="color: #555555;">2021-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  curl            5.0.0   <span style="color: #555555;">2023-01-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  data.table      1.14.8  <span style="color: #555555;">2023-02-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  datamods        1.4.0   <span style="color: #555555;">2022-11-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc            1.4.2   <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest          0.6.31  <span style="color: #555555;">2022-12-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  doFuture        1.0.0   <span style="color: #555555;">2023-03-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dplyr           1.1.2   <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  DT            * 0.27    <span style="color: #555555;">2023-01-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ellipsis        0.3.2   <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  eulerr          7.0.0   <span style="color: #555555;">2022-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate        0.21    <span style="color: #555555;">2023-05-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fansi           1.0.4   <span style="color: #555555;">2023-01-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  farver          2.1.1   <span style="color: #555555;">2022-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap         1.1.1   <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  forcats         1.0.0   <span style="color: #555555;">2023-01-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  foreach         1.5.2   <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  foreign         0.8-84  <span style="color: #555555;">2022-12-06</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  fs              1.6.2   <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  future          1.32.0  <span style="color: #555555;">2023-03-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  future.apply    1.10.0  <span style="color: #555555;">2022-11-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics        0.1.3   <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggplot2         3.4.2   <span style="color: #555555;">2023-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  globals         0.16.2  <span style="color: #555555;">2022-11-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue            1.6.2   <span style="color: #555555;">2022-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable          0.3.3   <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtools          3.9.4   <span style="color: #555555;">2022-11-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  haven           2.5.2   <span style="color: #555555;">2023-02-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  highr           0.10    <span style="color: #555555;">2022-12-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  hms             1.1.3   <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools       0.5.5   <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets     1.6.2   <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httpuv          1.6.9   <span style="color: #555555;">2023-02-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httr            1.4.5   <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ISAnalytics   * 1.10.1  <span style="color: #555555;">2023-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  iterators       1.0.14  <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib       0.1.4   <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite        1.8.4   <span style="color: #555555;">2022-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr           1.42    <span style="color: #555555;">2023-01-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  later           1.3.1   <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lattice         0.21-8  <span style="color: #555555;">2023-04-05</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  lifecycle       1.0.3   <span style="color: #555555;">2022-10-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  listenv         0.9.0   <span style="color: #555555;">2022-12-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lubridate       1.9.2   <span style="color: #555555;">2023-02-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr        2.0.3   <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  memoise         2.0.1   <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mime            0.12    <span style="color: #555555;">2021-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mnormt          2.1.1   <span style="color: #555555;">2022-09-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  munsell         0.5.0   <span style="color: #555555;">2018-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  nlme            3.1-162 <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  openxlsx        4.2.5.2 <span style="color: #555555;">2023-02-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  parallelly      1.35.0  <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  phosphoricons   0.1.2   <span style="color: #555555;">2022-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pillar          1.9.0   <span style="color: #555555;">2023-03-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig       2.0.3   <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown         2.0.7   <span style="color: #555555;">2022-12-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">any (@2.0.7)</span></span></span>
<span><span class="co">#&gt;  plyr            1.8.8   <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  polyclip        1.10-4  <span style="color: #555555;">2022-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  polylabelr      0.2.0   <span style="color: #555555;">2020-04-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  progressr       0.13.0  <span style="color: #555555;">2023-01-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  promises        1.2.0.1 <span style="color: #555555;">2021-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  psych           2.3.3   <span style="color: #555555;">2023-03-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr           1.0.1   <span style="color: #555555;">2023-01-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R.methodsS3     1.8.2   <span style="color: #555555;">2022-06-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R.oo            1.25.0  <span style="color: #555555;">2022-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R.utils         2.12.2  <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R6              2.5.1   <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg            1.2.5   <span style="color: #555555;">2023-01-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp            1.0.10  <span style="color: #555555;">2023-01-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  reactable       0.4.4   <span style="color: #555555;">2023-03-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  readr           2.1.4   <span style="color: #555555;">2023-02-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  readxl          1.4.2   <span style="color: #555555;">2023-02-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RefManageR    * 1.4.0   <span style="color: #555555;">2022-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rio             0.5.29  <span style="color: #555555;">2021-11-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rlang           1.1.1   <span style="color: #555555;">2023-04-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown       2.21    <span style="color: #555555;">2023-03-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rprojroot       2.0.3   <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sass            0.4.6   <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales          1.2.1   <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sessioninfo   * 1.2.2   <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shiny           1.7.4   <span style="color: #555555;">2022-12-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shinybusy       0.3.1   <span style="color: #555555;">2022-05-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shinyWidgets    0.7.6   <span style="color: #555555;">2023-01-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringi         1.7.12  <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr         1.5.0   <span style="color: #555555;">2022-12-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  systemfonts     1.0.4   <span style="color: #555555;">2022-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping     0.3.6   <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble          3.2.1   <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyr           1.3.0   <span style="color: #555555;">2023-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect      1.2.0   <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  timechange      0.2.0   <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tzdb            0.3.0   <span style="color: #555555;">2022-03-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  utf8            1.2.3   <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs           0.6.2   <span style="color: #555555;">2023-04-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vroom           1.6.3   <span style="color: #555555;">2023-04-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr           2.5.0   <span style="color: #555555;">2022-03-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  writexl         1.4.2   <span style="color: #555555;">2023-01-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun            0.39    <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xml2            1.3.4   <span style="color: #555555;">2023-04-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xtable          1.8-4   <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  yaml            2.3.7   <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  zip             2.3.0   <span style="color: #555555;">2023-04-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/4.3.0/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/4.3.0/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.17/BiocStyle" class="external-link">BiocStyle</a></em>
(Oleś, 2023) with <em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em> (Xie,
2023) and <em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em>
(Allaire, Xie, Dervieux, McPherson, Luraschi, Ushey, Atkins, Wickham,
Cheng, Chang, and Iannone, 2023) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em>
(McLean, 2017).</p>
<p>
<a id="bib-allaire2023rmarkdown"></a><a href="#cite-allaire2023rmarkdown">[1]</a><cite>
J. Allaire, Y. Xie, C. Dervieux, et al. <em>rmarkdown: Dynamic Documents
for R</em>. R package version 2.21. 2023. URL:
<a href="https://github.com/rstudio/rmarkdown" class="external-link">https://github.com/rstudio/rmarkdown</a>.</cite>
</p>
<p>
<a id="bib-giuliospinozzivispa2"></a><a href="#cite-giuliospinozzivispa2">[2]</a><cite>
S. B. Giulio Spinozzi Andrea Calabria. “VISPA2: a scalable pipeline for
high-throughput identification and annotation of vector integration
sites”. In: <em>BMC Bioinformatics</em> (Nov. 25, 2017). DOI:
<a href="https://doi.org/10.1186/s12859-017-1937-9" class="external-link">10.1186/s12859-017-1937-9</a>.</cite>
</p>
<p>
<a id="bib-mclean2017refmanager"></a><a href="#cite-mclean2017refmanager">[3]</a><cite>
M. W. McLean. “RefManageR: Import and Manage BibTeX and BibLaTeX
References in R”. In: <em>The Journal of Open Source Software</em>
(2017). DOI:
<a href="https://doi.org/10.21105/joss.00338" class="external-link">10.21105/joss.00338</a>.</cite>
</p>
<p>
<a id="bib-ole2023biocstyle"></a><a href="#cite-ole2023biocstyle">[4]</a><cite>
A. Oleś. <em>BiocStyle: Standard styles for vignettes and other
Bioconductor documents</em>. R package version 2.28.0. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.BiocStyle" class="external-link">10.18129/B9.bioc.BiocStyle</a>.
URL:
<a href="https://bioconductor.org/packages/BiocStyle" class="external-link">https://bioconductor.org/packages/BiocStyle</a>.</cite>
</p>
<p>
<a id="bib-xie2023knitr"></a><a href="#cite-xie2023knitr">[5]</a><cite>
Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report
Generation in R</em>. R package version 1.42. 2023. URL:
<a href="https://yihui.org/knitr/" class="external-link">https://yihui.org/knitr/</a>.</cite>
</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Giulia Pais, Andrea Calabria, Giulio Spinozzi.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
